{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# enable automatic reloading of the notebook\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fad188da5ea44a2586d555f51cc0a9a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "11aac94806af4d0c9db02c19a80c11ab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n",
      "  return self.fget.__get__(instance, owner)()\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:392: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:397: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Load model directly\n",
    "from jsonformer import Jsonformer\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"AdaptLLM/medicine-chat\")\n",
    "model = AutoModelForCausalLM.from_pretrained(\"AdaptLLM/medicine-chat\", device_map=\"auto\", load_in_4bit=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/utils.py:1518: UserWarning: You have modified the pretrained model configuration to control generation. This is a deprecated strategy to control generation and will be removed soon, in a future version. Please use and modify the model generation configuration (see https://huggingface.co/docs/transformers/generation_strategies#default-text-generation-configuration )\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/bitsandbytes/nn/modules.py:226: UserWarning: Input type into Linear4bit is torch.float16, but bnb_4bit_compute_dtype=torch.float32 (default). This will lead to slow inference or training speed.\n",
      "  warnings.warn(f'Input type into Linear4bit is torch.float16, but bnb_4bit_compute_dtype=torch.float32 (default). This will lead to slow inference or training speed.')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### User Input:\n",
      "Please extract information from the text between !!! by using only categorical words or choosing from options Yes or No. Do not provide any additional text or information. The output must be in the following format:\n",
      "\n",
      "{\n",
      "    \"Sex\": \"\",\n",
      "    \"Age\": \"\",\n",
      "    \"Treatment\": \"\",\n",
      "    \"Patient had ECG done?\": \"\",\n",
      "    \"Patient had palpitations?\": \"\",\n",
      "    \"Patient had leg operation?\": \"\",\n",
      "    \"Rehabilitation time\": \"\"\n",
      "    \"Patient finished with treatment?\": \"\",\n",
      "    \"Patient died\": \"\",\n",
      "}\n",
      "\n",
      "!!!\n",
      "A 28-year-old previously healthy man presented with a 6-week history of palpitations.\n",
      "The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea.\n",
      "Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings.\n",
      "An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway.\n",
      "Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2).\n",
      "The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead).\n",
      "Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
      "The patient underwent an electrophysiologic study with mapping of the accessory pathway, followed by radiofrequency ablation (interruption of the pathway using the heat generated by electromagnetic waves at the tip of an ablation catheter).\n",
      "His post-ablation ECG showed a prolonged PR interval and an odd “second” QRS complex in leads III, aVF and V2–V4 (Fig.1Bottom), a consequence of abnormal impulse conduction in the “atrialized” right ventricle.\n",
      "The patient reported no recurrence of palpitations at follow-up 6 months after the ablation.\n",
      "!!!\n",
      "\n",
      "\n",
      "\n",
      "### Assistant Output:\n",
      "Sex: Male\n",
      "Age: 28 years old\n",
      "Treatment: Electrophysiologic study and radiofrequency ablation of accessory pathway\n",
      "Patient had ECG done: Yes\n",
      "Patient had palpitations: Yes\n",
      "Patient had leg operation: No\n",
      "Rehabilitation time: 6 months\n",
      "Patient finished with treatment: Yes\n",
      "Patient died: No\n"
     ]
    }
   ],
   "source": [
    "user_input = '''Please extract information from the text between !!! by using only categorical words or choosing from options Yes or No. Do not provide any additional text or information. The output must be in the following format:\n",
    "\n",
    "{\n",
    "    \"Sex\": \"\",\n",
    "    \"Age\": \"\",\n",
    "    \"Treatment\": \"\",\n",
    "    \"Patient had ECG done?\": \"\",\n",
    "    \"Patient had palpitations?\": \"\",\n",
    "    \"Patient had leg operation?\": \"\",\n",
    "    \"Rehabilitation time\": \"\"\n",
    "    \"Patient finished with treatment?\": \"\",\n",
    "    \"Patient died\": \"\",\n",
    "}\n",
    "\n",
    "!!!\n",
    "A 28-year-old previously healthy man presented with a 6-week history of palpitations.\n",
    "The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea.\n",
    "Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings.\n",
    "An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway.\n",
    "Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2).\n",
    "The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead).\n",
    "Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
    "The patient underwent an electrophysiologic study with mapping of the accessory pathway, followed by radiofrequency ablation (interruption of the pathway using the heat generated by electromagnetic waves at the tip of an ablation catheter).\n",
    "His post-ablation ECG showed a prolonged PR interval and an odd “second” QRS complex in leads III, aVF and V2–V4 (Fig.1Bottom), a consequence of abnormal impulse conduction in the “atrialized” right ventricle.\n",
    "The patient reported no recurrence of palpitations at follow-up 6 months after the ablation.\n",
    "!!!\n",
    "\n",
    "'''\n",
    "\n",
    "# Apply the prompt template and system prompt of LLaMA-2-Chat demo for chat models (NOTE: NO prompt template is required for base models!)\n",
    "our_system_prompt = \"\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n\" # Please do NOT change this\n",
    "\n",
    "prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{user_input} [/INST]\"\n",
    "\n",
    "# # NOTE:\n",
    "# # If you want to apply your own system prompt, please integrate it into the instruction part following our system prompt like this:\n",
    "# your_system_prompt = \"Please, answer this question faithfully.\"\n",
    "# prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{your_system_prompt}\\n{user_input} [/INST]\"\n",
    "\n",
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{user_input}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_input = '''Please extract information from the text between !!! by using only categorical words or choosing from options Yes or No. Do not provide any additional text or information. Please provide the response in the form of a Python list. It should begin with “[“ and end with “]”:\n",
    "\n",
    "[\n",
    "    \"Sex\": \"\",\n",
    "    \"Age\": \"\",\n",
    "    \"Treatment\": \"\",\n",
    "    \"Patient had ECG done?\": \"\",\n",
    "    \"Patient had palpitations?\": \"\",\n",
    "    \"Patient had leg operation?\": \"\",\n",
    "    \"Rehabilitation time\": \"\"\n",
    "    \"Patient finished with treatment?\": \"\",\n",
    "    \"Patient died\": \"\",\n",
    "]\n",
    "\n",
    "!!!\n",
    "A 28-year-old previously healthy man presented with a 6-week history of palpitations.\n",
    "The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea.\n",
    "Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings.\n",
    "An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway.\n",
    "Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2).\n",
    "The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead).\n",
    "Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
    "The patient underwent an electrophysiologic study with mapping of the accessory pathway, followed by radiofrequency ablation (interruption of the pathway using the heat generated by electromagnetic waves at the tip of an ablation catheter).\n",
    "His post-ablation ECG showed a prolonged PR interval and an odd “second” QRS complex in leads III, aVF and V2–V4 (Fig.1Bottom), a consequence of abnormal impulse conduction in the “atrialized” right ventricle.\n",
    "The patient reported no recurrence of palpitations at follow-up 6 months after the ablation.\n",
    "!!!\n",
    "\n",
    "'''\n",
    "\n",
    "# Apply the prompt template and system prompt of LLaMA-2-Chat demo for chat models (NOTE: NO prompt template is required for base models!)\n",
    "our_system_prompt = \"\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n\" # Please do NOT change this\n",
    "\n",
    "prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{user_input} [/INST]\"\n",
    "\n",
    "# # NOTE:\n",
    "# # If you want to apply your own system prompt, please integrate it into the instruction part following our system prompt like this:\n",
    "# your_system_prompt = \"Please, answer this question faithfully.\"\n",
    "# prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{your_system_prompt}\\n{user_input} [/INST]\"\n",
    "\n",
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{user_input}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "user_input = '''Please extract information from the text between !!! by using only categorical words or choosing from options Yes or No. Do not provide any additional text or information.\n",
    "\n",
    "!!!\n",
    "A 28-year-old previously healthy man presented with a 6-week history of palpitations.\n",
    "The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea.\n",
    "Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings.\n",
    "An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway.\n",
    "Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2).\n",
    "The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead).\n",
    "Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
    "The patient underwent an electrophysiologic study with mapping of the accessory pathway, followed by radiofrequency ablation (interruption of the pathway using the heat generated by electromagnetic waves at the tip of an ablation catheter).\n",
    "His post-ablation ECG showed a prolonged PR interval and an odd “second” QRS complex in leads III, aVF and V2–V4 (Fig.1Bottom), a consequence of abnormal impulse conduction in the “atrialized” right ventricle.\n",
    "The patient reported no recurrence of palpitations at follow-up 6 months after the ablation.\n",
    "!!!\n",
    "\n",
    "Format the information using the following schema: \n",
    "'''\n",
    "#prompt = \"Generate a person's information based on the following schema:\"\n",
    "json_schema = {\n",
    "    \"type\": \"object\",\n",
    "    \"properties\": {\n",
    "        \"sex\": {\"type\": \"string\"},\n",
    "        \"age\": {\"type\": \"string\"}, # for some reason, \"type\": \"number\" does not work with the medical-chat model (investigate?)\n",
    "        \"Treatment\": {\"type\": \"string\"},\n",
    "        \"Patient had ECG done\": {\"type\": \"boolean\"},\n",
    "        \"Patient died\": {\"type\": \"boolean\"},\n",
    "    }\n",
    "}\n",
    "\n",
    "# Apply the prompt template and system prompt of LLaMA-2-Chat demo for chat models (NOTE: NO prompt template is required for base models!)\n",
    "our_system_prompt = \"\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n\" # Please do NOT change this\n",
    "\n",
    "prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{user_input} [/INST]\"\n",
    "\n",
    "\n",
    "jsonformer = Jsonformer(model, tokenizer, json_schema, user_input)\n",
    "generated_data = jsonformer()\n",
    "\n",
    "print(generated_data)\n",
    "\n",
    "# # NOTE:\n",
    "# # If you want to apply your own system prompt, please integrate it into the instruction part following our system prompt like this:\n",
    "# your_system_prompt = \"Please, answer this question faithfully.\"\n",
    "# prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{your_system_prompt}\\n{user_input} [/INST]\"\n",
    "\n",
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{user_input}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonformer import Jsonformer\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(\"databricks/dolly-v2-3b\", device_map=\"auto\", load_in_4bit=True)\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"databricks/dolly-v2-3b\")\n",
    "\n",
    "json_schema = {\n",
    "    \"type\": \"object\",\n",
    "    \"properties\": {\n",
    "        \"sex\": {\"type\": \"string\"},\n",
    "        \"age\": {\"type\": \"number\"},\n",
    "        \"Treatment\": {\"type\": \"string\"},\n",
    "        \"Patient had ECG done\": {\"type\": \"boolean\"},\n",
    "        \"Patient died\": {\"type\": \"boolean\"},\n",
    "    }\n",
    "}\n",
    "\n",
    "prompt = \"\"\"Please extract information from the text between !!! by using only categorical words or choosing from options Yes or No. Do not provide any additional text or information. Please generate a person's information based on the following schema:\n",
    "\n",
    "\n",
    "!!!\n",
    "A 28-year-old previously healthy man presented with a 6-week history of palpitations.\n",
    "The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea.\n",
    "Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings.\n",
    "An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway.\n",
    "Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2).\n",
    "The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead).\n",
    "Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
    "The patient underwent an electrophysiologic study with mapping of the accessory pathway, followed by radiofrequency ablation (interruption of the pathway using the heat generated by electromagnetic waves at the tip of an ablation catheter).\n",
    "His post-ablation ECG showed a prolonged PR interval and an odd “second” QRS complex in leads III, aVF and V2–V4 (Fig.1Bottom), a consequence of abnormal impulse conduction in the “atrialized” right ventricle.\n",
    "The patient reported no recurrence of palpitations at follow-up 6 months after the ablation.\n",
    "!!!\n",
    "\"\"\"\n",
    "jsonformer = Jsonformer(model, tokenizer, json_schema, prompt)\n",
    "generated_data = jsonformer()\n",
    "\n",
    "print(generated_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = os.getcwd()\n",
    "\n",
    "parent_directory_path = os.path.dirname(path)\n",
    "\n",
    "data_directory = \"data/maccrobat/MACCROBAT2020\"\n",
    "\n",
    "path = os.path.join(parent_directory_path, data_directory)\n",
    "file_list = os.listdir(path)\n",
    "\n",
    "# Retrive txt and ann files\n",
    "txt_files = [file for file in file_list if file.endswith('.txt')]\n",
    "ann_files = [file for file in file_list if file.endswith('.ann')]\n",
    "\n",
    "# Tuple together same txt and ann files\n",
    "file_tuples = [(txt, txt[:-4] + '.ann') for txt in txt_files if txt[:-4] + '.ann' in ann_files]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('26444414.txt', '26444414.ann'),\n",
       " ('28079821.txt', '28079821.ann'),\n",
       " ('23033875.txt', '23033875.ann'),\n",
       " ('28767567.txt', '28767567.ann'),\n",
       " ('27741115.txt', '27741115.ann'),\n",
       " ('25572898.txt', '25572898.ann'),\n",
       " ('25246819.txt', '25246819.ann'),\n",
       " ('28353604.txt', '28353604.ann'),\n",
       " ('25410034.txt', '25410034.ann'),\n",
       " ('26530965.txt', '26530965.ann'),\n",
       " ('20146086.txt', '20146086.ann'),\n",
       " ('21308977.txt', '21308977.ann'),\n",
       " ('22520024.txt', '22520024.ann'),\n",
       " ('28353613.txt', '28353613.ann'),\n",
       " ('27218632.txt', '27218632.ann'),\n",
       " ('18666334.txt', '18666334.ann'),\n",
       " ('21477357.txt', '21477357.ann'),\n",
       " ('24526194.txt', '24526194.ann'),\n",
       " ('26285706.txt', '26285706.ann'),\n",
       " ('27928148.txt', '27928148.ann'),\n",
       " ('22791498.txt', '22791498.ann'),\n",
       " ('21527041.txt', '21527041.ann'),\n",
       " ('28193213.txt', '28193213.ann'),\n",
       " ('24518095.txt', '24518095.ann'),\n",
       " ('19860925.txt', '19860925.ann'),\n",
       " ('26405496.txt', '26405496.ann'),\n",
       " ('24957905.txt', '24957905.ann'),\n",
       " ('25210224.txt', '25210224.ann'),\n",
       " ('26469535.txt', '26469535.ann'),\n",
       " ('26670309.txt', '26670309.ann'),\n",
       " ('28202869.txt', '28202869.ann'),\n",
       " ('28120581.txt', '28120581.ann'),\n",
       " ('21923918.txt', '21923918.ann'),\n",
       " ('21067996.txt', '21067996.ann'),\n",
       " ('28115731.txt', '28115731.ann'),\n",
       " ('27683825.txt', '27683825.ann'),\n",
       " ('27661040.txt', '27661040.ann'),\n",
       " ('26327988.txt', '26327988.ann'),\n",
       " ('22665582.txt', '22665582.ann'),\n",
       " ('26474553.txt', '26474553.ann'),\n",
       " ('26361431.txt', '26361431.ann'),\n",
       " ('28353569.txt', '28353569.ann'),\n",
       " ('28403092.txt', '28403092.ann'),\n",
       " ('23124805.txt', '23124805.ann'),\n",
       " ('28121940.txt', '28121940.ann'),\n",
       " ('19214295.txt', '19214295.ann'),\n",
       " ('24781756.txt', '24781756.ann'),\n",
       " ('26629302.txt', '26629302.ann'),\n",
       " ('25793030.txt', '25793030.ann'),\n",
       " ('15939911.txt', '15939911.ann'),\n",
       " ('23076693.txt', '23076693.ann'),\n",
       " ('27990013.txt', '27990013.ann'),\n",
       " ('28173879.txt', '28173879.ann'),\n",
       " ('26350418.txt', '26350418.ann'),\n",
       " ('28190872.txt', '28190872.ann'),\n",
       " ('27793101.txt', '27793101.ann'),\n",
       " ('25858931.txt', '25858931.ann'),\n",
       " ('28250304.txt', '28250304.ann'),\n",
       " ('19860006.txt', '19860006.ann'),\n",
       " ('28154281.txt', '28154281.ann'),\n",
       " ('26336183.txt', '26336183.ann'),\n",
       " ('27057898.txt', '27057898.ann'),\n",
       " ('25024632.txt', '25024632.ann'),\n",
       " ('26309459.txt', '26309459.ann'),\n",
       " ('28292056.txt', '28292056.ann'),\n",
       " ('26584481.txt', '26584481.ann'),\n",
       " ('26175648.txt', '26175648.ann'),\n",
       " ('23897372.txt', '23897372.ann'),\n",
       " ('25410883.txt', '25410883.ann'),\n",
       " ('28151916.txt', '28151916.ann'),\n",
       " ('23678274.txt', '23678274.ann'),\n",
       " ('25884600.txt', '25884600.ann'),\n",
       " ('23468586.txt', '23468586.ann'),\n",
       " ('28248891.txt', '28248891.ann'),\n",
       " ('28154669.txt', '28154669.ann'),\n",
       " ('28239141.txt', '28239141.ann'),\n",
       " ('26106249.txt', '26106249.ann'),\n",
       " ('23155491.txt', '23155491.ann'),\n",
       " ('27064109.txt', '27064109.ann'),\n",
       " ('26313770.txt', '26313770.ann'),\n",
       " ('27196481.txt', '27196481.ann'),\n",
       " ('28202865.txt', '28202865.ann'),\n",
       " ('26264228.txt', '26264228.ann'),\n",
       " ('20977862.txt', '20977862.ann'),\n",
       " ('26457578.txt', '26457578.ann'),\n",
       " ('19860007.txt', '19860007.ann'),\n",
       " ('27842605.txt', '27842605.ann'),\n",
       " ('27821134.txt', '27821134.ann'),\n",
       " ('28090049.txt', '28090049.ann'),\n",
       " ('27059701.txt', '27059701.ann'),\n",
       " ('28320420.txt', '28320420.ann'),\n",
       " ('19009665.txt', '19009665.ann'),\n",
       " ('28595573.txt', '28595573.ann'),\n",
       " ('26714786.txt', '26714786.ann'),\n",
       " ('27846860.txt', '27846860.ann'),\n",
       " ('22218279.txt', '22218279.ann'),\n",
       " ('28154700.txt', '28154700.ann'),\n",
       " ('28403086.txt', '28403086.ann'),\n",
       " ('26656340.txt', '26656340.ann'),\n",
       " ('22814979.txt', '22814979.ann'),\n",
       " ('27906105.txt', '27906105.ann'),\n",
       " ('28353596.txt', '28353596.ann'),\n",
       " ('28321073.txt', '28321073.ann'),\n",
       " ('18561524.txt', '18561524.ann'),\n",
       " ('26266396.txt', '26266396.ann'),\n",
       " ('25023062.txt', '25023062.ann'),\n",
       " ('28033278.txt', '28033278.ann'),\n",
       " ('28196820.txt', '28196820.ann'),\n",
       " ('18815636.txt', '18815636.ann'),\n",
       " ('28151882.txt', '28151882.ann'),\n",
       " ('21505579.txt', '21505579.ann'),\n",
       " ('28296775.txt', '28296775.ann'),\n",
       " ('28103924.txt', '28103924.ann'),\n",
       " ('25155594.txt', '25155594.ann'),\n",
       " ('28353556.txt', '28353556.ann'),\n",
       " ('28265107.txt', '28265107.ann'),\n",
       " ('21720478.txt', '21720478.ann'),\n",
       " ('28272214.txt', '28272214.ann'),\n",
       " ('25661749.txt', '25661749.ann'),\n",
       " ('27974938.txt', '27974938.ann'),\n",
       " ('28321071.txt', '28321071.ann'),\n",
       " ('18416479.txt', '18416479.ann'),\n",
       " ('28202862.txt', '28202862.ann'),\n",
       " ('28538413.txt', '28538413.ann'),\n",
       " ('25934795.txt', '25934795.ann'),\n",
       " ('21254744.txt', '21254744.ann'),\n",
       " ('28207542.txt', '28207542.ann'),\n",
       " ('26692730.txt', '26692730.ann'),\n",
       " ('20671919.txt', '20671919.ann'),\n",
       " ('19610147.txt', '19610147.ann'),\n",
       " ('25139918.txt', '25139918.ann'),\n",
       " ('25926582.txt', '25926582.ann'),\n",
       " ('27773410.txt', '27773410.ann'),\n",
       " ('19307547.txt', '19307547.ann'),\n",
       " ('18258107.txt', '18258107.ann'),\n",
       " ('26395443.txt', '26395443.ann'),\n",
       " ('23035161.txt', '23035161.ann'),\n",
       " ('27980272.txt', '27980272.ann'),\n",
       " ('24043987.txt', '24043987.ann'),\n",
       " ('28321070.txt', '28321070.ann'),\n",
       " ('28403099.txt', '28403099.ann'),\n",
       " ('16778410.txt', '16778410.ann'),\n",
       " ('26523273.txt', '26523273.ann'),\n",
       " ('28154287.txt', '28154287.ann'),\n",
       " ('19816630.txt', '19816630.ann'),\n",
       " ('27004009.txt', '27004009.ann'),\n",
       " ('26675562.txt', '26675562.ann'),\n",
       " ('22781096.txt', '22781096.ann'),\n",
       " ('24161539.txt', '24161539.ann'),\n",
       " ('23312850.txt', '23312850.ann'),\n",
       " ('28100279.txt', '28100279.ann'),\n",
       " ('27100441.txt', '27100441.ann'),\n",
       " ('26361640.txt', '26361640.ann'),\n",
       " ('28272235.txt', '28272235.ann'),\n",
       " ('25743872.txt', '25743872.ann'),\n",
       " ('26664317.txt', '26664317.ann'),\n",
       " ('25853982.txt', '25853982.ann'),\n",
       " ('25293719.txt', '25293719.ann'),\n",
       " ('21129213.txt', '21129213.ann'),\n",
       " ('28559815.txt', '28559815.ann'),\n",
       " ('25295501.txt', '25295501.ann'),\n",
       " ('28057913.txt', '28057913.ann'),\n",
       " ('28353561.txt', '28353561.ann'),\n",
       " ('27998312.txt', '27998312.ann'),\n",
       " ('28151860.txt', '28151860.ann'),\n",
       " ('22719160.txt', '22719160.ann'),\n",
       " ('27842595.txt', '27842595.ann'),\n",
       " ('26445413.txt', '26445413.ann'),\n",
       " ('28383413.txt', '28383413.ann'),\n",
       " ('25370695.txt', '25370695.ann'),\n",
       " ('28100235.txt', '28100235.ann'),\n",
       " ('27130218.txt', '27130218.ann'),\n",
       " ('23242090.txt', '23242090.ann'),\n",
       " ('25759562.txt', '25759562.ann'),\n",
       " ('28353588.txt', '28353588.ann'),\n",
       " ('28250406.txt', '28250406.ann'),\n",
       " ('28353558.txt', '28353558.ann'),\n",
       " ('24294397.txt', '24294397.ann'),\n",
       " ('22515939.txt', '22515939.ann'),\n",
       " ('22514576.txt', '22514576.ann'),\n",
       " ('24654246.txt', '24654246.ann'),\n",
       " ('25721834.txt', '25721834.ann'),\n",
       " ('28422883.txt', '28422883.ann'),\n",
       " ('28296749.txt', '28296749.ann'),\n",
       " ('26683938.txt', '26683938.ann'),\n",
       " ('18236639.txt', '18236639.ann'),\n",
       " ('21672201.txt', '21672201.ann'),\n",
       " ('28248858.txt', '28248858.ann'),\n",
       " ('23864579.txt', '23864579.ann'),\n",
       " ('26257516.txt', '26257516.ann'),\n",
       " ('27904130.txt', '27904130.ann'),\n",
       " ('18787726.txt', '18787726.ann'),\n",
       " ('23077697.txt', '23077697.ann'),\n",
       " ('27749582.txt', '27749582.ann'),\n",
       " ('27980261.txt', '27980261.ann'),\n",
       " ('17803823.txt', '17803823.ann'),\n",
       " ('24898994.txt', '24898994.ann'),\n",
       " ('26216058.txt', '26216058.ann'),\n",
       " ('26228535.txt', '26228535.ann'),\n",
       " ('28216610.txt', '28216610.ann')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_tuples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41\n",
      "{'Area', 'Personal_background', 'Subject', 'Quantitative_concept', 'Qualitative_concept', 'Medication', 'Duration', 'Coreference', 'Dosage', 'History', 'Outcome', 'Distance', 'Severity', 'Mass', 'Height', 'Frequency', 'Biological_structure', 'Date', 'Activity', 'Color', 'Detailed_description', 'Nonbiological_location', 'Sign_symptom', 'Disease_disorder', 'Weight', 'Shape', 'Age', 'Texture', 'Administration', 'Sex', 'Clinical_event', 'Time', 'Family_history', 'Other_event', 'Lab_value', 'Other_entity', 'Occupation', 'Diagnostic_procedure', 'Biological_attribute', 'Therapeutic_procedure', 'Volume'}\n"
     ]
    }
   ],
   "source": [
    "label_types = []\n",
    "\n",
    "\n",
    "for Annfile in ann_files:\n",
    "    pathToAnnFile = os.path.join(path, Annfile)\n",
    "    Annfile = open(pathToAnnFile, \"r\")\n",
    "    allAnnLines = [re.split(r'\\t+', tag.rstrip('\\t')) for tag in Annfile if tag[0][0].startswith(('T'))]\n",
    "\n",
    "    for annLine in allAnnLines:\n",
    "        label_types.append(annLine[1].split()[0])\n",
    "\n",
    "label_types = set(label_types)\n",
    "print(len(label_types))\n",
    "print(label_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mySet = {\n",
    "    'Diagnostic_procedure',\n",
    "    'Sign_symptom',\n",
    "    'Biological_structure',\n",
    "    'Detailed_description',\n",
    "    'Age',\n",
    "    'Lab_value'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53\n",
      "[['T1', 'Age 2 13', '58-year-old\\n'], ['T2', 'Sex 14 17', 'man\\n'], ['T3', 'Sign_symptom 42 57', 'general fatigue\\n'], ['T4', 'Sign_symptom 69 75', 'anemia\\n'], ['T5', 'Severity 62 68', 'severe\\n'], ['T6', 'Duration 80 94', 'several months\\n'], ['T7', 'Diagnostic_procedure 100 117', 'hemoglobin levels\\n'], ['T8', 'Lab_value 123 131', '6.6 g/dl\\n'], ['T9', 'History 167 185', 'no medical history\\n'], ['T10', 'History 190 215', 'did not take any medicine\\n'], ['T11', 'Diagnostic_procedure 217 243', 'Esophagogastroduodenoscopy\\n'], ['T12', 'Diagnostic_procedure 248 259', 'colonoscopy\\n'], ['T13', 'Sign_symptom 291 299', 'bleeding\\n'], ['T14', 'Diagnostic_procedure 311 330', 'computer tomography\\n'], ['T15', 'Biological_structure 301 310', 'Abdominal\\n'], ['T16', 'Sign_symptom 361 366', 'tumor\\n'], ['T17', 'Biological_structure 374 389', 'small intestine\\n'], ['T18', 'Detailed_description 347 360', 'hypervascular\\n'], ['T19', 'Distance 342 346', '2-cm\\n'], ['T20', 'Diagnostic_procedure 404 407', 'DBE\\n'], ['T21', 'Sign_symptom 464 470', 'lesion\\n'], ['T22', 'Detailed_description 453 463', 'tumor-like\\n'], ['T23', 'Biological_structure 442 452', 'submucosal\\n'], ['T24', 'Color 433 440', 'reddish\\n'], ['T25', 'Distance 419 432', '2-cm-diameter\\n'], ['T26', 'Biological_structure 502 509', 'jejunum\\n'], ['T27', 'Detailed_description 476 494', 'surface ulceration\\n'], ['T28', 'Biological_structure 525 560', '20 cm away from the Treitz ligament\\n'], ['T29', 'Diagnostic_procedure 589 595', 'biopsy\\n'], ['T30', 'Therapeutic_procedure 838 847', 'resection\\n'], ['T31', 'Detailed_description 828 837', 'segmental\\n'], ['T32', 'Detailed_description 806 827', 'laparoscopic-assisted\\n'], ['T33', 'Biological_structure 855 862', 'jejunum\\n'], ['T34', 'Detailed_description 872 897', 'dissection of lymph nodes\\n'], ['T35', 'Disease_disorder 785 791', 'cancer\\n'], ['T36', 'Disease_disorder 762 780', 'malignant lymphoma\\n'], ['T37', 'Disease_disorder 723 753', 'gastrointestinal stromal tumor\\n'], ['T38', 'Disease_disorder 704 721', 'small bowel tumor\\n'], ['T39', 'Disease_disorder 755 759', 'GIST\\n'], ['T40', 'Area 957 967', '19\\u2009×\\u200916 mm\\n'], ['T41', 'Sign_symptom 927 932', 'tumor\\n'], ['T42', 'Detailed_description 918 926', 'resected\\n'], ['T43', 'Diagnostic_procedure 989 998', 'Histology\\n'], ['T44', 'Diagnostic_procedure 899 910', 'Examination\\n'], ['T45', 'Disease_disorder 1097 1099', 'PG\\n'], ['T46', 'Sign_symptom 1051 1069', 'granulation tissue\\n'], ['T47', 'Sign_symptom 1012 1046', 'proliferation of blood capillaries\\n'], ['T48', 'Clinical_event 1125 1135', 'discharged\\n'], ['T49', 'Date 1139 1158', 'postoperative day 9\\n'], ['T50', 'Sign_symptom 1167 1179', 'complication\\n'], ['T51', 'Sign_symptom 1188 1194', 'anemia\\n'], ['T52', 'Lab_value 1195 1203', 'improved\\n'], ['T53', 'Medication 1235 1244', 'oral iron\\n']]\n",
      "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
      "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
      "He had no medical history and did not take any medicine.\n",
      "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
      "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
      "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
      "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
      "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
      "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
      "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
      "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for txtAnnPair in file_tuples[:1]:\n",
    "    pathToTxtFile = os.path.join(path, txtAnnPair[0])\n",
    "    pathToAnnFile = os.path.join(path, txtAnnPair[1])\n",
    "\n",
    "    Txtfile = open(pathToTxtFile, \"r\")\n",
    "    Txtfile = Txtfile.readlines()\n",
    "    Txtfile = \"\".join(Txtfile)\n",
    "    \n",
    "    Annfile = open(pathToAnnFile, \"r\")\n",
    "    allAnnLines = [re.split(r'\\t+', tag.rstrip('\\t')) for tag in Annfile if tag[0][0].startswith(('T'))]\n",
    "    print(len(allAnnLines))\n",
    "\n",
    "\n",
    "    removed = []\n",
    "    previous = -10\n",
    "    for annLine in allAnnLines.copy():\n",
    "        currentStart = int((annLine[1].split()[1]))\n",
    "        if currentStart == previous:\n",
    "            removed.append(annLine)\n",
    "            allAnnLines.remove(annLine)\n",
    "        previous = currentStart\n",
    "\n",
    "    print(allAnnLines)\n",
    "    print(Txtfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IDEAS for better results\n",
    "\n",
    "On generated and words in dataset perfrom lematization!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3aaadc63b7424a938d5a206fea0608fd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:389: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:394: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#from jsonformer import Jsonformer\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(\"AdaptLLM/medicine-chat\", device_map=\"auto\", load_in_4bit=True)\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"AdaptLLM/medicine-chat\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Both `max_new_tokens` (=150) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### User Input:\n",
      "Identify sign symptoms mentioned in text between !!!. For every detected sign symptom extract only one word and do not provide additional information.\n",
      "\n",
      "!!!\n",
      "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
      "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
      "He had no medical history and did not take any medicine.\n",
      "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
      "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
      "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
      "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
      "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
      "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
      "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
      "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
      "!!!\n",
      "\n",
      "\n",
      "### Assistant Output:\n",
      "The sign symptoms mentioned in the text are:\n",
      "\n",
      "1. General fatigue\n",
      "2. Severe anemia\n",
      "3. No medical history or medication\n",
      "4. No significant bleeding detected during esophagogastroduodenoscopy or colonoscopy\n",
      "5. Hypervascular tumor detected in the small intestine during abdominal computer tomography\n",
      "6. Reddish, submucosal tumor-like lesion with surface ulceration detected in the jejunum during oral DBE\n",
      "7. Diagnosis of small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer\n",
      "8. Lap\n"
     ]
    }
   ],
   "source": [
    "user_input = \"\"\"Identify sign symptoms mentioned in text between !!!. For every detected sign symptom extract only one word and do not provide additional information.\n",
    "\n",
    "!!!\n",
    "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
    "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
    "He had no medical history and did not take any medicine.\n",
    "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
    "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
    "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
    "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
    "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
    "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
    "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
    "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
    "!!!\n",
    "\"\"\"\n",
    "our_system_prompt = \"\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n\" # Please do NOT change this\n",
    "\n",
    "prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{user_input} [/INST]\"\n",
    "\n",
    "# # NOTE:\n",
    "# # If you want to apply your own system prompt, please integrate it into the instruction part following our system prompt like this:\n",
    "# your_system_prompt = \"Please, answer this question faithfully.\"\n",
    "# prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{your_system_prompt}\\n{user_input} [/INST]\"\n",
    "\n",
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096, max_new_tokens=150)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{user_input}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### User Input:\n",
      "Identify and sign symptoms mentioned in text between !!!. For every detected sign symptom extract only one word and do not provide additional information.\n",
      "\n",
      "\n",
      "!!!\n",
      "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
      "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
      "He had no medical history and did not take any medicine.\n",
      "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
      "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
      "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
      "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
      "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
      "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
      "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
      "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
      "!!!\n",
      "\n",
      "\n",
      "### Assistant Output:\n",
      "The text mentions several symptoms that the 58-year-old man was experiencing, including general fatigue and severe anemia. The text also mentions that the man did not have any medical history or take any medication. The text goes on to describe the results of various medical tests, including esophagogastroduodenoscopy and colonoscopy, which did not reveal any significant bleeding. Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine, which was detected by oral DBE. The text also mentions that the man was diagnosed with a small bowel tumor, which was suspected to be a gastrointestinal stromal tumor, malignant lymphoma, or cancer. The text goes on to describe the treatment plan, which involved laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes. The text also mentions that the resected tumor measured 19 × 16 mm in diameter and was consistent with PG. Finally, the text mentions that the patient was discharged without complication and his anemia improved gradually after surgery.\n"
     ]
    }
   ],
   "source": [
    "user_input = \"\"\"Identify and sign symptoms mentioned in text between !!!. For every detected sign symptom extract only one word and do not provide additional information.\n",
    "\n",
    "\n",
    "!!!\n",
    "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
    "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
    "He had no medical history and did not take any medicine.\n",
    "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
    "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
    "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
    "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
    "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
    "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
    "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
    "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
    "!!!\n",
    "\"\"\"\n",
    "our_system_prompt = \"\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n\" # Please do NOT change this\n",
    "\n",
    "prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{user_input} [/INST]\"\n",
    "\n",
    "# # NOTE:\n",
    "# # If you want to apply your own system prompt, please integrate it into the instruction part following our system prompt like this:\n",
    "# your_system_prompt = \"Please, answer this question faithfully.\"\n",
    "# prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{your_system_prompt}\\n{user_input} [/INST]\"\n",
    "\n",
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{user_input}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "generator = transformers.pipeline(\n",
    "    model = model,\n",
    "    tokenizer=tokenizer,\n",
    "    return_full_text = True, # langchain expects full text\n",
    "    task='text-generation',\n",
    "    #stopping_criteria=stopping_criteria, # without this model rambles during chat\n",
    "    temperature=0.1, # 'randomness' of outputs, 0.0 is the min and 1.0 is the max\n",
    "    max_new_tokens=512, # max number of tokens to generate in the output\n",
    "    repetition_penalty=1.1 # without this output begins repeating\n",
    ")\n",
    "\n",
    "llm = HuggingFacePipeline(pipeline=generator)\n",
    "# creating prompt for large language model\n",
    "pre_prompt = \"\"\"[INST] <<SYS>>\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe. Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n<</SYS>>\\n\\nGenerate the next agent response by answering the question. Answer it as succinctly as possible. You are provided several documents with titles. If the answer comes from different documents please mention all possibilities in your answer and use the titles to separate between topics or domains. If you cannot answer the question from the given documents, please state that you do not have an answer.\\n\"\"\"\n",
    "prompt = pre_prompt + \"CONTEXT:\\n\\n{context}\\n\" +\"Question : {question}\" + \"[\\INST]\"\n",
    "llama_prompt = PromptTemplate(template=prompt, input_variables=[\"context\", \"question\"])\n",
    "# integrate prompt with LLM\n",
    "chain = ConversationalRetrievalChain.from_llm(llm, loaded_vectorstore.as_retriever(), combine_docs_chain_kwargs={\"prompt\": llama_prompt}, return_source_documents=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d03dcde7868d4509bfe1aaf9edea50a8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:389: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:394: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#from jsonformer import Jsonformer\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(\"AdaptLLM/medicine-chat\", device_map=\"auto\", load_in_4bit=True)\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"AdaptLLM/medicine-chat\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_input = \"\"\"Identify sign symptoms mentioned in text between !!!. For every detected sign symptom extract only one word and do not provide additional information.\n",
    "\n",
    "!!!\n",
    "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
    "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
    "He had no medical history and did not take any medicine.\n",
    "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
    "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
    "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
    "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
    "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
    "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
    "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
    "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
    "!!!\n",
    "\"\"\"\n",
    "our_system_prompt = \"\\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe. Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\\n\\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\\n\" # Please do NOT change this\n",
    "\n",
    "prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{user_input} [/INST]\"\n",
    "\n",
    "# # NOTE:\n",
    "# # If you want to apply your own system prompt, please integrate it into the instruction part following our system prompt like this:\n",
    "# your_system_prompt = \"Please, answer this question faithfully.\"\n",
    "# prompt = f\"<s>[INST] <<SYS>>{our_system_prompt}<</SYS>>\\n\\n{your_system_prompt}\\n{user_input} [/INST]\"\n",
    "\n",
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096, max_new_tokens=150)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{user_input}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = \"\"\"<s> [INST] <<SYS>>\n",
    "Your goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\n",
    "\n",
    "\n",
    "```TypeScript\n",
    "\n",
    "patient: { // A patient is an individual admitted to or seeking medical attention at a hospital or healthcare facility. \n",
    " age: Array<string> // Patients' age.\n",
    " sex: Array<string> // Patients' sex.\n",
    " sign symptom: Array<string> // Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\n",
    " biological structure: Array<string> // Biological structures refer to the anatomical components and organs mentioned in the input.\n",
    " diagnostic procedure: Array<string> // A diagnostic procedure is a medical test or examination.\n",
    " lab value: Array<string> // Laboratory values refer to the results of tests.\n",
    "}\n",
    "```\n",
    "\n",
    "\n",
    "Please output the extracted information in JSON format. Do not output anything except for the extracted information. Do not add any clarifying information. Do not add any fields that are not in the schema. If the text contains attributes that do not appear in the schema, please ignore them. All output must be in JSON format and follow the schema specified above. Wrap the JSON in <json> tags.\n",
    "Information is specificly extracted by the Input befor your output. Do not use blood as biological structure it is not mentioned in the text i gave you.\n",
    "\n",
    "\n",
    "\n",
    "Input: A 32-year-old man\n",
    "Output: <json>{\"patient\": {\"age\": [\"32-year-old\"]}}</json>\n",
    "Input: A 32-year-old man.\n",
    "Output: <json>{\"patient\": {\"sex\": [\"Male\"]}}</json>\n",
    "Input: A 50-year-old woman.\n",
    "Output: <json>{\"patient\": {\"sex\": [\"Female\"]}}</json>\n",
    "Input: The patient has a fever temperature of 38.\n",
    "Output: <json>{\"patient\": {\"sign symptom\": [\"fever\"]}}</json>\n",
    "Input: The patient complains of abdominal pain.\n",
    "Output: <json>{\"patient\": {\"sign symptom\": [\"abdominal pain\"]}}</json>\n",
    "Input: <</SYS>> A 58-year-old man had been suffering from general fatigue and severe anemia for several months. His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl). He had no medical history and did not take any medicine. Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding. Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1). Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2). We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions. Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes. Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3). Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4). The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery. [/INST]\n",
    "Output: \n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = \"\"\"[INST]\n",
    "Your goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\n",
    "\n",
    "\n",
    "```TypeScript\n",
    "\n",
    "patient: { // A patient is an individual admitted to or seeking medical attention at a hospital or healthcare facility. \n",
    " age: Array<string> // Patients' age.\n",
    " sex: Array<string> // Patients' sex.\n",
    " sign symptom: Array<string> // Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\n",
    " biological structure: Array<string> // Biological structures refer to the anatomical components and organs mentioned in the input.\n",
    " diagnostic procedure: Array<string> // A diagnostic procedure is a medical test or examination.\n",
    " lab value: Array<string> // Laboratory values refer to the results of tests.\n",
    "}\n",
    "```\n",
    "\n",
    "\n",
    "Please output the extracted information in JSON format. Do not output anything except for the extracted information. Do not add any clarifying information. Do not add any fields that are not in the schema. If the text contains attributes that do not appear in the schema, please ignore them. All output must be in JSON format and follow the schema specified above. Wrap the JSON in <json> tags.\n",
    "Information is specificly extracted by the Input befor your output. Do not use blood as biological structure it is not mentioned in the text i gave you.\n",
    "\n",
    "\n",
    "\n",
    "Input: His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl). He had no medical history and did not take any medicine. Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding. Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1). Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2). We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions. Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes. Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3). Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4). The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery. \n",
    "Output: [/INST]\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Both `max_new_tokens` (=200) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### User Input:\n",
      "[INST]\n",
      "Your goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\n",
      "\n",
      "\n",
      "```TypeScript\n",
      "\n",
      "patient: { // A patient is an individual admitted to or seeking medical attention at a hospital or healthcare facility. \n",
      " age: Array<string> // Patients' age.\n",
      " sex: Array<string> // Patients' sex.\n",
      " sign symptom: Array<string> // Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\n",
      " biological structure: Array<string> // Biological structures refer to the anatomical components and organs mentioned in the input.\n",
      " diagnostic procedure: Array<string> // A diagnostic procedure is a medical test or examination.\n",
      " lab value: Array<string> // Laboratory values refer to the results of tests.\n",
      "}\n",
      "```\n",
      "\n",
      "\n",
      "Please output the extracted information in JSON format. Do not output anything except for the extracted information. Do not add any clarifying information. Do not add any fields that are not in the schema. If the text contains attributes that do not appear in the schema, please ignore them. All output must be in JSON format and follow the schema specified above. Wrap the JSON in <json> tags.\n",
      "Information is specificly extracted by the Input befor your output. Do not use blood as biological structure it is not mentioned in the text i gave you.\n",
      "\n",
      "\n",
      "\n",
      "Input: His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl). He had no medical history and did not take any medicine. Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding. Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1). Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2). We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions. Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes. Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3). Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4). The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery. \n",
      "Output: We have no infprmation about patients age. [/INST]\n",
      "\n",
      "\n",
      "### Assistant Output:\n",
      "\n",
      "```\n",
      "{\n",
      "  \"patient\": {\n",
      "    \"age\": [\"6.6 g/dl (normal range: 12-16 g/dl)\"],\n",
      "    \"sex\": [\"His\"],\n",
      "    \"sign symptom\": [\"Hemoglobin levels were 6.6 g/dl (normal range: 12-16 g/dl). He had no medical history and did not take any medicine.\"],\n",
      "    \"biological structure\": [\"Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding. Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1). Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the\n"
     ]
    }
   ],
   "source": [
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096, max_new_tokens=200)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "print(f'### User Input:\\n{prompt}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TESTING KOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e1fab2a7964a40118f3a16a9bbaa953a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/8 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dc1ef0c6f4f54117a97b3bce79a2d1e4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer_config.json:   0%|          | 0.00/4.08k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ff4e19dfde5244f08dbf130dbe1b75c6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "585a916af3c346c5b641a23050451692",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer.json:   0%|          | 0.00/1.85M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e33ea646a7e7420c9d6d91239f23228d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "added_tokens.json:   0%|          | 0.00/344 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0508fcaf54e44b0cb349e2385d80a076",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "special_tokens_map.json:   0%|          | 0.00/736 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\n"
     ]
    }
   ],
   "source": [
    "from langchain.llms import HuggingFacePipeline\n",
    "import torch\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline, AutoModelForSeq2SeqLM\n",
    "from kor import create_extraction_chain, Object, Text\n",
    "\n",
    "#model_id = 'AdaptLLM/medicine-chat'# go for a smaller model if you dont have the VRAM\n",
    "model_id = 'epfl-llm/meditron-7b'\n",
    "access_token = \"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\"\n",
    "model = AutoModelForCausalLM.from_pretrained(model_id, device_map=\"auto\", load_in_4bit=True, token=access_token)\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_id, token=access_token)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Original example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pipeline' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m pipe \u001b[38;5;241m=\u001b[39m \u001b[43mpipeline\u001b[49m(\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtext-generation\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      3\u001b[0m     model\u001b[38;5;241m=\u001b[39mmodel, \n\u001b[1;32m      4\u001b[0m     tokenizer\u001b[38;5;241m=\u001b[39mtokenizer, \n\u001b[1;32m      5\u001b[0m     max_length\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1000\u001b[39m,\n\u001b[1;32m      6\u001b[0m     temperature\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.1\u001b[39m \n\u001b[1;32m      7\u001b[0m )\n\u001b[1;32m      9\u001b[0m local_llm \u001b[38;5;241m=\u001b[39m HuggingFacePipeline(pipeline\u001b[38;5;241m=\u001b[39mpipe)\n\u001b[1;32m     12\u001b[0m schema \u001b[38;5;241m=\u001b[39m Object(\n\u001b[1;32m     13\u001b[0m     \u001b[38;5;28mid\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplayer\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     14\u001b[0m     description\u001b[38;5;241m=\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     48\u001b[0m     many\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m     49\u001b[0m )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pipeline' is not defined"
     ]
    }
   ],
   "source": [
    "pipe = pipeline(\n",
    "    \"text-generation\",\n",
    "    model=model, \n",
    "    tokenizer=tokenizer, \n",
    "    max_length=4096,\n",
    "    temperature=0.1 \n",
    "    max_new_tokens=150\n",
    ")\n",
    "\n",
    "local_llm = HuggingFacePipeline(pipeline=pipe)\n",
    "\n",
    "\n",
    "schema = Object(\n",
    "    id=\"player\",\n",
    "    description=(\n",
    "        \"User is controlling a music player to select songs, pause or start them or play\"\n",
    "        \" music by a particular artist.\"\n",
    "    ),\n",
    "    attributes=[\n",
    "        Text(\n",
    "            id=\"song\",\n",
    "            description=\"User wants to play this song\",\n",
    "            examples=[],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"album\",\n",
    "            description=\"User wants to play this album\",\n",
    "            examples=[],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"artist\",\n",
    "            description=\"Music by the given artist\",\n",
    "            examples=[(\"Songs by paul simon\", \"paul simon\")],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"action\",\n",
    "            description=\"Action to take one of: `play`, `stop`, `next`, `previous`.\",\n",
    "            examples=[\n",
    "                (\"Please stop the music\", \"stop\"),\n",
    "                (\"play something\", \"play\"),\n",
    "                (\"play a song\", \"play\"),\n",
    "                (\"next song\", \"next\"),\n",
    "            ],\n",
    "        ),\n",
    "    ],\n",
    "    many=False,\n",
    ")\n",
    "\n",
    "chain = create_extraction_chain(local_llm, schema, encoder_or_encoder_class='json')\n",
    "chain.run(\"play songs by paul simon and led zeppelin and the doors\")['data']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mySet = {\n",
    "    'Diagnostic_procedure',\n",
    "    'Sign_symptom',\n",
    "    'Biological_structure',\n",
    "    'Detailed_description',\n",
    "    'Age',\n",
    "    'Lab_value'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'patient': {'age': ['58-year-old'],\n",
       "  'sex': ['Male'],\n",
       "  'sign symptom': ['general fatigue'],\n",
       "  'biological structure': ['small intestine'],\n",
       "  'diagnostic procedure': ['esophagogastroduodenoscopy'],\n",
       "  'lab value': ['6.6 g/dl']}}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain.prompts import PromptTemplate\n",
    "\n",
    "pipe = pipeline(\n",
    "    \"text-generation\",\n",
    "    model=model, \n",
    "    tokenizer=tokenizer, \n",
    "    max_length=2048,\n",
    "    #do_sample=True,\n",
    "    #temperature=0.0001,\n",
    "    #max_new_tokens=300\n",
    ")\n",
    "\n",
    "local_llm = HuggingFacePipeline(pipeline=pipe)\n",
    "\n",
    "#[INST] <<SYS>>\\n<</SYS>> [/INST]\n",
    "\n",
    "instruction_template = PromptTemplate(\n",
    "    input_variables=[\"format_instructions\", \"type_description\"],\n",
    "    template=(\n",
    "        \"[INST]\\n Your goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\\n\\n\"\n",
    "        \"{type_description}\\n\\n\" # Can comment out\n",
    "        \"{format_instructions}\\n\\n\"\n",
    "        #\"Here are some examples:\\n\\n\"\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "schema = Object(\n",
    "    id=\"patient\",\n",
    "    description=(\n",
    "        \"A patient is an individual admitted to or seeking medical attention\"\n",
    "        \" at a hospital or healthcare facility.\"\n",
    "    ),\n",
    "    attributes=[\n",
    "        Text(\n",
    "            id=\"age\",\n",
    "            description=\"Patients' age.\",\n",
    "            examples=[\n",
    "                (\"A 32-year-old man\", \"32-year-old\")\n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"sex\",\n",
    "            description=\"Patients' sex.\",\n",
    "            examples=[\n",
    "                (\"A 32-year-old man.\", \"Male\"),\n",
    "                (\"A 50-year-old woman.\", \"Female\")\n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"sign symptom\",\n",
    "            description=\"Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\",\n",
    "            examples=[\n",
    "                  (\"The patient has a fever temperature of 38.\", \"fever\"),\n",
    "                  (\"The patient complains of abdominal pain.\", \"abdominal pain\")\n",
    "            ],\n",
    "              many=True,\n",
    "          ),\n",
    "         Text(\n",
    "             id=\"biological structure\",\n",
    "             description=\"Biological structures refer to the anatomical components and organs.\",\n",
    "             examples=[\n",
    "                 (\"The patient lost strength in the left arm.\", \"arm\"),\n",
    "                 (\"Computed tomography (CT) of the chest.\", \"chest\"),\n",
    "                 (\"Discomfort in the left upper quadrant of the abdomen.\", \"abdomen\"),\n",
    "             ],\n",
    "             many=True,\n",
    "         ),\n",
    "        Text(\n",
    "            id=\"diagnostic procedure\",\n",
    "            description=\"A diagnostic procedure is a medical test or examination.\",\n",
    "            examples=[\n",
    "                (\"An electrocardiogram revealed normal sinus rhythm\", \"electrocardiogram\"),\n",
    "                (\"Physical examination yielded unremarkable findings.\", \"physical examination\"),\n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"lab value\",\n",
    "            description=\"Laboratory values refer to the results of tests.\",\n",
    "            examples=[\n",
    "                (\"Surgical margins were tumor-free.\", \"tumor-free\"),\n",
    "                (\"His temperature was 38.4 °C.\", \"38.4 °C\"),\n",
    "                (\"Heart rate 130/minute\", \"130/minute\"),\n",
    "                (\"Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\", \"0.096 mmol/l\"), \n",
    "                (\"Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\", \"1.6\"), \n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "    ],\n",
    "    many=False,\n",
    ")\n",
    "\n",
    "user_input = \"\"\"\\n\n",
    "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
    "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
    "He had no medical history and did not take any medicine.\n",
    "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
    "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1). [/INST]\"\"\"\n",
    "\n",
    "prompt = f\"{user_input}\"\n",
    "\n",
    "#prompt = user_input\n",
    "\n",
    "chain = create_extraction_chain(local_llm, schema, instruction_template=instruction_template, encoder_or_encoder_class='json')\n",
    "chain.run(prompt)['data']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INST]\n",
      " Your goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\n",
      "\n",
      "```TypeScript\n",
      "\n",
      "patient: { // A patient is an individual admitted to or seeking medical attention at a hospital or healthcare facility.\n",
      " age: Array<string> // Patients' age.\n",
      " sex: Array<string> // Patients' sex.\n",
      " sign symptom: Array<string> // Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\n",
      " biological structure: Array<string> // Biological structures refer to the anatomical components and organs.\n",
      " diagnostic procedure: Array<string> // A diagnostic procedure is a medical test or examination.\n",
      " lab value: Array<string> // Laboratory values refer to the results of tests.\n",
      "}\n",
      "```\n",
      "\n",
      "\n",
      "Please output the extracted information in JSON format. Do not output anything except for the extracted information. Do not add any clarifying information. Do not add any fields that are not in the schema. If the text contains attributes that do not appear in the schema, please ignore them. All output must be in JSON format and follow the schema specified above. Wrap the JSON in <json> tags.\n",
      "\n",
      "\n",
      "\n",
      "Input: A 32-year-old man\n",
      "Output: <json>{\"patient\": {\"age\": [\"32-year-old\"]}}</json>\n",
      "Input: A 32-year-old man.\n",
      "Output: <json>{\"patient\": {\"sex\": [\"Male\"]}}</json>\n",
      "Input: A 50-year-old woman.\n",
      "Output: <json>{\"patient\": {\"sex\": [\"Female\"]}}</json>\n",
      "Input: The patient has a fever temperature of 38.\n",
      "Output: <json>{\"patient\": {\"sign symptom\": [\"fever\"]}}</json>\n",
      "Input: The patient complains of abdominal pain.\n",
      "Output: <json>{\"patient\": {\"sign symptom\": [\"abdominal pain\"]}}</json>\n",
      "Input: The patient lost strength in the left arm.\n",
      "Output: <json>{\"patient\": {\"biological structure\": [\"arm\"]}}</json>\n",
      "Input: Computed tomography (CT) of the chest.\n",
      "Output: <json>{\"patient\": {\"biological structure\": [\"chest\"]}}</json>\n",
      "Input: Discomfort in the left upper quadrant of the abdomen.\n",
      "Output: <json>{\"patient\": {\"biological structure\": [\"abdomen\"]}}</json>\n",
      "Input: An electrocardiogram revealed normal sinus rhythm\n",
      "Output: <json>{\"patient\": {\"diagnostic procedure\": [\"electrocardiogram\"]}}</json>\n",
      "Input: Physical examination yielded unremarkable findings.\n",
      "Output: <json>{\"patient\": {\"diagnostic procedure\": [\"physical examination\"]}}</json>\n",
      "Input: Surgical margins were tumor-free.\n",
      "Output: <json>{\"patient\": {\"lab value\": [\"tumor-free\"]}}</json>\n",
      "Input: His temperature was 38.4 °C.\n",
      "Output: <json>{\"patient\": {\"lab value\": [\"38.4 °C\"]}}</json>\n",
      "Input: Heart rate 130/minute\n",
      "Output: <json>{\"patient\": {\"lab value\": [\"130/minute\"]}}</json>\n",
      "Input: Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\n",
      "Output: <json>{\"patient\": {\"lab value\": [\"0.096 mmol/l\"]}}</json>\n",
      "Input: Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\n",
      "Output: <json>{\"patient\": {\"lab value\": [\"1.6\"]}}</json>\n",
      "Input: [user_input]\n",
      "Output:\n"
     ]
    }
   ],
   "source": [
    "print(chain.prompt.format_prompt(\"[user_input]\").to_string())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TESTING WITH FOR LOOP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
      "{'patient': {'age': ['58'], 'sex': ['male'], 'sign symptom': ['general fatigue', 'severe anemia'], 'biological structure': ['blood'], 'diagnostic procedure': ['blood test'], 'lab value': ['anemia', '38.4 °C']}}\n",
      "---------------\n",
      "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
      "{'patient': {'lab value': ['6.6 g/dl']}}\n",
      "---------------\n",
      "He had no medical history and did not take any medicine.\n",
      "{'patient': {'age': ['0'], 'sex': ['M'], 'sign symptom': ['none'], 'biological structure': ['none'], 'diagnostic procedure': ['none'], 'lab value': ['38.4 °C']}}\n",
      "---------------\n",
      "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
      "{'patient': {'diagnostic procedure': ['Esophagogastroduodenoscopy', 'colonoscopy'], 'lab value': ['No significant bleeding found']}}\n",
      "---------------\n",
      "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
      "{'patient': {'diagnostic procedure': ['Abdominal computer tomography'], 'lab value': ['2-cm hypervascular tumor in the small intestine (Fig.1)']}}\n",
      "---------------\n",
      "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
      "{}\n",
      "---------------\n",
      "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
      "{'patient': {'diagnostic procedure': ['biopsy'], 'lab value': ['difficult to stop bleeding in the case of hypervascular lesions']}}\n",
      "---------------\n",
      "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
      "{'patient': {'lab value': ['38.4 °C', 'laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes']}}\n",
      "---------------\n",
      "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
      "{'patient': {'diagnostic procedure': ['Examination of the resected tumor showed that it measured 19×16 mm in diameter (Fig.3)'], 'lab value': ['38.4 °C']}}\n",
      "---------------\n",
      "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
      "{'patient': {'diagnostic procedure': ['Histology'], 'lab value': ['PG', '38.4 °C']}}\n",
      "---------------\n",
      "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
      "{'patient': {'age': ['38.4 °C'], 'sex': ['male'], 'sign symptom': ['no'], 'biological structure': ['no'], 'diagnostic procedure': ['no'], 'lab value': ['38.4 °C']}}\n",
      "---------------\n",
      "\n",
      "{'patient': {'age': ['38'], 'sex': ['male'], 'sign symptom': ['fever'], 'biological structure': ['body temperature'], 'diagnostic procedure': ['temperature reading'], 'lab value': ['38.4 °C']}}\n",
      "---------------\n"
     ]
    }
   ],
   "source": [
    "from langchain.prompts import PromptTemplate\n",
    "\n",
    "for txtAnnPair in file_tuples[:1]:\n",
    "    pathToTxtFile = os.path.join(path, txtAnnPair[0])\n",
    "    pathToAnnFile = os.path.join(path, txtAnnPair[1])\n",
    "\n",
    "    Txtfile = open(pathToTxtFile, \"r\")\n",
    "    Txtfile = Txtfile.readlines()\n",
    "    Txtfile = \"\".join(Txtfile)\n",
    "    \n",
    "    # removed = []\n",
    "    # previous = -10\n",
    "    # for annLine in allAnnLines.copy():\n",
    "    #     currentStart = int((annLine[1].split()[1]))\n",
    "    #     if currentStart == previous:\n",
    "    #         removed.append(annLine)\n",
    "    #         allAnnLines.remove(annLine)\n",
    "    #     previous = currentStart\n",
    "\n",
    "    for sentence in Txtfile.split(\"\\n\"):\n",
    "        print(sentence)\n",
    "        pipe = pipeline(\n",
    "            \"text-generation\",\n",
    "            model=model, \n",
    "            tokenizer=tokenizer, \n",
    "            max_length=4096,\n",
    "        )\n",
    "\n",
    "        local_llm = HuggingFacePipeline(pipeline=pipe)\n",
    "\n",
    "        instruction_template = PromptTemplate(\n",
    "            input_variables=[\"format_instructions\", \"type_description\"],\n",
    "            template=(\n",
    "                \"[INST] <<SYS>>\\nYour goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\\n<</SYS>>\\n\\n\"\n",
    "                \"{type_description}\\n\\n\" # Can comment out\n",
    "                \"{format_instructions}\\n\\n\"\n",
    "                #\"Here are some examples:\\n\\n\"\n",
    "            ),\n",
    "        )\n",
    "\n",
    "\n",
    "\n",
    "        schema = Object(\n",
    "            id=\"patient\",\n",
    "            description=(\n",
    "                \"A patient is an individual admitted to or seeking medical attention\"\n",
    "                \" at a hospital or healthcare facility.\"\n",
    "            ),\n",
    "            attributes=[\n",
    "                Text(\n",
    "                    id=\"age\",\n",
    "                    description=\"Patients' age.\",\n",
    "                    examples=[\n",
    "                        #(\"A 32-year-old\", \"32-year-old\"),\n",
    "                    ],\n",
    "                    many=True,\n",
    "                ),\n",
    "                Text(\n",
    "                    id=\"sex\",\n",
    "                    description=\"Patients' sex.\",\n",
    "                    examples=[\n",
    "                        #(\"A 32-year-old man.\", \"Male\"),\n",
    "                        #(\"A 50-year-old woman.\", \"Female\")\n",
    "                    ],\n",
    "                    many=True,\n",
    "                ),\n",
    "                Text(\n",
    "                    id=\"sign symptom\",\n",
    "                    description=\"Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\",\n",
    "                    examples=[\n",
    "                        #(\"The patient has a fever temperature of 38.\", \"fever\"),\n",
    "                        #(\"The patient complains of abdominal pain.\", \"abdominal pain\")\n",
    "                    ],\n",
    "                    many=True,\n",
    "                ),\n",
    "                Text(\n",
    "                    id=\"biological structure\",\n",
    "                    description=\"Biological structures refer to the anatomical body components.\",\n",
    "                    examples=[\n",
    "                        #(\"The patient lost strength in the left arm.\", \"arm\"),\n",
    "                        #(\"Computed tomography (CT) of the chest.\", \"chest\"),\n",
    "                        #(\"Discomfort in the left upper quadrant of the abdomen.\", \"abdomen\"),\n",
    "                    ],\n",
    "                    many=True,\n",
    "                ),\n",
    "                Text(\n",
    "                    id=\"diagnostic procedure\",\n",
    "                    description=\"A diagnostic procedure is a medical test or examination.\",\n",
    "                    examples=[\n",
    "                        # (\"An electrocardiogram revealed normal sinus rhythm\", \"electrocardiogram\"),\n",
    "                        # (\"Physical examination yielded unremarkable findings.\", \"physical examination\"),\n",
    "                    ],\n",
    "                    many=True,\n",
    "                ),\n",
    "                Text(\n",
    "                    id=\"lab value\",\n",
    "                    description=\"Laboratory values refer to the results of tests or diagnostic procedures.\",\n",
    "                    examples=[\n",
    "                        # (\"Surgical margins were tumor-free.\", \"tumor-free\"),\n",
    "                         (\"His temperature was 38.4 °C.\", \"38.4 °C\"),\n",
    "                        # (\"Heart rate 130/minute\", \"130/minute\"),\n",
    "                        # (\"His hemoglobin levels were 6.6 g/dl\", \"6.6 g/dl\")\n",
    "                        #(\"Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\", \"0.096 mmol/l\"), \n",
    "                        #(\"Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\", \"1.6\"), \n",
    "                    ],\n",
    "                    many=True,\n",
    "                ),\n",
    "            ],\n",
    "            many=False,\n",
    "        )\n",
    "\n",
    "        #user_input = \"\"\"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.[/INST]\"\"\"\n",
    "\n",
    "        prompt = f\"{sentence}[/INST]\"\n",
    "\n",
    "        #prompt = user_input\n",
    "\n",
    "        chain = create_extraction_chain(local_llm, schema, instruction_template=instruction_template, encoder_or_encoder_class='json')\n",
    "        print(chain.run(prompt)['data'])\n",
    "        print(\"---------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
      "----\n",
      "His hemoglobin levels were 6.6 g/dl (normal range: 12–16 g/dl).\n",
      "----\n",
      "He had no medical history and did not take any medicine.\n",
      "----\n",
      "Esophagogastroduodenoscopy and colonoscopy did not reveal any significant bleeding.\n",
      "----\n",
      "Abdominal computer tomography revealed a 2-cm hypervascular tumor in the small intestine (Fig.1).\n",
      "----\n",
      "Oral DBE detected a 2-cm-diameter reddish, submucosal tumor-like lesion with surface ulceration in the jejunum, approximately 20 cm away from the Treitz ligament (Fig.2).\n",
      "----\n",
      "We did not perform biopsy because it can be difficult to stop bleeding in the case of hypervascular lesions.\n",
      "----\n",
      "Under the diagnosis of a small bowel tumor, gastrointestinal stromal tumor (GIST), malignant lymphoma, or cancer, we performed laparoscopic-assisted segmental resection of the jejunum with the dissection of lymph nodes.\n",
      "----\n",
      "Examination of the resected tumor showed that it measured 19 × 16 mm in diameter (Fig.3).\n",
      "----\n",
      "Histology revealed the proliferation of blood capillaries and granulation tissue, which was consistent with PG (Fig.4).\n",
      "----\n",
      "The patient was discharged on postoperative day 9 without complication and his anemia improved gradually without the need for oral iron after surgery.\n",
      "----\n",
      "\n",
      "----\n",
      "A 67-year-old woman was sent to emergency department because of fever and sore throat.\n",
      "----\n",
      "Her medical history included ESRD with regular hemodialysis 3 times per week for 15 years, hyperuricemia and gout.\n",
      "----\n",
      "She also had comorbidity such as dyslipidemia, hypertension, secondary hyperparathyroidism, and diabetes.\n",
      "----\n",
      "The hemogram revealed a white blood cell count (WBC) of 700/μL, with 2% neutrophils, 94% lymphocytes and 2% monocytes, hemoglobin 11.1 g/dL, and platelet count, 131,000/μL.\n",
      "----\n",
      "Under the impression of febrile neutropenia and acute pharyngitis, she was admitted to our hematologic ward for further survey and management.\n",
      "----\n",
      "Broad-spectrum antibiotics with piperacillin 2 g and tazobactam 0.25 gm i.v.\n",
      "----\n",
      "q8h had been administered and her infection sign resolved gradually.\n",
      "----\n",
      "Upon admission, we reviewed her oral medication: glipizide 5 mg tid, saxagliptin 2.5 mg qd, fenofibrate 600 mg qd, aluminum hydroxide 324 mg tid, folic acid 5 mg qd, calcium carbonate 1000 mg tid, and febuxostat 40 mg qd.\n",
      "----\n",
      "Besides, she also received epoetin-beta 2000 iu i.v.tiw.\n",
      "----\n",
      "Except for febuxostat, all the other drugs had been used for more than 1 year.\n",
      "----\n",
      "Febuxostat was administered 2½ months before admission for inadequate serum uric acid control by allopurinol 50 mg qd.\n",
      "----\n",
      "Two weeks before febuxostat exposure, routine laboratory test revealed WBC 6000/μL and serum uric acid level 9.8 mg/dL.\n",
      "----\n",
      "Febuxostat was discontinued thereafter due to the causal relationship of agranulocytosis cannot be excluded.\n",
      "----\n",
      "Besides, we also surveyed viral infection and autoimmune disorder.\n",
      "----\n",
      "There were no clinical or laboratory evidence of Epstein–Barr virus, cytomegalovirus, or human immunodeficiency virus infections; antinuclear antibody (ANA) and antiextractable nuclear antigen (anti-ENA) were both negative.\n",
      "----\n",
      "Bone marrow examination during hospitalization showed hypocellular marrow with a marked decrease in myeloid component but no evidence of hematologic neoplasms.\n",
      "----\n",
      "Chromosome analysis of bone marrow was normal karyotype.\n",
      "----\n",
      "The patient denied history of radiation or chemicals exposure.\n",
      "----\n",
      "After stopping febuxostat for 17 days, her neutropenia improved significantly (WBC 2100/μL, and neutrophil 66%), without any granulocyte colony-stimulating factor (G-CSF) support.\n",
      "----\n",
      "After discharge, her WBC and differential count was completely normal during follow-up (Fig.1).\n",
      "----\n",
      "This study was approved by our institutional review board.\n",
      "----\n",
      "\n",
      "----\n",
      "The 31-year-old white primipara with twin pregnancy was admitted to hospital in the 38th week of gestation with elevated blood pressure (150/100 mmHg).\n",
      "----\n",
      "After receiving antihypertensive treatment blood pressure was 120/80.\n",
      "----\n",
      "All laboratory variables, including plasma proteins, were within their respective reference intervals.\n",
      "----\n",
      "Few hours after admission to the hospital the contractions started, and the caesarean section was performed because of vertex-transverse presentation of twins.\n",
      "----\n",
      "Liveborn female and male were delivered.\n",
      "----\n",
      "The postoperative course was initially inconspicuous.\n",
      "----\n",
      "Four hours postpartum she experienced sudden epigastric pain.\n",
      "----\n",
      "The blood pressure rose to 190/130.\n",
      "----\n",
      "Laboratory findings showed haemolysis, thrombocytopenia, and an increase in serum creatinine and aminotransferases (Table I).\n",
      "----\n",
      "Intravenous magnesium sulphate was administered.\n",
      "----\n",
      "Abdominal ultrasound disclosed an empty uterine cavity without placenta residue.\n",
      "----\n",
      "Five hours postpartum, the patient was transferred to the intensive care unit because of poor urine output, drowsiness, and suspicion of DIC.\n",
      "----\n",
      "On examination she was sleepy and disoriented.\n",
      "----\n",
      "The combination of haemolysis, thrombocytopenia, and elevated liver enzymes suggested a postpartum HELLP syndrome, complicated by DIC.\n",
      "----\n",
      "The patient was rehydrated, and treatment was instituted with fresh frozen plasma, red cell transfusion, fresh platelets, and kybernin P (antithrombin III).\n",
      "----\n",
      "High serum urea, creatinine, and persistent anuria were compatible with acute renal failure.\n",
      "----\n",
      "A high dose of furosemide failed to increase diuresis.\n",
      "----\n",
      "Sixteen hours postpartum her blood pressure was 60/40 and oxygen saturation 70%.\n",
      "----\n",
      "Few minutes later the patient had acute cardiac arrest, and resuscitation started.\n",
      "----\n",
      "Resuscitation was successful, and normal heart action was re-established; blood pressure rose to 150/110, and oxygen saturation was 90%.\n",
      "----\n",
      "Laboratory findings were deteriorated, and development of acute renal failure indicated the necessity of urgent dialysis.\n",
      "----\n",
      "Following dialysis, the patient was stable, oxygen saturation 98%.\n",
      "----\n",
      "Twenty hours postpartum patient developed again cardiopulmonary arrest.\n",
      "----\n",
      "Despite resuscitation attempts the outcome was lethal.\n",
      "----\n",
      "Due to the fact that death occurred in the hospital a short time after a surgical intervention, a medico-legal autopsy was ordered.\n",
      "----\n",
      "An autopsy revealed oedema of the brain and lungs as well as dilatation of right and left ventricles.\n",
      "----\n",
      "Hydrothorax (900 mL), hydroperitoneum (2500 mL), and hydropericardium (200 mL) were present.\n",
      "----\n",
      "The field of operation of the caesarean section was unremarkable.\n",
      "----\n",
      "Petechial and suffusional haemorrhages were observed under the pleura, endocardium, in the mucosae of the renal pelvis, and peritoneum of the small and large bowel.\n",
      "----\n",
      "Histology revealed periportal hepatocellular necrosis, bloodless glomeruli with swollen and vacuolated intracapillary cells, as well as confluent haemorrhages in kidneys, liver, and spleen.\n",
      "----\n",
      "Death was attributed to multiple organ failure due to DIC as a consequence of HELLP syndrome.\n",
      "----\n",
      "\n",
      "----\n"
     ]
    }
   ],
   "source": [
    "for txtAnnPair in file_tuples[:3]:\n",
    "    pathToTxtFile = os.path.join(path, txtAnnPair[0])\n",
    "    pathToAnnFile = os.path.join(path, txtAnnPair[1])\n",
    "\n",
    "    Txtfile = open(pathToTxtFile, \"r\")\n",
    "    Txtfile = Txtfile.readlines()\n",
    "    Txtfile = \"\".join(Txtfile)\n",
    "    \n",
    "    # removed = []\n",
    "    # previous = -10\n",
    "    # for annLine in allAnnLines.copy():\n",
    "    #     currentStart = int((annLine[1].split()[1]))\n",
    "    #     if currentStart == previous:\n",
    "    #         removed.append(annLine)\n",
    "    #         allAnnLines.remove(annLine)\n",
    "    #     previous = currentStart\n",
    "\n",
    "    for sentence in Txtfile.split(\"\\n\"):\n",
    "        print(sentence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# / TESTING WITH FOR LOOP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FINE-TUNING with PEFT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from peft import LoraConfig, TaskType\n",
    "\n",
    "peft_config = LoraConfig(task_type=TaskType.CAUSAL_LM, inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2982303e8f7f469c94e0715282370d84",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:389: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:394: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoModelForCausalLM, AutoTokenizer, TrainingArguments, Trainer\n",
    "\n",
    "model_id = 'AdaptLLM/medicine-chat'# go for a smaller model if you dont have the VRAM\n",
    "model = AutoModelForCausalLM.from_pretrained(model_id, device_map=\"auto\", load_in_4bit=True)\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trainable params: 4,194,304 || all params: 6,742,618,112 || trainable%: 0.06220586618327525\n"
     ]
    }
   ],
   "source": [
    "from peft import get_peft_model\n",
    "\n",
    "model = get_peft_model(model, peft_config)\n",
    "model.print_trainable_parameters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_prompt(sample):\n",
    "    return f\"\"\"\n",
    "Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n",
    "\n",
    "### Instruction:\n",
    "{sample[\"instruction\"]}\n",
    "\n",
    "### Input:\n",
    "{sample[\"input\"]}\n",
    "\n",
    "### Response:\n",
    "{sample[\"output\"]}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_args = TrainingArguments(\n",
    "    output_dir=\"../models/AdaptLLM/medicine-chat-Erik\",\n",
    "    learning_rate=1e-3,\n",
    "    per_device_train_batch_size=1,\n",
    "    #per_device_eval_batch_size=32,\n",
    "    num_train_epochs=30,\n",
    "    #weight_decay=0.01,\n",
    "    #evaluation_strategy=\"epoch\",\n",
    "    #save_strategy=\"epoch\",\n",
    "    #load_best_model_at_end=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = {\n",
    "    \"instruction\": \"Your goal is to extract structured information from the user's input.\",\n",
    "    \"input\": \"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\",\n",
    "    \"output\": \"\"\"{'age': ['58'], \n",
    "            'sex': ['male'], \n",
    "            'sign symptom': ['general fatigue', 'severe anemia']\n",
    "    }\n",
    "    \"\"\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.DataFrame(columns=[\"instruction\", \"input\", \"output\"])\n",
    "df.loc[0] = [\"Your goal is to extract structured information from the user's input.\", \"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\", \"\"\"{'age': ['58'], \n",
    "            'sex': ['male'], \n",
    "            'sign symptom': ['general fatigue', 'severe anemia']\n",
    "    }\n",
    "    \"\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>instruction</th>\n",
       "      <th>input</th>\n",
       "      <th>output</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Your goal is to extract structured information...</td>\n",
       "      <td>A 58-year-old man had been suffering from gene...</td>\n",
       "      <td>{'age': ['58'], \\n            'sex': ['male'],...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         instruction  \\\n",
       "0  Your goal is to extract structured information...   \n",
       "\n",
       "                                               input  \\\n",
       "0  A 58-year-old man had been suffering from gene...   \n",
       "\n",
       "                                              output  \n",
       "0  {'age': ['58'], \\n            'sex': ['male'],...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    Your goal is to extract structured information...\n",
       "Name: instruction, dtype: object"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"instruction\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'column_names'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_2827299/2814384113.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m trainer = SFTTrainer(\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mmodel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtraining_args\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mtrain_dataset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m                  \u001b[0;31m#tokenized_datasets[\"train\"],\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, model, args, data_collator, train_dataset, eval_dataset, tokenizer, model_init, compute_metrics, callbacks, optimizers, preprocess_logits_for_metrics, peft_config, dataset_text_field, packing, formatting_func, max_seq_length, infinite, num_of_sequences, chars_per_token, dataset_num_proc, dataset_batch_size, neftune_noise_alpha, model_init_kwargs, dataset_kwargs)\u001b[0m\n\u001b[1;32m    254\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdataset_kwargs\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    256\u001b[0m             \u001b[0mdataset_kwargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    257\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtrain_dataset\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 258\u001b[0;31m             train_dataset = self._prepare_dataset(\n\u001b[0m\u001b[1;32m    259\u001b[0m                 \u001b[0mtrain_dataset\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    260\u001b[0m                 \u001b[0mtokenizer\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m                 \u001b[0mpacking\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, dataset, tokenizer, packing, dataset_text_field, max_seq_length, formatting_func, num_of_sequences, chars_per_token, remove_unused_columns, append_concat_token, add_special_tokens)\u001b[0m\n\u001b[1;32m    367\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mIterableDataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mConstantLengthDataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    368\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mdataset\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    369\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    370\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpacking\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 371\u001b[0;31m             return self._prepare_non_packed_dataloader(\n\u001b[0m\u001b[1;32m    372\u001b[0m                 \u001b[0mtokenizer\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    373\u001b[0m                 \u001b[0mdataset\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m                 \u001b[0mdataset_text_field\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, tokenizer, dataset, dataset_text_field, max_seq_length, formatting_func, add_special_tokens, remove_unused_columns)\u001b[0m\n\u001b[1;32m    427\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"input_ids\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0moutputs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"input_ids\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"attention_mask\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0moutputs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"attention_mask\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    428\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    429\u001b[0m         \u001b[0msignature_columns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"input_ids\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"labels\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"attention_mask\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    430\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 431\u001b[0;31m         \u001b[0mextra_colmuns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumn_names\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msignature_columns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    432\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    433\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mremove_unused_columns\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextra_colmuns\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m             warnings.warn(\n",
      "\u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5898\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5899\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5900\u001b[0m         ):\n\u001b[1;32m   5901\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5902\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'column_names'"
     ]
    }
   ],
   "source": [
    "trainer = Trainer(\n",
    "    model=model,\n",
    "    args=training_args,\n",
    "    train_dataset=df,                  #tokenized_datasets[\"train\"],\n",
    "    #eval_dataset=tokenized_datasets[\"test\"],\n",
    "    formatting_func=format_prompt,\n",
    "    tokenizer=tokenizer,\n",
    "    #data_collator=data_collator,\n",
    "    #compute_metrics=compute_metrics,\n",
    "    #add metrics\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['s', 'b', 'c']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=[[\"Your goal is to extract structured information from the user's input.\", \"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\", \"\"\"{'age': ['58'], \n",
    "            'sex': ['male'], \n",
    "            'sign symptom': ['general fatigue', 'severe anemia']\n",
    "    }\n",
    "    \"\"\"]\n",
    "df[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyarrow as pa\n",
    "import pyarrow.dataset as ds\n",
    "import pandas as pd\n",
    "from datasets import Dataset\n",
    "\n",
    "df = pd.DataFrame({'instruction': [\"Your goal is to extract structured information from the user's input.\"], \n",
    "                   'input': [\"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\"],\n",
    "                   'output' : [\"{'age': ['58'],\\n'sex': ['male'],\\n'sign symptom': ['general fatigue', 'severe anemia']}\"]\n",
    "                   })\n",
    "dataset = ds.dataset(pa.Table.from_pandas(df).to_batches())\n",
    "\n",
    "### convert to Huggingface dataset\n",
    "hg_dataset = Dataset(pa.Table.from_pandas(df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'instruction': \"Your goal is to extract structured information from the user's input.\",\n",
       " 'input': 'A 58-year-old man had been suffering from general fatigue and severe anemia for several months.',\n",
       " 'output': \"{'age': ['58'],\\n'sex': ['male'],\\n'sign symptom': ['general fatigue', 'severe anemia']}\"}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg_dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset Size: 10178\n"
     ]
    }
   ],
   "source": [
    "from datasets import load_dataset\n",
    "\n",
    "dataset = load_dataset(\"medalpaca/medical_meadow_medqa\", split=\"train\")\n",
    "\n",
    "print(f\"Dataset Size: {len(dataset)}\")\n",
    "#print(dataset[randrange(len(dataset))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "90bacd5931a74500a20cb4c51cf1db6b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:389: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:394: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig\n",
    "\n",
    "\n",
    "model_name = 'AdaptLLM/medicine-chat'\n",
    "use_flash_attention = False\n",
    "\n",
    "# BitsAndBytesConfig int-4 config\n",
    "bnb_config = BitsAndBytesConfig(\n",
    "    load_in_4bit=True,\n",
    "    bnb_4bit_use_double_quant=True,\n",
    "    bnb_4bit_quant_type=\"nf4\",\n",
    "    bnb_4bit_compute_dtype=torch.bfloat16,\n",
    ")\n",
    "\n",
    "# Load model\n",
    "model = AutoModelForCausalLM.from_pretrained(\n",
    "    model_name,\n",
    "    quantization_config=bnb_config,\n",
    "    use_cache=False,\n",
    "    use_flash_attention_2=use_flash_attention,\n",
    "    device_map=\"auto\",\n",
    "    torch_dtype=torch.float16\n",
    ")\n",
    "\n",
    "model.config.pretraining_tp = 1\n",
    "\n",
    "# Load tokenizer\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "tokenizer.pad_token = tokenizer.eos_token\n",
    "tokenizer.padding_side = \"right\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_prompt(sample):\n",
    "    return [f\"\"\"<s>[INST] <<SYS>>\n",
    "            Below is an instruction that describes a task.\n",
    "            <</SYS>>\n",
    "            ### Instruction:\n",
    "            {sample[\"instruction\"]}\n",
    "\n",
    "            ### Input:\n",
    "            {sample[\"input\"]}\n",
    "\n",
    "            ### Response:\n",
    "            {sample[\"output\"]}\n",
    "            [/INST]\n",
    "            \"\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"<s>[INST] <<SYS>>\\nBelow is an instruction that describes a task.\\n<</SYS>>\\n### Instruction:\\nYour goal is to extract structured information from the user's input.\\n\\n### Input:\\nA 58-year-old man had been suffering from general fatigue and severe anemia for several months.\\n\\n### Response:\\n{'age': ['58'],\\n'sex': ['male'],\\n'sign symptom': ['general fatigue', 'severe anemia']}\\n[/INST]\\n\""
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "format_prompt(hg_dataset[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from peft import LoraConfig, prepare_model_for_kbit_training, get_peft_model\n",
    "\n",
    "# LoRA config based on QLoRA paper\n",
    "peft_config = LoraConfig(\n",
    "    lora_alpha=32,\n",
    "    lora_dropout=0.1,\n",
    "    r=16,\n",
    "    bias=\"none\",\n",
    "    task_type=\"CAUSAL_LM\",\n",
    ")\n",
    "# Prepare model for training\n",
    "model = prepare_model_for_kbit_training(model)\n",
    "model = get_peft_model(model, peft_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import TrainingArguments\n",
    "\n",
    "args = TrainingArguments(\n",
    "    output_dir=\"../models/AdaptLLM/medicine-chat-Erik\",\n",
    "    num_train_epochs=30,\n",
    "    #per_device_train_batch_size=4,\n",
    "    gradient_accumulation_steps=1,\n",
    "    gradient_checkpointing=True,\n",
    "    #optim=\"paged_adamw_32bit\",\n",
    "    logging_steps=10,\n",
    "    save_strategy=\"epoch\",\n",
    "    learning_rate=2e-4,\n",
    "    fp16=True,\n",
    "    max_grad_norm=0.3,\n",
    "    warmup_ratio=0.03,\n",
    "    lr_scheduler_type=\"constant\",\n",
    "    disable_tqdm=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3788b378206b4052a931344d4569b1f6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/1 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Detected kernel version 5.4.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.\n"
     ]
    }
   ],
   "source": [
    "from trl import SFTTrainer\n",
    "\n",
    "max_seq_length = 100 # max sequence length for model and packing of the dataset\n",
    "\n",
    "trainer = SFTTrainer(\n",
    "    model=model,\n",
    "    train_dataset=hg_dataset,\n",
    "    peft_config=peft_config,\n",
    "    max_seq_length=max_seq_length,\n",
    "    tokenizer=tokenizer,\n",
    "    packing=False,\n",
    "    formatting_func=format_prompt,\n",
    "    args=args,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "You're using a LlamaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "      \n",
       "      <progress value='30' max='30' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      [30/30 00:50, Epoch 30/30]\n",
       "    </div>\n",
       "    <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       " <tr style=\"text-align: left;\">\n",
       "      <th>Step</th>\n",
       "      <th>Training Loss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>1.810100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>0.390600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>0.222700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-1 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-2 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-3 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-4 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-5 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-6 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-7 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-8 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-9 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-10 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-11 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-12 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-13 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-14 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-15 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-16 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-17 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-18 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-19 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-20 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-21 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-22 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-23 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-24 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-25 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-26 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-27 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-28 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-29 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "Checkpoint destination directory ../models/AdaptLLM/medicine-chat-Erik/checkpoint-30 already exists and is non-empty.Saving will proceed but saved results may be invalid.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "TrainOutput(global_step=30, training_loss=0.807794459660848, metrics={'train_runtime': 51.815, 'train_samples_per_second': 0.579, 'train_steps_per_second': 0.579, 'total_flos': 119083253760000.0, 'train_loss': 0.807794459660848, 'epoch': 30.0})"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trainer.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:61: UserWarning: None of the inputs have requires_grad=True. Gradients will be None\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "prompt = \"\"\"<s>[INST] <<SYS>>\n",
    "Below is an instruction that describes a task.\n",
    "<</SYS>>\n",
    "### Instruction:\n",
    "Your goal is to extract structured information from the user's input.\n",
    "\n",
    "### Input:\n",
    "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
    "\n",
    "### Response:\n",
    "[/INST]\n",
    "\"\"\"\n",
    "\n",
    "input_ids = tokenizer(prompt, return_tensors=\"pt\", truncation=True).input_ids.cuda()\n",
    "outputs = model.generate(input_ids=input_ids, max_new_tokens=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:61: UserWarning: None of the inputs have requires_grad=True. Gradients will be None\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=4096)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "answer_start = int(input_ids.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "print(pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pred' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m### User Input:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mprompt\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m### Assistant Output:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mpred\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pred' is not defined"
     ]
    }
   ],
   "source": [
    "print(f'### User Input:\\n{prompt}\\n\\n### Assistant Output:\\n{pred}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated Response:\n",
      "# Below is the structured response.\n",
      "\n",
      "Structured Response:\n",
      "\n",
      "Structured Response:\n",
      "\n",
      "{'age': ['58'],\n",
      " 'sex': ['Male'],\n",
      " 'symptoms': ['General fatigue'],\n",
      " 'severity': ['Severe'],\n",
      " 'duration': ['Several months'],\n",
      " 'diagnosis': ['Anemia'],\n",
      " 'treatment': ['Your doctor may recommend a blood transfusion to treat your\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"Generated Response:\\n{tokenizer.batch_decode(outputs.detach().cpu().numpy(), skip_special_tokens=True)[0][len(prompt):]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([[    1, 29871,    13, 21140,   340,   338,   385, 15278,   393, 16612,\n",
      "           263,  3414, 29892,  3300,  2859,   411,   385,  1881,   393,  8128,\n",
      "          4340,  3030, 29889, 14350,   263,  2933,   393,  7128,  2486,  1614,\n",
      "          2167,   278,  2009, 29889,    13,    13,  2277, 29937,  2799,  4080,\n",
      "         29901,    13, 10858,  7306,   338,   304,  6597,  2281,  2955,  2472,\n",
      "           515,   278,  1404, 29915, 29879,  1881, 29889,    13,    13,  2277,\n",
      "         29937, 10567, 29901,    13, 29909, 29871, 29945, 29947, 29899,  6360,\n",
      "         29899,  1025,   767,   750,  1063, 23164,   515,  2498,  9950, 12137,\n",
      "           322, 22261,   385, 29747,   363,  3196,  7378, 29889,    13,    13,\n",
      "          2277, 29937, 13291, 29901,    13, 10998,   482,  2396,  6024, 29945,\n",
      "         29947,  7464,  6024, 14167,  2396,  6024, 19202,  2033,  1118, 11117,\n",
      "           273, 29747,  2396,  6024,   344,  9359,  2033,  1118, 11117,  2484,\n",
      "           271,   358,  2396,   518, 10998,   262,  1037,   559, 29918,  1127,\n",
      "         29918, 14073,   397, 29918,  3729, 29879,  2396,  6024,  1609, 29918,\n",
      "          3286, 29888,  3958,  2033,  6525,   930,    13,    13,  5618,   916,\n",
      "         18671,   508,   367,  7436,   297,   445, 10483, 29973,    13,    13,\n",
      "          2277, 29937,  7338,  1461,   292, 29901,    13, 10998,   482,  2396,\n",
      "           518,  1839, 29945, 29947,  2033,  1402, 11117, 14167,  2396,   518,\n",
      "          1839, 19202,  2033,  1402, 11117,   273, 29747,  2396,   518,  1839,\n",
      "           344,  9359,  2033, 29962,  1118]], device='cuda:0')\n"
     ]
    }
   ],
   "source": [
    "print(outputs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=[[\"Your goal is to extract structured information from the user's input.\", \"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\", \"\"\"{'age': ['58'], \n",
    "            'sex': ['male'], \n",
    "            'sign symptom': ['general fatigue', 'severe anemia']\n",
    "    }\n",
    "    \"\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset Size: 10178\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "datasets.arrow_dataset.Dataset"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Train on completions only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "034e7ead1cd84390b85cec84c38eb188",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:389: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:394: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from peft import LoraConfig, TaskType\n",
    "from trl import SFTTrainer, DataCollatorForCompletionOnlyLM\n",
    "\n",
    "# TODO: check the lora config parameters, what do they mean and what they impact\n",
    "peft_config = LoraConfig(task_type=TaskType.CAUSAL_LM, inference_mode=False, r=64, lora_alpha=32, lora_dropout=0.1)\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer, TrainingArguments, Trainer\n",
    "from trl import AutoModelForCausalLMWithValueHead\n",
    "\n",
    "model_id = 'AdaptLLM/medicine-chat'# go for a smaller model if you dont have the VRAM\n",
    "model = AutoModelForCausalLM.from_pretrained(model_id, device_map=\"auto\", load_in_4bit=True) # , peft_config=peft_config\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_id)\n",
    "\n",
    "#from peft import get_peft_model\n",
    "\n",
    "#model = get_peft_model(model, peft_config)\n",
    "#model.print_trainable_parameters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset({\n",
      "    features: ['input', 'output'],\n",
      "    num_rows: 1\n",
      "})\n"
     ]
    }
   ],
   "source": [
    "#Dataset\n",
    "import pyarrow as pa\n",
    "import pyarrow.dataset as ds\n",
    "import pandas as pd\n",
    "from datasets import Dataset\n",
    "\n",
    "df = pd.DataFrame({'input': [\"A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\"],\n",
    "                   'output' : [\"{'age': ['58'],\\n'sex': ['male'],\\n'sign symptom': ['general fatigue', 'severe anemia']}\"],\n",
    "                   })\n",
    "dataset = ds.dataset(pa.Table.from_pandas(df).to_batches())\n",
    "\n",
    "### convert to Huggingface dataset\n",
    "hg_dataset = Dataset(pa.Table.from_pandas(df))\n",
    "print(hg_dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datasets import load_from_disk\n",
    "\n",
    "hg_dataset = load_from_disk(\"../data/dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg_dataset = hg_dataset.train_test_split(train_size=0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dataset({\n",
       "    features: ['input', 'output'],\n",
       "    num_rows: 180\n",
       "})"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg_dataset[\"train\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bd1590c2e94b4e028852c13c0761f3b6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/180 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py:290: UserWarning: You passed a tokenizer with `padding_side` not equal to `right` to the SFTTrainer. This might lead to some unexpected behaviour due to overflow issues when training a model in half-precision. You might consider adding `tokenizer.padding_side = 'right'` to your code.\n",
      "  warnings.warn(\n",
      "Detected kernel version 5.4.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.\n"
     ]
    }
   ],
   "source": [
    "def formatting_prompts_func(example):\n",
    "    output_texts = []\n",
    "    for i in range(len(example['input'])):\n",
    "        text = f\"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not match: Age, Sex, Biological_structure, Sign_symptom, Diagnostic_procedure, Detailed_description or Lab_value.\\n### Input: {example['input'][i]}\\n ### Output: {example['output'][i]}\"\n",
    "        output_texts.append(text)\n",
    "    return output_texts\n",
    "\n",
    "response_template = \"### Output:\"\n",
    "collator = DataCollatorForCompletionOnlyLM(response_template, tokenizer=tokenizer)\n",
    "\n",
    "# TODO: play around with the traning arguments, specifically the optimizer ones (learning rate, batch size, epochs)\n",
    "# Try putting in training only one example at time... batch sitze...\n",
    "args = TrainingArguments(\n",
    "    output_dir=\"../models/AdaptLLM/medicine-chat-Erik\",\n",
    "    learning_rate=2e-4,\n",
    "    num_train_epochs=10,\n",
    "    logging_steps=10,\n",
    "    gradient_checkpointing=True,\n",
    "    per_device_train_batch_size = 2,\n",
    "    #save_strategy=\"epoch\",\n",
    ")\n",
    "\n",
    "# TODO: check the SFTTrainer parameters and how to train the model\n",
    "trainer = SFTTrainer(\n",
    "    model,\n",
    "    train_dataset=hg_dataset[\"train\"],\n",
    "    formatting_func=formatting_prompts_func,\n",
    "    data_collator=collator,\n",
    "    max_seq_length=2048,\n",
    "    args=args,\n",
    "    peft_config=peft_config,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "You're using a LlamaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.\n",
      "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`...\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "      \n",
       "      <progress value='75' max='900' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      [ 75/900 08:18 < 1:33:50, 0.15 it/s, Epoch 0.82/10]\n",
       "    </div>\n",
       "    <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       " <tr style=\"text-align: left;\">\n",
       "      <th>Step</th>\n",
       "      <th>Training Loss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>0.746600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>0.606500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>0.499300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40</td>\n",
       "      <td>0.483800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50</td>\n",
       "      <td>0.455900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>60</td>\n",
       "      <td>0.437100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>70</td>\n",
       "      <td>0.433500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mtrainer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py:323\u001b[0m, in \u001b[0;36mSFTTrainer.train\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    320\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mneftune_noise_alpha \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_trainer_supports_neftune:\n\u001b[1;32m    321\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_trl_activate_neftune(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel)\n\u001b[0;32m--> 323\u001b[0m output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;66;03m# After training we make sure to retrieve back the original forward pass method\u001b[39;00m\n\u001b[1;32m    326\u001b[0m \u001b[38;5;66;03m# for the embedding layer by removing the forward post hook.\u001b[39;00m\n\u001b[1;32m    327\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mneftune_noise_alpha \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_trainer_supports_neftune:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/trainer.py:1537\u001b[0m, in \u001b[0;36mTrainer.train\u001b[0;34m(self, resume_from_checkpoint, trial, ignore_keys_for_eval, **kwargs)\u001b[0m\n\u001b[1;32m   1535\u001b[0m         hf_hub_utils\u001b[38;5;241m.\u001b[39menable_progress_bars()\n\u001b[1;32m   1536\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1537\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43minner_training_loop\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1538\u001b[0m \u001b[43m        \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1539\u001b[0m \u001b[43m        \u001b[49m\u001b[43mresume_from_checkpoint\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresume_from_checkpoint\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1540\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtrial\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrial\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1541\u001b[0m \u001b[43m        \u001b[49m\u001b[43mignore_keys_for_eval\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mignore_keys_for_eval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1542\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/trainer.py:1854\u001b[0m, in \u001b[0;36mTrainer._inner_training_loop\u001b[0;34m(self, batch_size, args, resume_from_checkpoint, trial, ignore_keys_for_eval)\u001b[0m\n\u001b[1;32m   1851\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcontrol \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcallback_handler\u001b[38;5;241m.\u001b[39mon_step_begin(args, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstate, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcontrol)\n\u001b[1;32m   1853\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maccelerator\u001b[38;5;241m.\u001b[39maccumulate(model):\n\u001b[0;32m-> 1854\u001b[0m     tr_loss_step \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtraining_step\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1856\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m   1857\u001b[0m     args\u001b[38;5;241m.\u001b[39mlogging_nan_inf_filter\n\u001b[1;32m   1858\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_torch_tpu_available()\n\u001b[1;32m   1859\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m (torch\u001b[38;5;241m.\u001b[39misnan(tr_loss_step) \u001b[38;5;129;01mor\u001b[39;00m torch\u001b[38;5;241m.\u001b[39misinf(tr_loss_step))\n\u001b[1;32m   1860\u001b[0m ):\n\u001b[1;32m   1861\u001b[0m     \u001b[38;5;66;03m# if loss is nan or inf simply add the average of previous logged losses\u001b[39;00m\n\u001b[1;32m   1862\u001b[0m     tr_loss \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m tr_loss \u001b[38;5;241m/\u001b[39m (\u001b[38;5;241m1\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstate\u001b[38;5;241m.\u001b[39mglobal_step \u001b[38;5;241m-\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_globalstep_last_logged)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/trainer.py:2744\u001b[0m, in \u001b[0;36mTrainer.training_step\u001b[0;34m(self, model, inputs)\u001b[0m\n\u001b[1;32m   2742\u001b[0m         scaled_loss\u001b[38;5;241m.\u001b[39mbackward()\n\u001b[1;32m   2743\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 2744\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maccelerator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\u001b[43mloss\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2746\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m loss\u001b[38;5;241m.\u001b[39mdetach() \u001b[38;5;241m/\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs\u001b[38;5;241m.\u001b[39mgradient_accumulation_steps\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/accelerate/accelerator.py:1964\u001b[0m, in \u001b[0;36mAccelerator.backward\u001b[0;34m(self, loss, **kwargs)\u001b[0m\n\u001b[1;32m   1962\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mscaler\u001b[38;5;241m.\u001b[39mscale(loss)\u001b[38;5;241m.\u001b[39mbackward(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m   1963\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1964\u001b[0m     \u001b[43mloss\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/_tensor.py:492\u001b[0m, in \u001b[0;36mTensor.backward\u001b[0;34m(self, gradient, retain_graph, create_graph, inputs)\u001b[0m\n\u001b[1;32m    482\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m has_torch_function_unary(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    483\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m handle_torch_function(\n\u001b[1;32m    484\u001b[0m         Tensor\u001b[38;5;241m.\u001b[39mbackward,\n\u001b[1;32m    485\u001b[0m         (\u001b[38;5;28mself\u001b[39m,),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    490\u001b[0m         inputs\u001b[38;5;241m=\u001b[39minputs,\n\u001b[1;32m    491\u001b[0m     )\n\u001b[0;32m--> 492\u001b[0m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mautograd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgradient\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minputs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/autograd/__init__.py:251\u001b[0m, in \u001b[0;36mbackward\u001b[0;34m(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)\u001b[0m\n\u001b[1;32m    246\u001b[0m     retain_graph \u001b[38;5;241m=\u001b[39m create_graph\n\u001b[1;32m    248\u001b[0m \u001b[38;5;66;03m# The reason we repeat the same comment below is that\u001b[39;00m\n\u001b[1;32m    249\u001b[0m \u001b[38;5;66;03m# some Python versions print out the first line of a multi-line function\u001b[39;00m\n\u001b[1;32m    250\u001b[0m \u001b[38;5;66;03m# calls in the traceback and some print out the last line\u001b[39;00m\n\u001b[0;32m--> 251\u001b[0m \u001b[43mVariable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_execution_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_backward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Calls into the C++ engine to run the backward pass\u001b[39;49;00m\n\u001b[1;32m    252\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtensors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgrad_tensors_\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    255\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    256\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mallow_unreachable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    258\u001b[0m \u001b[43m    \u001b[49m\u001b[43maccumulate_grad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    259\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "trainer.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.save_pretrained(\"../models/AdaptLLM/medicine-chat-Erik\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aac10f3dde974073a66b478e26405cf7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n",
      "  return self.fget.__get__(instance, owner)()\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:392: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:397: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from peft import PeftConfig, PeftModel\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "import json\n",
    "import tqdm\n",
    "\n",
    "device = \"cuda\"\n",
    "\n",
    "base_model_name = \"AdaptLLM/medicine-chat\" #path/to/your/model/or/name/on/hub\"\n",
    "adapter_model_name = \"../models/AdaptLLM/medicine-chat-Erik-shuffled\"\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(base_model_name, load_in_4bit=True)\n",
    "model = PeftModel.from_pretrained(model, adapter_model_name).to(device)\n",
    "#model = model.merge_and_unload()\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(base_model_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt = \"\"\"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not match: Age, Sex, Biological_structure, Sign_symptom, Diagnostic_procedure, Detailed_description or Lab_value.\n",
    "### Input: Emily Johnson, a 42 year old woman, visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month. Alongside the discomfort, she experienced sporadic nausea and vomiting. During the examination, a diagnostic ultrasound revealed gallstones in the gallbladder, indicative of cholecystitis.\n",
    "\n",
    "### Output:\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "generate() takes 1 positional argument but 2 were given",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 18\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m#prompt = \"Generate a person's information based on the following schema:\"\u001b[39;00m\n\u001b[1;32m     17\u001b[0m jsonformer \u001b[38;5;241m=\u001b[39m Jsonformer(model, tokenizer, json_schema, prompt)\n\u001b[0;32m---> 18\u001b[0m generated_data \u001b[38;5;241m=\u001b[39m \u001b[43mjsonformer\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28mprint\u001b[39m(generated_data)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/jsonformer/main.py:242\u001b[0m, in \u001b[0;36mJsonformer.__call__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    240\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Dict[\u001b[38;5;28mstr\u001b[39m, Any]:\n\u001b[1;32m    241\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalue \u001b[38;5;241m=\u001b[39m {}\n\u001b[0;32m--> 242\u001b[0m     generated_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate_object\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    243\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjson_schema\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mproperties\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    245\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m generated_data\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/jsonformer/main.py:147\u001b[0m, in \u001b[0;36mJsonformer.generate_object\u001b[0;34m(self, properties, obj)\u001b[0m\n\u001b[1;32m    145\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m key, schema \u001b[38;5;129;01min\u001b[39;00m properties\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m    146\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[generate_object] generating value for\u001b[39m\u001b[38;5;124m\"\u001b[39m, key)\n\u001b[0;32m--> 147\u001b[0m     obj[key] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mschema\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    148\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m obj\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/jsonformer/main.py:162\u001b[0m, in \u001b[0;36mJsonformer.generate_value\u001b[0;34m(self, schema, obj, key)\u001b[0m\n\u001b[1;32m    160\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    161\u001b[0m         obj\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgeneration_marker)\n\u001b[0;32m--> 162\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate_number\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m schema_type \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mboolean\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    164\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m key:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/jsonformer/main.py:61\u001b[0m, in \u001b[0;36mJsonformer.generate_number\u001b[0;34m(self, temperature, iterations)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m[generate_number]\u001b[39m\u001b[38;5;124m\"\u001b[39m, prompt, is_prompt\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     58\u001b[0m input_tokens \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39mencode(prompt, return_tensors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpt\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39mto(\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39mdevice\n\u001b[1;32m     60\u001b[0m )\n\u001b[0;32m---> 61\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     62\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_tokens\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     63\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmax_new_tokens\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_number_tokens\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     64\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_return_sequences\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     65\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlogits_processor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnumber_logit_processor\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     66\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstopping_criteria\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\n\u001b[1;32m     67\u001b[0m \u001b[43m        \u001b[49m\u001b[43mNumberStoppingCriteria\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtokenizer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43minput_tokens\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     68\u001b[0m \u001b[43m    \u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     69\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtemperature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemperature\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtemperature\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     70\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpad_token_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtokenizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meos_token_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     71\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     72\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtokenizer\u001b[38;5;241m.\u001b[39mdecode(response[\u001b[38;5;241m0\u001b[39m], skip_special_tokens\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     74\u001b[0m response \u001b[38;5;241m=\u001b[39m response[\u001b[38;5;28mlen\u001b[39m(prompt) :]\n",
      "\u001b[0;31mTypeError\u001b[0m: generate() takes 1 positional argument but 2 were given"
     ]
    }
   ],
   "source": [
    "from jsonformer import Jsonformer\n",
    "\n",
    "json_schema = {\n",
    "    \"type\": \"object\",\n",
    "    \"properties\": {\n",
    "        \"age\": {\"type\": \"number\"},\n",
    "        \"sex\": {\"type\": \"string\"},\n",
    "        #\"Sign_symptom\": \"array\",\n",
    "        #\"Sign_symptom\": {\n",
    "        #    \"type\": \"array\",\n",
    "        #    \"items\": {\"type\": \"string\"}\n",
    "        #}\n",
    "    }\n",
    "}\n",
    "\n",
    "#prompt = \"Generate a person's information based on the following schema:\"\n",
    "jsonformer = Jsonformer(model, tokenizer, json_schema, prompt)\n",
    "generated_data = jsonformer()\n",
    "\n",
    "print(generated_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0328a89a95ef4d9aa286061c6edd61a5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "config.json:   0%|          | 0.00/818 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6fe3cf334b5d4c54a7bbb2342a98f93e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "pytorch_model.bin:   0%|          | 0.00/23.8G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mjsonformer\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Jsonformer\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtransformers\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AutoModelForCausalLM, AutoTokenizer\n\u001b[0;32m----> 4\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43mAutoModelForCausalLM\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdatabricks/dolly-v2-12b\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m tokenizer \u001b[38;5;241m=\u001b[39m AutoTokenizer\u001b[38;5;241m.\u001b[39mfrom_pretrained(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdatabricks/dolly-v2-12b\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      7\u001b[0m json_schema \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtype\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mproperties\u001b[39m\u001b[38;5;124m\"\u001b[39m: {\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     17\u001b[0m     }\n\u001b[1;32m     18\u001b[0m }\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/auto/auto_factory.py:566\u001b[0m, in \u001b[0;36m_BaseAutoModelClass.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, *model_args, **kwargs)\u001b[0m\n\u001b[1;32m    564\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(config) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_model_mapping\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m    565\u001b[0m     model_class \u001b[38;5;241m=\u001b[39m _get_model_class(config, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_model_mapping)\n\u001b[0;32m--> 566\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmodel_class\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    567\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpretrained_model_name_or_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmodel_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mhub_kwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m    568\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    569\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    570\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnrecognized configuration class \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mconfig\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m for this kind of AutoModel: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    571\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mModel type should be one of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(c\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mfor\u001b[39;00m\u001b[38;5;250m \u001b[39mc\u001b[38;5;250m \u001b[39m\u001b[38;5;129;01min\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_model_mapping\u001b[38;5;241m.\u001b[39mkeys())\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    572\u001b[0m )\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/modeling_utils.py:3412\u001b[0m, in \u001b[0;36mPreTrainedModel.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, config, cache_dir, ignore_mismatched_sizes, force_download, local_files_only, token, revision, use_safetensors, *model_args, **kwargs)\u001b[0m\n\u001b[1;32m   3409\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   3410\u001b[0m         \u001b[38;5;66;03m# This repo has no safetensors file of any kind, we switch to PyTorch.\u001b[39;00m\n\u001b[1;32m   3411\u001b[0m         filename \u001b[38;5;241m=\u001b[39m _add_variant(WEIGHTS_NAME, variant)\n\u001b[0;32m-> 3412\u001b[0m         resolved_archive_file \u001b[38;5;241m=\u001b[39m \u001b[43mcached_file\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3413\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpretrained_model_name_or_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcached_file_kwargs\u001b[49m\n\u001b[1;32m   3414\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3415\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m resolved_archive_file \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m filename \u001b[38;5;241m==\u001b[39m _add_variant(WEIGHTS_NAME, variant):\n\u001b[1;32m   3416\u001b[0m     \u001b[38;5;66;03m# Maybe the checkpoint is sharded, we try to grab the index name in this case.\u001b[39;00m\n\u001b[1;32m   3417\u001b[0m     resolved_archive_file \u001b[38;5;241m=\u001b[39m cached_file(\n\u001b[1;32m   3418\u001b[0m         pretrained_model_name_or_path,\n\u001b[1;32m   3419\u001b[0m         _add_variant(WEIGHTS_INDEX_NAME, variant),\n\u001b[1;32m   3420\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mcached_file_kwargs,\n\u001b[1;32m   3421\u001b[0m     )\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/utils/hub.py:385\u001b[0m, in \u001b[0;36mcached_file\u001b[0;34m(path_or_repo_id, filename, cache_dir, force_download, resume_download, proxies, token, revision, local_files_only, subfolder, repo_type, user_agent, _raise_exceptions_for_missing_entries, _raise_exceptions_for_connection_errors, _commit_hash, **deprecated_kwargs)\u001b[0m\n\u001b[1;32m    382\u001b[0m user_agent \u001b[38;5;241m=\u001b[39m http_user_agent(user_agent)\n\u001b[1;32m    383\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    384\u001b[0m     \u001b[38;5;66;03m# Load from URL or cache if already cached\u001b[39;00m\n\u001b[0;32m--> 385\u001b[0m     resolved_file \u001b[38;5;241m=\u001b[39m \u001b[43mhf_hub_download\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    386\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpath_or_repo_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    387\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    388\u001b[0m \u001b[43m        \u001b[49m\u001b[43msubfolder\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msubfolder\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43msubfolder\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    389\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrepo_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrepo_type\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    390\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrevision\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrevision\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    391\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcache_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcache_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    392\u001b[0m \u001b[43m        \u001b[49m\u001b[43muser_agent\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muser_agent\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    393\u001b[0m \u001b[43m        \u001b[49m\u001b[43mforce_download\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mforce_download\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    394\u001b[0m \u001b[43m        \u001b[49m\u001b[43mproxies\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mproxies\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    395\u001b[0m \u001b[43m        \u001b[49m\u001b[43mresume_download\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresume_download\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    396\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtoken\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtoken\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    397\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlocal_files_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlocal_files_only\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    398\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    399\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m GatedRepoError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    400\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mEnvironmentError\u001b[39;00m(\n\u001b[1;32m    401\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou are trying to access a gated repo.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mMake sure to request access at \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    402\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://huggingface.co/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpath_or_repo_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m and pass a token having permission to this repo either \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    403\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mby logging in with `huggingface-cli login` or by passing `token=<your_token>`.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    404\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/huggingface_hub/utils/_validators.py:118\u001b[0m, in \u001b[0;36mvalidate_hf_hub_args.<locals>._inner_fn\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m check_use_auth_token:\n\u001b[1;32m    116\u001b[0m     kwargs \u001b[38;5;241m=\u001b[39m smoothly_deprecate_use_auth_token(fn_name\u001b[38;5;241m=\u001b[39mfn\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m, has_token\u001b[38;5;241m=\u001b[39mhas_token, kwargs\u001b[38;5;241m=\u001b[39mkwargs)\n\u001b[0;32m--> 118\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/huggingface_hub/file_download.py:1457\u001b[0m, in \u001b[0;36mhf_hub_download\u001b[0;34m(repo_id, filename, subfolder, repo_type, revision, library_name, library_version, cache_dir, local_dir, local_dir_use_symlinks, user_agent, force_download, force_filename, proxies, etag_timeout, resume_download, token, local_files_only, legacy_cache_layout, endpoint)\u001b[0m\n\u001b[1;32m   1454\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m local_dir \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1455\u001b[0m             _check_disk_space(expected_size, local_dir)\n\u001b[0;32m-> 1457\u001b[0m     \u001b[43mhttp_get\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1458\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl_to_download\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1459\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemp_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1460\u001b[0m \u001b[43m        \u001b[49m\u001b[43mproxies\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mproxies\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1461\u001b[0m \u001b[43m        \u001b[49m\u001b[43mresume_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresume_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1462\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1463\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexpected_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpected_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1464\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1466\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m local_dir \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1467\u001b[0m     logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mStoring \u001b[39m\u001b[38;5;132;01m{\u001b[39;00murl\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m in cache at \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mblob_path\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/huggingface_hub/file_download.py:524\u001b[0m, in \u001b[0;36mhttp_get\u001b[0;34m(url, temp_file, proxies, resume_size, headers, expected_size, _nb_retries)\u001b[0m\n\u001b[1;32m    522\u001b[0m new_resume_size \u001b[38;5;241m=\u001b[39m resume_size\n\u001b[1;32m    523\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 524\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m chunk \u001b[38;5;129;01min\u001b[39;00m r\u001b[38;5;241m.\u001b[39miter_content(chunk_size\u001b[38;5;241m=\u001b[39mDOWNLOAD_CHUNK_SIZE):\n\u001b[1;32m    525\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m chunk:  \u001b[38;5;66;03m# filter out keep-alive new chunks\u001b[39;00m\n\u001b[1;32m    526\u001b[0m             progress\u001b[38;5;241m.\u001b[39mupdate(\u001b[38;5;28mlen\u001b[39m(chunk))\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/requests/models.py:816\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[0;34m()\u001b[0m\n\u001b[1;32m    814\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    815\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 816\u001b[0m         \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mstream(chunk_size, decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    817\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m ProtocolError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    818\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ChunkedEncodingError(e)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/urllib3/response.py:934\u001b[0m, in \u001b[0;36mHTTPResponse.stream\u001b[0;34m(self, amt, decode_content)\u001b[0m\n\u001b[1;32m    932\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    933\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_fp_closed(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decoded_buffer) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 934\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mamt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    936\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m data:\n\u001b[1;32m    937\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m data\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/urllib3/response.py:877\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt, decode_content, cache_content)\u001b[0m\n\u001b[1;32m    874\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decoded_buffer) \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m amt:\n\u001b[1;32m    875\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decoded_buffer\u001b[38;5;241m.\u001b[39mget(amt)\n\u001b[0;32m--> 877\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raw_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    879\u001b[0m flush_decoder \u001b[38;5;241m=\u001b[39m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m (amt \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data)\n\u001b[1;32m    881\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decoded_buffer) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/urllib3/response.py:812\u001b[0m, in \u001b[0;36mHTTPResponse._raw_read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    809\u001b[0m fp_closed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclosed\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    811\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_error_catcher():\n\u001b[0;32m--> 812\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fp_closed \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    813\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m amt \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data:\n\u001b[1;32m    814\u001b[0m         \u001b[38;5;66;03m# Platform-specific: Buggy versions of Python.\u001b[39;00m\n\u001b[1;32m    815\u001b[0m         \u001b[38;5;66;03m# Close the connection when no data is returned\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    820\u001b[0m         \u001b[38;5;66;03m# not properly close the connection in all cases. There is\u001b[39;00m\n\u001b[1;32m    821\u001b[0m         \u001b[38;5;66;03m# no harm in redundantly calling close.\u001b[39;00m\n\u001b[1;32m    822\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/urllib3/response.py:789\u001b[0m, in \u001b[0;36mHTTPResponse._fp_read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    787\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    788\u001b[0m     chunk_amt \u001b[38;5;241m=\u001b[39m max_chunk_amt\n\u001b[0;32m--> 789\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchunk_amt\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    790\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data:\n\u001b[1;32m    791\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:463\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    460\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    461\u001b[0m     \u001b[38;5;66;03m# Amount is given, implement using readinto\u001b[39;00m\n\u001b[1;32m    462\u001b[0m     b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mbytearray\u001b[39m(amt)\n\u001b[0;32m--> 463\u001b[0m     n \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadinto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    464\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mmemoryview\u001b[39m(b)[:n]\u001b[38;5;241m.\u001b[39mtobytes()\n\u001b[1;32m    465\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    466\u001b[0m     \u001b[38;5;66;03m# Amount is not given (unbounded read) so we must check self.length\u001b[39;00m\n\u001b[1;32m    467\u001b[0m     \u001b[38;5;66;03m# and self.chunked\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/http/client.py:507\u001b[0m, in \u001b[0;36mHTTPResponse.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    502\u001b[0m         b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmemoryview\u001b[39m(b)[\u001b[38;5;241m0\u001b[39m:\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength]\n\u001b[1;32m    504\u001b[0m \u001b[38;5;66;03m# we do not use _safe_read() here because this may be a .will_close\u001b[39;00m\n\u001b[1;32m    505\u001b[0m \u001b[38;5;66;03m# connection, and the user is reading more bytes than will be provided\u001b[39;00m\n\u001b[1;32m    506\u001b[0m \u001b[38;5;66;03m# (for example, reading in 1k chunks)\u001b[39;00m\n\u001b[0;32m--> 507\u001b[0m n \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadinto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    508\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m n \u001b[38;5;129;01mand\u001b[39;00m b:\n\u001b[1;32m    509\u001b[0m     \u001b[38;5;66;03m# Ideally, we would raise IncompleteRead if the content-length\u001b[39;00m\n\u001b[1;32m    510\u001b[0m     \u001b[38;5;66;03m# wasn't satisfied, but it might break compatibility.\u001b[39;00m\n\u001b[1;32m    511\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_close_conn()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/socket.py:704\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    703\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 704\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    706\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/ssl.py:1241\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1237\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1238\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1239\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1240\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1241\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1242\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1243\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/ssl.py:1099\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1097\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1098\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1099\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1100\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1101\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from jsonformer import Jsonformer\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(\"databricks/dolly-v2-12b\")\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"databricks/dolly-v2-12b\")\n",
    "\n",
    "json_schema = {\n",
    "    \"type\": \"object\",\n",
    "    \"properties\": {\n",
    "        \"name\": {\"type\": \"string\"},\n",
    "        \"age\": {\"type\": \"number\"},\n",
    "        \"is_student\": {\"type\": \"boolean\"},\n",
    "        \"courses\": {\n",
    "            \"type\": \"array\",\n",
    "            \"items\": {\"type\": \"string\"}\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "prompt = \"Generate a person's information based on the following schema:\"\n",
    "jsonformer = Jsonformer(model, tokenizer, json_schema, prompt)\n",
    "generated_data = jsonformer()\n",
    "\n",
    "print(generated_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:392: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:397: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/bitsandbytes/nn/modules.py:226: UserWarning: Input type into Linear4bit is torch.float16, but bnb_4bit_compute_dtype=torch.float32 (default). This will lead to slow inference or training speed.\n",
      "  warnings.warn(f'Input type into Linear4bit is torch.float16, but bnb_4bit_compute_dtype=torch.float32 (default). This will lead to slow inference or training speed.')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1. Age: 42\n",
      "2. Sex: Female\n",
      "3. Biological_structure: Gallbladder\n",
      "4. Sign_symptom: Abdominal pain\n",
      "5. Diagnostic_procedure: Ultrasound\n",
      "6. Detailed_description: Gallstones in the gallbladder\n",
      "7. Lab_value: Not provided\n",
      "\n",
      "### Hints:\n",
      "\n",
      "1. Age: The patient's age is 42 years old.\n",
      "2. Sex: The patient is a female.\n",
      "3. Biological_structure: The patient's abdominal pain is caused by gallstones in the gallbladder.\n",
      "4. Sign_symptom: The patient is experiencing abdominal pain and nausea.\n",
      "5. Diagnostic_procedure: An ultrasound was performed to diagnose the patient's condition.\n",
      "6. Detailed_description: The patient has gallstones in the gallbladder.\n",
      "7. Lab_value: Not provided in the input.\n"
     ]
    }
   ],
   "source": [
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=1000)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "print(pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pred = pred.replace(\"\\'\", \"\\\"\")\n",
    "sep = \"}\"\n",
    "pred = pred.split(sep, 1)[0] + sep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\n1. Age: 42\\n2. Sex: Female\\n3. Biological_structure: Gallbladder\\n4. Sign_symptom: Abdominal pain\\n5. Diagnostic_procedure: Ultrasound\\n6. Detailed_description: Gallstones in the gallbladder\\n7. Lab_value: Not provided\\n\\n### Hints:\\n\\n1. Age: The patient's age is 42 years old.\\n2. Sex: The patient is a female.\\n3. Biological_structure: The patient's abdominal pain is caused by gallstones in the gallbladder.\\n4. Sign_symptom: The patient is experiencing abdominal pain and nausea.\\n5. Diagnostic_procedure: An ultrasound was performed to diagnose the patient's condition.\\n6. Detailed_description: The patient has gallstones in the gallbladder.\\n7. Lab_value: Not provided in the input.\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "#employee_string = '{\"name\": \"erik\"}'\n",
    "json_pred = json.loads(pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['abdominal', 'gallbladder', 'upper right quadrant']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_pred[\"Biological_structure\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'json' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m json_object \u001b[38;5;241m=\u001b[39m \u001b[43mjson\u001b[49m\u001b[38;5;241m.\u001b[39mloads(pred)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'json' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datasets import load_from_disk\n",
    "\n",
    "hg_dataset = load_from_disk(\"../data/datasetTestShuffledEntireAllEntities\")\n",
    "#hg_dataset = hg_dataset.train_test_split(train_size=0.9)\n",
    "\n",
    "#test_dataset=hg_dataset[\"test\"]\n",
    "test_dataset = hg_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dataset({\n",
       "    features: ['input', 'output', 'keys'],\n",
       "    num_rows: 20\n",
       "})"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This way wanted output and to compare with generated output. za vsako poglej če je v generiranem\n",
    "#test_dataset[0][\"output\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatting_input_func(example):\n",
    "    return f\"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not match: Age, Sex, Biological_structure, Sign_symptom, Diagnostic_procedure, Detailed_description, Lab_value.\\n### Input: {example['input']}\\n ### Output: \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatting_input_func(example):\n",
    "    return f\"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Lab_value. Do not add any extra entities.\\n\\n### Input:\\n{example['input']}\\n\\n### Output: \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not match: Age, Sex, Biological_structure, Sign_symptom, Diagnostic_procedure, Detailed_description or Lab_value.\\n### Input: Our patient was a 7-year-old Italian boy born after an uneventful gestation of normal duration.\\nAt the age of 16 months, he presented with a clinically evident enlarged abdomen and was referred for oncological examination.\\nInitial tests revealed anemia, thrombocytopenia, and splenomegaly.\\nA bone marrow biopsy revealed the presence of foam cells, which led to suspicion of lysosomal storage disease.\\nBiochemical testing revealed elevated level of acid phosphatase (47.8 IU/L [normal range 5–7 IU/L]) and chitotriosidase activity (508 nmol/mg protein [normal range 5.9–41.0 nmol/mg protein]), as well as reduced beta-glucosidase activity (2 nmol/mg/protein [normal range 4.5–18.0 nmol/mg/protein]).\\nMolecular analysis showed homozygous L444P mutations in the GBA gene (OMIM reference 606463), confirming the diagnosis of chronic NGD.\\nThe patient began ERT at a dosage of 60 U/kg every 2 weeks at the age of 18 months.\\nThereafter, when the patient reached the age of 30 months, we decided to combine ERT with SRT with miglustat.\\nThis clinical decision was approved by our institution’s ethics committee, and informed consent was obtained from the patient’s parents before commencing combination therapy.\\nThe dose of miglustat was adjusted according to the patient’s body surface area and was uptitrated during the first 4 weeks with the following scheme: one-third of target dose during weeks 1 and 2, two-thirds of target dose during weeks 3 and 4, and target dose (100 mg three times daily) after 1 month.\\nFrom 2 weeks before starting miglustat therapy, the patient also followed specific dietary modifications, avoiding high intake of carbohydrate-containing food in single meals, especially foods high in disaccharides, such as sucrose and maltose, to ensure acceptable gastrointestinal tolerability.\\nHe experienced mild episodes of diarrhea after commencing miglustat therapy, which decreased in frequency/severity over time.\\nFrom the start of ERT/miglustat combination treatment, we observed a reduction in splenomegaly and a gradual normalization of hematological values and plasma angiotensin-converting enzyme activity (Table 1).\\nPlasma chitotriosidase was evaluated at the time of diagnosis and then approximately every 6 months during follow-up.\\nValues showed an initial reduction after the start of ERT, with a further, substantial, and sustained decrease after commencement of miglustat treatment (Fig.1).\\nThe patient has been followed according to recommended guidelines for GD, which specify a complete neurological examination, clinical evaluation of ocular movements, and psychological evaluations every 6–12 months [10].\\nAfter 5 years of combination therapy and follow-up, the patient did not show any signs of neurological impairment.\\nAs of February 2016, he had not displayed any epileptic crises and had demonstrated clinical performance and cooperation.\\nHe showed good muscular tone and trophism, normal reflexes with a slight hyperreflexia in his legs, and a negative Romberg sign.\\nHis toe and heel deambulation was normal.\\nIn particular, ocular evaluations revealed no evidence of saccadic movement velocity reduction and normal visual evoked potentials.\\nThe patient’s auditory brain responses were also normal.\\nIn addition, he has not demonstrated any cognitive impairment, and he has regularly attended school with good performance since the age of 5 years.\\n\\n ### Output: \""
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "formatting_input_func(test_dataset[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"A 68-year-old man was referred by his optometrist to HES with suspected LTG due to repeatedly irregular visual field test results, advanced optic disc cupping, normal intraocular pressures (IOPs) and a family history of glaucoma.\\nThe patient subjectively felt that vision in his ‘good’ left eye (LE), which normally had a visual acuity of 6/6 N5, started to deteriorate 6\\u2005months earlier; at the point of referral it was best corrected to 6/7.5 N6.\\nHis right eye (RE) was known to be amblyopic with a visual acuity of 6/18 N12.\\nHis medical history included considerable risk factors for systemic vasculopathy, such as hypertension, hypercholesterolaemia, 50 pack-years of smoking and type 2 diabetes with no diabetic retinopathy.\\nDespite detailed questioning, he denied any new systemic symptoms apart from experiencing increased lethargy.\\nClinical examination at HES revealed advanced bilateral cupped optic discs with a cup-to-disc ratio right 0.9 (90%) and left 0.8 (80%; figure 1).\\nFurthermore, the findings from the referring optometrist of equal and reactive pupils, normal eye movements and IOPs in mid-teens were confirmed.\\nComputerised perimetry showed constricted visual fields in both eyes, albeit with considerably raised false-negative error rates in excess of 30% (figure 2A).\\nSince this can be a result of poor concentration and attention span with testing compounded, especially with an amblyopic RE, the field test was repeated 4\\u2005weeks later with similar results (figure 2B).\\nAt this visit, the patient experienced further deterioration of his ‘good’ LE.\\nVisual acuities measured 6/24 (loss of one Snellen line) N36 in his amblyopic RE, and 6/18 (loss of two Snellen lines) N36 in his LE, which now also showed depressed colour vision.\\nAn urgent scan of the patient's head and orbit was requested to rule out a compressive lesion and blood analysis was carried out to screen for other causes of optic neuropathy including vitamin B12, folate levels and inflammatory markers.\\nMRI revealed a large mass centred at the pituitary fossa measuring approximately 4.5\\u2005cm×5.1\\u2005cm× 4.8\\u2005cm (figure 3A–C) extending into the suprasellar cistern and infrasellar sphenoid sinuses.\\nThe optic chiasm and both optic nerves were displaced superiorly and compressed on both sides.\\nThe mass showed heterogeneous T1 and T2 signals with multiple fluid levels most likely due to previous haemorrhage and low-signal components suggesting calcifications.\\nThe appearance was probably due to a craniopharyngioma or haemorrhagic pituitary adenoma.\\nThe patient was referred urgently to the neurosurgery unit and was found to have markedly raised prolactin levels of greater than 200\\u2005000\\u2005mIU/L, confirming the diagnosis of a giant prolactinoma.\\nIn cooperation with the local endocrinology service, he was started on cabergoline 500\\u2005µg and dexamethasone 8\\u2005mg.\\nWithin a week this led to a dramatic improvement of the patient's visual fields and acuities to 6/18 RE and to 6/9 LE.\\nHis prolactin levels showed a similar dramatic response with levels coming down to 336\\u2005mIU/L.\\nThree months later his visual acuity in his RE and LE were 6/18 and 6/5, respectively, and visual field examination showed striking recovery with almost normal results (figure 2C).\\n\""
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dataset['input'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entity: Age. Do not add any extra entities. Example: {'age': ['32-year-old']}\\n\\n### Input:\\ntesting!\\n\\n### Output:\""
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.prompts import prompts\n",
    "\n",
    "prompt = prompts.age(\"testing!\")\n",
    "prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Age': ['68-year-old']}\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.generate import generate\n",
    "\n",
    "gene = generate.gene(model, tokenizer, prompt)\n",
    "json_pred = generate.json_format(gene)\n",
    "print(json_pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['None']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"None\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Something went wrong with json.loads(pred)\n",
      "\n",
      "{\"Sign_symptom\": [\"CD20+ B lymphocyte infiltrate\", \"arteriolar thrombus\", \"ascites\", \"biologic manifestation\", \"biologic sign of hemolysis\", \"deposit\", \"diarrhea\", \"diarrhea\", \"edema\", \"enlargement\", \"follicles\", \"hematuria\", \"hemiplegia\", \"histological lesions\", \"histological lesions\", \"hyalinization\", \"hyaline-vascular multicentric Castleman disease\", \"infiltrate\", \"infection\", \"ischemic brain lesions\", \"ischemic lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\", \"lesions\",}\n",
      "Something went wrong with json.loads(pred)\n",
      "\n",
      "{\"Lab_value\": [\"140/70 mm Hg\", \"14%\", \"39°C\", \"825 pg/mL\", \"86 μmol/L\", \"87%\", \"decreased\", \"decreased\", \"elevated\", \"elevated\", \"highly elevated\", \"low\", \"normal\", \"normalized\", \"normalized\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\", \"no\",}\n",
      "Something went wrong with json.loads(pred)\n",
      "\n",
      "{\"Diagnostic_procedure\": [\"CD20+ B lymphocyte\", \"CT\", \"CT\", \"Clinical presentation\", \"Clinical manifestations\", \"C-reactive protein\", \"HIV\", \"HHV-8\", \"HHV-8 lymph node tissue staining\", \"Immunofluorescence study\", \"Lymph node biopsy\", \"MRI\", \"MRI\", \"PAS\", \"Plasma\", \"Plasma\", \"Plasma\", \"Plasma\", \"Physical examination\", \"Rh\", \"Serum\", \"Serum\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF\", \"VEGF}\n",
      "Activity\n",
      "----------\n",
      "Administration\n",
      "----------\n",
      "Age\n",
      "['54-year-old']\n",
      "['54-year-old']\n",
      "['54-year-old']\n",
      "1.0 1.0 1.0\n",
      "----------------------\n",
      "Area\n",
      "----------\n",
      "Biological_attribute\n",
      "----------\n",
      "Biological_structure\n",
      "['interstitium', 'Thoraco-abdominal', 'arteriolar', 'Lymph node', 'axillar', 'lymph node', 'follicles', 'neurologic', 'glomeruli', 'germinal center', 'kidney', 'liver', 'spleen', 'Cervical', 'brain', 'ankles', 'mantle zone', 'lower and upper limbs', 'renal', 'joints', 'glomerular']\n",
      "['interstitium', 'kidney', 'neurologic', 'both lower and upper limbs', 'glomeruli', 'renal', 'spleen', 'Cervical', 'axillar', 'brain', 'joints', 'thoraco-abdominal', 'lymph node', 'peritubularcapillaritis']\n",
      "['neurologic', 'interstitium', 'kidney', 'glomeruli', 'renal', 'spleen', 'Cervical', 'axillar', 'brain', 'joints', 'lymph node']\n",
      "0.7857142857142857 0.5238095238095238 0.6285714285714286\n",
      "----------------------\n",
      "Clinical_event\n",
      "----------\n",
      "Color\n",
      "----------\n",
      "Coreference\n",
      "----------\n",
      "Date\n",
      "----------\n",
      "Detailed_description\n",
      "----------\n",
      "Diagnostic_procedure\n",
      "----------\n",
      "Disease_disorder\n",
      "----------\n",
      "Distance\n",
      "----------\n",
      "Dosage\n",
      "----------\n",
      "Duration\n",
      "----------\n",
      "Family_history\n",
      "----------\n",
      "Frequency\n",
      "----------\n",
      "Height\n",
      "----------\n",
      "History\n",
      "----------\n",
      "Lab_value\n",
      "----------\n",
      "Medication\n",
      "----------\n",
      "Nonbiological_location\n",
      "----------\n",
      "Other_event\n",
      "----------\n",
      "Outcome\n",
      "----------\n",
      "Personal_background\n",
      "----------\n",
      "Qualitative_concept\n",
      "----------\n",
      "Quantitative_concept\n",
      "----------\n",
      "Severity\n",
      "----------\n",
      "Sex\n",
      "['None']\n",
      "['male']\n",
      "[]\n",
      "0.0 0.0 0\n",
      "----------------------\n",
      "Shape\n",
      "----------\n",
      "Sign_symptom\n",
      "----------\n",
      "Subject\n",
      "----------\n",
      "Texture\n",
      "----------\n",
      "Therapeutic_procedure\n",
      "----------\n",
      "Time\n",
      "----------\n",
      "Volume\n",
      "----------\n",
      "Weight\n",
      "----------\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import importlib\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.generate import generate\n",
    "\n",
    "compare = \"\"\n",
    "#entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\"]\n",
    "#entities = [\"Age\", \"Sex\"]\n",
    "#entities = [\"Age\", \"Biological_structure\", \"Sex\"]\n",
    "#entities = [\"Sign_symptom\"]\n",
    "entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\", \"Biological_structure\", \"Diagnostic_procedure\"]\n",
    "\n",
    "text = test_dataset['input'][12]\n",
    "extract = generate.extract(model, tokenizer, entities, text)\n",
    "\n",
    "for x in test_dataset['output'][12]:\n",
    "        print(x)\n",
    "        if x not in extract:\n",
    "            print(\"----------\")\n",
    "            continue\n",
    "\n",
    "\n",
    "        if test_dataset['output'][12][x] is None:\n",
    "            real = [\"None\"]\n",
    "        else:\n",
    "            real = list(set(test_dataset['output'][12][x]))\n",
    "\n",
    "        gene = list(set(extract[x]))\n",
    "\n",
    "        match = list(set([r for r in real if r in gene]))\n",
    "\n",
    "        print(real)\n",
    "        print(gene)\n",
    "        print(match)\n",
    "\n",
    "        lenmatch = len(match)\n",
    "        lenreal = len(real)\n",
    "        lengene = len(gene)\n",
    "            \n",
    "        text = f\"{x} \\nReal: {test_dataset['output'][12][x]} \\nGene: {extract[x]} \\nMatch: {match} \\n\\n\"\n",
    "        compare = compare + text\n",
    "\n",
    "        precision = lenmatch/lengene\n",
    "        recall = lenmatch/lenreal\n",
    "        if precision == 0 and recall == 0:\n",
    "             f1 = 0\n",
    "        else:\n",
    "            f1 = 2*(precision*recall) / (precision+recall)\n",
    "        print(precision, recall, f1)\n",
    "        print(\"----------------------\")\n",
    "    #print(precision, recall, f1)\n",
    "\n",
    "    #F1score[x].append(f1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biological_structure\n",
      "['chest', 'brain', 'skin', 'quadriceps muscle', 'cardiac', 'muscle', 'Head', 'Chest', 'subcortical', 'ventricles', 'Eye', 'cortical', 'head', 'white matter cortical', 'occipital region', 'Brain', 'pulmonary']\n",
      "['chest', 'ventricular', 'cardiac', 'Chest', 'ventricles', 'occipital region', 'Brain']\n",
      "['chest', 'cardiac', 'Chest', 'ventricles', 'occipital region', 'Brain']\n",
      "0.8571428571428571 0.35294117647058826 0.5\n",
      "----------------------\n",
      "Detailed_description\n",
      "----------\n",
      "Diagnostic_procedure\n",
      "['Birth weight', 'head circumference', 'lactate', 'sequencing', 'ultrasound', 'Physical exam', 'alanine', 'cytochrome c oxidase stain', 'US', 'Echo', 'organic acids', 'CT', 'ABR', 'biopsy', 'Hearing test', 'Metabolic work up', 'Pathology', 'acylcarnitines', 'ammonia', 'X-ray', 'Apgar', 'examination']\n",
      "['head circumference', 'lactate', 'Physical exam', 'alanine', 'cytochrome c oxidase stain', 'lactic acidosis', 'Echo', 'CT', 'ABR', 'biopsy', 'Hearing test', 'Metabolic work up', 'Pathology', 'Eye examination', 'acylcarnitines', 'cardiac Echo', 'ammonia', 'Head CT', 'exome sequencing', 'X-ray', 'urine organic acids']\n",
      "['Hearing test', 'head circumference', 'lactate', 'Metabolic work up', 'Pathology', 'ammonia', 'X-ray', 'Echo', 'Physical exam', 'alanine', 'cytochrome c oxidase stain', 'acylcarnitines', 'CT', 'ABR', 'biopsy']\n",
      "0.7142857142857143 0.6818181818181818 0.6976744186046512\n",
      "----------------------\n",
      "Lab_value\n",
      "['elevated', '2.3\\u2009kg', '10th percentile', '1400\\u2009μm', '27\\u2009mm', '32\\u2009cm', '9', 'grew rapidly', '10', 'adding 13 centimeters', 'normal', 'decreased', '3/6', '277\\u2009μm']\n",
      "['subcortial and white matter', '9', 'grew rapidly', '13 centimeters', '10', 'rapidly', 'normal', '277\\u2009μm']\n",
      "['9', 'grew rapidly', '10', 'normal', '277\\u2009μm']\n",
      "0.625 0.35714285714285715 0.45454545454545453\n",
      "----------------------\n",
      "Sex\n",
      "['male']\n",
      "['male']\n",
      "['male']\n",
      "1.0 1.0 1.0\n",
      "----------------------\n",
      "Sign_symptom\n",
      "['edema', 'dilated', 'dilatation', 'lactic acidosis', 'hypoxemia', 'systolic murmur', 'fevers', 'tricuspid regurge', 'hemorrhage', 'LVH', 'cardiomegaly', 'ketones', 'left ventricular hypertrophy', 'pulmonary hypertension', 'PHTN']\n",
      "['subcortial and white matter cortical hemorrhage', 'hypotonia', 'dilatation', 'infections', 'metabolic acidosis', 'hypoxemia', 'febres', 'shunt malfunctioning', 'febrile illness', 'hypertrophy', 'cardiomegaly', 'hypertrophic', 'left ventricular hypertrophy', 'pulmonary hypertension', 'pulmonary edema', 'murmur']\n",
      "['dilatation', 'hypoxemia', 'cardiomegaly', 'left ventricular hypertrophy', 'pulmonary hypertension']\n",
      "0.3125 0.3333333333333333 0.3225806451612903\n",
      "----------------------\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import importlib\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.generate import generate\n",
    "\n",
    "compare = \"\"\n",
    "#entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\"]\n",
    "#entities = [\"Age\", \"Sex\"]\n",
    "#entities = [\"Age\", \"Biological_structure\", \"Sex\"]\n",
    "#entities = [\"Sign_symptom\"]\n",
    "entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\", \"Biological_structure\", \"Diagnostic_procedure\"]\n",
    "\n",
    "text = test_dataset['input'][13]\n",
    "extract = generate.extract(model, tokenizer, entities, text)\n",
    "\n",
    "for x in json.loads(test_dataset['output'][11]):\n",
    "        print(x)\n",
    "        if x not in extract:\n",
    "            print(\"----------\")\n",
    "            continue\n",
    "           \n",
    "\n",
    "        real = list(set(json.loads(test_dataset['output'][13])[x]))\n",
    "        gene = list(set(extract[x]))\n",
    "\n",
    "        match = list(set([r for r in real if r in gene]))\n",
    "\n",
    "        print(real)\n",
    "        print(gene)\n",
    "        print(match)\n",
    "\n",
    "        lenmatch = len(match)\n",
    "        lenreal = len(real)\n",
    "        lengene = len(gene)\n",
    "            \n",
    "        text = f\"{x} \\nReal: {json.loads(test_dataset['output'][13])[x]} \\nGene: {extract[x]} \\nMatch: {match} \\n\\n\"\n",
    "        compare = compare + text\n",
    "\n",
    "        precision = lenmatch/lengene\n",
    "        recall = lenmatch/lenreal\n",
    "        f1 = 2*(precision*recall) / (precision+recall)\n",
    "        print(precision, recall, f1)\n",
    "        print(\"----------------------\")\n",
    "    #print(precision, recall, f1)\n",
    "\n",
    "    #F1score[x].append(f1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Age': ['15 weeks',\n",
       "  '2 years',\n",
       "  '2 years',\n",
       "  '30 months',\n",
       "  '4 months',\n",
       "  '4 months',\n",
       "  'age 10 weeks',\n",
       "  'age 2 years',\n",
       "  'age 35 days',\n",
       "  'age 4 months',\n",
       "  'age 5 days',\n",
       "  'first months',\n",
       "  'first years',\n",
       "  'fourth child'],\n",
       " 'Sex': ['male'],\n",
       " 'Sign_symptom': ['cardiomegaly',\n",
       "  'dilatation',\n",
       "  'dilatation',\n",
       "  'febrile illness',\n",
       "  'febres',\n",
       "  'hypoxemia',\n",
       "  'hypertrophic',\n",
       "  'hypertrophy',\n",
       "  'hypertrophy',\n",
       "  'hypertrophy',\n",
       "  'hypotonia',\n",
       "  'infections',\n",
       "  'left ventricular hypertrophy',\n",
       "  'left ventricular hypertrophy',\n",
       "  'metabolic acidosis',\n",
       "  'murmur',\n",
       "  'pulmonary edema',\n",
       "  'pulmonary hypertension',\n",
       "  'pulmonary hypertension',\n",
       "  'shunt malfunctioning',\n",
       "  'subcortial and white matter cortical hemorrhage'],\n",
       " 'Lab_value': ['10',\n",
       "  '13 centimeters',\n",
       "  '277\\u2009μm',\n",
       "  '9',\n",
       "  'grew rapidly',\n",
       "  'normal',\n",
       "  'normal',\n",
       "  'normal',\n",
       "  'rapidly',\n",
       "  'subcortial and white matter'],\n",
       " 'Biological_structure': ['Brain',\n",
       "  'Chest',\n",
       "  'cardiac',\n",
       "  'chest',\n",
       "  'occipital region',\n",
       "  'ventricles',\n",
       "  'ventricular'],\n",
       " 'Diagnostic_procedure': ['ABR',\n",
       "  'CT',\n",
       "  'CT',\n",
       "  'Echo',\n",
       "  'Echo',\n",
       "  'Eye examination',\n",
       "  'Head CT',\n",
       "  'Hearing test',\n",
       "  'Metabolic work up',\n",
       "  'Pathology',\n",
       "  'Physical exam',\n",
       "  'X-ray',\n",
       "  'acylcarnitines',\n",
       "  'alanine',\n",
       "  'ammonia',\n",
       "  'biopsy',\n",
       "  'biopsy',\n",
       "  'cardiac Echo',\n",
       "  'cardiac Echo',\n",
       "  'cytochrome c oxidase stain',\n",
       "  'exome sequencing',\n",
       "  'head circumference',\n",
       "  'head circumference',\n",
       "  'head circumference',\n",
       "  'lactate',\n",
       "  'lactic acidosis',\n",
       "  'urine organic acids']}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Age': ['16-year-old'], 'Sex': ['female']}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output = {}\n",
    "output.update({'Age': ['16-year-old'], 'Sex': ['female']})\n",
    "output.update({})\n",
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A 16-year-old female suffered from an abdominal pain at right upper quadrant, lasting for more than 10 days.\\nShortly ahead of her medical consultation, 4 ascaris-like worms were vomited out, individually with a length of about 10\\u200acm.\\nShe had a previous onset of ascariasis, during which low-grade fever (38.4°C) occurred without apparent jaundice, diarrhea, or anemia.\\nPhysical examination revealed tenderness at right upper quadrant.\\nHer total leukocyte count was 11.2\\u200aG/L consisting of 5.2% eosinophils.\\nSerum and urine amylase were 386 and 928\\u200aU/L, respectively.\\nIn terms of liver functionality, the level of total bilirubin rose up to 23.2\\u200aμm/L, meanwhile the hepatic enzymes were similarly elevated (alanine aminotransferase 163\\u200aU/L; aspartate aminotransferase 96\\u200aU/L).\\nAbdominal ultrasound described the enlargement of the gallbladder, upper segment of common bile duct (1.5\\u200acm in diameter) and intrahepatic bile duct (1.3\\u200acm in diameter).\\nFurthermore, the intrahepatic bile duct was also discovered to be filled with echogenic and banded substances.\\nBased on such evidences, she was clinically diagnosed as biliary ascariasis with a concomitant manifestation of biliary duct infection.\\nThrough a laparoscopic exploration, the enlarged gallbladder and common bile duct were visually confirmed, as well as the massive storage of worms within the biliary tract.\\nSubsequently, a 1\\u200acm incision was made on the common bile duct in order to facilitate the extraction of the pathogenic worms, which were then placed into a specimen bag laparoscopically (Fig.1).\\nOnce the contained worms were virtually cleared out under laparoscope, chodedochoscopy was employed to seek for more hidden worms inside the biliary system (Fig.2).\\nAfter finishing the removal of ascarides, the biliary duct incision was well sutured, with a T-tube placed in situ (Fig.3).\\nEventually, a standard laparoscopic cholecystectomy was performed, followed by the removal of specimen bags and drainage tube placement (Fig.4).\\nThe total amount of worms inside the common bile duct summed up to more than 100 (Fig.5).\\nPostoperatively, she was treated with oral antihelminthic drugs for 3 days.\\nThe T-tube was rinsed with saline every day, and during the first 3 days, 6 more worms were additionally washed out.\\nThe patient experienced a smooth postoperative recovery and was finally discharged from hospital without any complications.\\n'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dataset['input'][15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Age': ['68-year-old'], 'Sex': ['man']}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Something went wrong with json.loads(pred)\n",
      "\n",
      "{\"Biological_structure\": [\"LAD\", \"LV\", \"LV\", \"Left ITA\", \"RA\", \"RA\", \"RITA\", \"RITA\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"LV\", \"L}\n",
      "Something went wrong with json.loads(pred)\n",
      "\n",
      "{\"Sign_symptom\": [\"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"ascariasis\", \"asc}\n",
      "Something went wrong with json.loads(pred)\n",
      "\n",
      "{\"Lab_value\": [\"+0.25 + 0.50 × 130°\", \"108 cm\", \"12 mm Hg\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"20/70\", \"}\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(\"..\")\n",
    "\n",
    "from scripts.generate import generate\n",
    "from collections import defaultdict\n",
    "\n",
    "F1score = defaultdict(list)\n",
    "\n",
    "compare = \"\"\n",
    "#entities = [\"Age\", \"Sex\"]\n",
    "#entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\"]\n",
    "entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\", \"Biological_structure\", \"Diagnostic_procedure\"]\n",
    "\n",
    "for instance in range(len(test_dataset['input'])):\n",
    "    text = test_dataset['input'][instance]\n",
    "    extract = generate.extract(model, tokenizer, entities, text, max_length=2500)\n",
    "\n",
    "    compare = compare + test_dataset[instance][\"input\"] + \"\\n\\n\"\n",
    "    \n",
    "    for x in json.loads(test_dataset['output'][instance]):\n",
    "\n",
    "        if x not in extract:\n",
    "            continue\n",
    "\n",
    "        real = list(set(json.loads(test_dataset['output'][instance])[x]))\n",
    "        gene = list(set(extract[x]))\n",
    "\n",
    "        match = list(set([r for r in real if r in gene]))\n",
    "\n",
    "        lenmatch = len(match)\n",
    "        lenreal = len(real)\n",
    "        lengene = len(gene)\n",
    "            \n",
    "        text = f\"{x} \\nReal: {json.loads(test_dataset['output'][instance])[x]} \\nGene: {extract[x]} \\nMatch: {match} \\n\\n\"\n",
    "        compare = compare + text\n",
    "\n",
    "        precision = lenmatch/lengene\n",
    "        recall = lenmatch/lenreal\n",
    "        f1 = 2*(precision*recall) / (precision+recall)\n",
    "        #print(precision, recall, f1)\n",
    "\n",
    "        F1score[x].append(f1) \n",
    "            \n",
    "            \n",
    "    compare = compare + \"\\n\\n\\n\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'Age': [0.6666666666666666,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              0.6666666666666666,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0],\n",
       "             'Biological_structure': [0.631578947368421,\n",
       "              0.4615384615384615,\n",
       "              0.9166666666666666,\n",
       "              0.5384615384615384,\n",
       "              0.7272727272727272,\n",
       "              0.4615384615384615,\n",
       "              0.8,\n",
       "              0.6000000000000001,\n",
       "              0.6666666666666666,\n",
       "              0.6666666666666667,\n",
       "              0.4,\n",
       "              0.4444444444444444,\n",
       "              0.5,\n",
       "              0.5384615384615384,\n",
       "              0.6,\n",
       "              0.6153846153846153,\n",
       "              0.5,\n",
       "              0.6666666666666667,\n",
       "              0.7586206896551724],\n",
       "             'Diagnostic_procedure': [0.5416666666666666,\n",
       "              0.4117647058823529,\n",
       "              0.7111111111111111,\n",
       "              0.7692307692307692,\n",
       "              0.6666666666666666,\n",
       "              0.18181818181818182,\n",
       "              0.3018867924528302,\n",
       "              0.7666666666666667,\n",
       "              0.6250000000000001,\n",
       "              0.6666666666666666,\n",
       "              0.7096774193548387,\n",
       "              0.26086956521739135,\n",
       "              0.7999999999999999,\n",
       "              0.6976744186046512,\n",
       "              0.5979381443298969,\n",
       "              0.6956521739130435,\n",
       "              0.823529411764706,\n",
       "              0.4761904761904762,\n",
       "              0.7333333333333334,\n",
       "              0.608695652173913],\n",
       "             'Lab_value': [0.39999999999999997,\n",
       "              0.6521739130434783,\n",
       "              0.7241379310344828,\n",
       "              0.8666666666666666,\n",
       "              0.4615384615384615,\n",
       "              0.6666666666666666,\n",
       "              0.34782608695652173,\n",
       "              0.7241379310344829,\n",
       "              0.5,\n",
       "              0.6666666666666666,\n",
       "              0.8,\n",
       "              0.6666666666666666,\n",
       "              0.631578947368421,\n",
       "              0.45454545454545453,\n",
       "              0.7032967032967034,\n",
       "              0.5925925925925927,\n",
       "              0.8,\n",
       "              0.4615384615384615,\n",
       "              0.56],\n",
       "             'Sex': [1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0,\n",
       "              1.0],\n",
       "             'Sign_symptom': [0.14285714285714288,\n",
       "              0.6923076923076924,\n",
       "              0.6190476190476191,\n",
       "              0.14285714285714288,\n",
       "              0.5714285714285713,\n",
       "              0.6666666666666666,\n",
       "              0.5333333333333333,\n",
       "              0.7499999999999999,\n",
       "              0.6666666666666666,\n",
       "              0.46511627906976744,\n",
       "              0.42857142857142855,\n",
       "              0.4444444444444444,\n",
       "              0.7586206896551724,\n",
       "              0.3225806451612903,\n",
       "              0.5230769230769231,\n",
       "              0.7142857142857143,\n",
       "              0.25,\n",
       "              0.6666666666666667,\n",
       "              0.48000000000000004]})"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "F1score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Age': 0.9649122807017543,\n",
       " 'Biological_structure': 0.6049456889890551,\n",
       " 'Diagnostic_procedure': 0.6023019411022081,\n",
       " 'Lab_value': 0.6147385868218805,\n",
       " 'Sex': 1.0,\n",
       " 'Sign_symptom': 0.5178172434787496}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def averageF1Dict(di):\n",
    "    return {k:sum(v)/len(v) for k,v in di.items()}\n",
    "\n",
    "averageF1Dict(F1score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"../results/comparePerEntity2.txt\", \"w\")\n",
    "f.write(compare)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scripts.prompts import prompts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "{\"Lab_value\": [\"1.3 cm in diameter\", \"1.5 cm in diameter\", \"10 cm\", \"11.2 G/L\", \"163 U/L\", \"23.2 μm/L\", \"38.4°C\", \"386\", \"928 U/L\", \"5.2% eosinophils\", \"96 U/L\", \"rose up\", \"similarly elevated\"]}\n"
     ]
    }
   ],
   "source": [
    "entities = [\"Age\", \"Sex\", \"Sign_symptom\", \"Lab_value\", \"Biological_structure\", \"Diagnostic_procedure\"]\n",
    "\n",
    "prompt = prompts.lab_value(text)\n",
    "ggenerated = generate.gene(model, tokenizer, prompt, max_length=2500)\n",
    "print(ggenerated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Lab_value': ['1.3\\u200acm in diameter',\n",
       "  '1.5\\u200acm in diameter',\n",
       "  '10\\u200acm',\n",
       "  '11.2\\u200aG/L',\n",
       "  '163\\u200aU/L',\n",
       "  '23.2\\u200aμm/L',\n",
       "  '38.4°C',\n",
       "  '386',\n",
       "  '928\\u200aU/L',\n",
       "  '5.2% eosinophils',\n",
       "  '96\\u200aU/L',\n",
       "  'rose up',\n",
       "  'similarly elevated']}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate.json_format(ggenerated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 16-year-old female suffered from an abdominal pain at right upper quadrant, lasting for more than 10 days.\n",
      "Shortly ahead of her medical consultation, 4 ascaris-like worms were vomited out, individually with a length of about 10 cm.\n",
      "She had a previous onset of ascariasis, during which low-grade fever (38.4°C) occurred without apparent jaundice, diarrhea, or anemia.\n",
      "Physical examination revealed tenderness at right upper quadrant.\n",
      "Her total leukocyte count was 11.2 G/L consisting of 5.2% eosinophils.\n",
      "Serum and urine amylase were 386 and 928 U/L, respectively.\n",
      "In terms of liver functionality, the level of total bilirubin rose up to 23.2 μm/L, meanwhile the hepatic enzymes were similarly elevated (alanine aminotransferase 163 U/L; aspartate aminotransferase 96 U/L).\n",
      "Abdominal ultrasound described the enlargement of the gallbladder, upper segment of common bile duct (1.5 cm in diameter) and intrahepatic bile duct (1.3 cm in diameter).\n",
      "Furthermore, the intrahepatic bile duct was also discovered to be filled with echogenic and banded substances.\n",
      "Based on such evidences, she was clinically diagnosed as biliary ascariasis with a concomitant manifestation of biliary duct infection.\n",
      "Through a laparoscopic exploration, the enlarged gallbladder and common bile duct were visually confirmed, as well as the massive storage of worms within the biliary tract.\n",
      "Subsequently, a 1 cm incision was made on the common bile duct in order to facilitate the extraction of the pathogenic worms, which were then placed into a specimen bag laparoscopically (Fig.1).\n",
      "Once the contained worms were virtually cleared out under laparoscope, chodedochoscopy was employed to seek for more hidden worms inside the biliary system (Fig.2).\n",
      "After finishing the removal of ascarides, the biliary duct incision was well sutured, with a T-tube placed in situ (Fig.3).\n",
      "Eventually, a standard laparoscopic cholecystectomy was performed, followed by the removal of specimen bags and drainage tube placement (Fig.4).\n",
      "The total amount of worms inside the common bile duct summed up to more than 100 (Fig.5).\n",
      "Postoperatively, she was treated with oral antihelminthic drugs for 3 days.\n",
      "The T-tube was rinsed with saline every day, and during the first 3 days, 6 more worms were additionally washed out.\n",
      "The patient experienced a smooth postoperative recovery and was finally discharged from hospital without any complications.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 68-year-old man was referred by his optometrist to HES with suspected LTG due to repeatedly irregular visual field test results, advanced optic disc cupping, normal intraocular pressures (IOPs) and a family history of glaucoma.\n",
      "The patient subjectively felt that vision in his ‘good’ left eye (LE), which normally had a visual acuity of 6/6 N5, started to deteriorate 6 months earlier; at the point of referral it was best corrected to 6/7.5 N6.\n",
      "His right eye (RE) was known to be amblyopic with a visual acuity of 6/18 N12.\n",
      "His medical history included considerable risk factors for systemic vasculopathy, such as hypertension, hypercholesterolaemia, 50 pack-years of smoking and type 2 diabetes with no diabetic retinopathy.\n",
      "Despite detailed questioning, he denied any new systemic symptoms apart from experiencing increased lethargy.\n",
      "Clinical examination at HES revealed advanced bilateral cupped optic discs with a cup-to-disc ratio right 0.9 (90%) and left 0.8 (80%; figure 1).\n",
      "Furthermore, the findings from the referring optometrist of equal and reactive pupils, normal eye movements and IOPs in mid-teens were confirmed.\n",
      "Computerised perimetry showed constricted visual fields in both eyes, albeit with considerably raised false-negative error rates in excess of 30% (figure 2A).\n",
      "Since this can be a result of poor concentration and attention span with testing compounded, especially with an amblyopic RE, the field test was repeated 4 weeks later with similar results (figure 2B).\n",
      "At this visit, the patient experienced further deterioration of his ‘good’ LE.\n",
      "Visual acuities measured 6/24 (loss of one Snellen line) N36 in his amblyopic RE, and 6/18 (loss of two Snellen lines) N36 in his LE, which now also showed depressed colour vision.\n",
      "An urgent scan of the patient's head and orbit was requested to rule out a compressive lesion and blood analysis was carried out to screen for other causes of optic neuropathy including vitamin B12, folate levels and inflammatory markers.\n",
      "MRI revealed a large mass centred at the pituitary fossa measuring approximately 4.5 cm×5.1 cm× 4.8 cm (figure 3A–C) extending into the suprasellar cistern and infrasellar sphenoid sinuses.\n",
      "The optic chiasm and both optic nerves were displaced superiorly and compressed on both sides.\n",
      "The mass showed heterogeneous T1 and T2 signals with multiple fluid levels most likely due to previous haemorrhage and low-signal components suggesting calcifications.\n",
      "The appearance was probably due to a craniopharyngioma or haemorrhagic pituitary adenoma.\n",
      "The patient was referred urgently to the neurosurgery unit and was found to have markedly raised prolactin levels of greater than 200 000 mIU/L, confirming the diagnosis of a giant prolactinoma.\n",
      "In cooperation with the local endocrinology service, he was started on cabergoline 500 µg and dexamethasone 8 mg.\n",
      "Within a week this led to a dramatic improvement of the patient's visual fields and acuities to 6/18 RE and to 6/9 LE.\n",
      "His prolactin levels showed a similar dramatic response with levels coming down to 336 mIU/L.\n",
      "Three months later his visual acuity in his RE and LE were 6/18 and 6/5, respectively, and visual field examination showed striking recovery with almost normal results (figure 2C).\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['68-year-old', 'mid-teens'] \n",
      "Gene: ['68-year-old'] \n",
      "Match: ['68-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['0.8', '0.9', '336\\u2005mIU/L', '50 pack-years', '6/18', '6/18', '6/18 N12', '6/5', '6/6 N5', '6/7.5 N6', '6/9', '80%', '90%', 'both', 'both', 'considerably raised', 'depressed', 'dramatic improvement', 'dramatic response with levels coming down', 'further deterioration', 'greater than 200\\u2005000\\u2005mIU/L', 'in excess of 30%', 'increased', 'markedly raised', 'multiple', 'normal', 'normal'] \n",
      "Gene: ['0.8', '0.9', '12', '18', '18', '18', '200\\u2005000\\u2005mIU/L', '336\\u2005mIU/L', '30%', '6/18', '6/18', '6/18', '6/5', '6/6', '6/6 N5', '6/7.5 N6', '6/9', '6/9', '6/18 N12', '6/18 RE', '6/24', '6/9 LE', '6/9 RE', 'almost normal', 'depressed', 'equal', 'false-negative', 'greater than', 'loss of one Snellen line', 'loss of two Snellen lines', 'normal', 'normal', 'normal', 'raised', 'raised', 'reactive', 'similar', 'similar', 'similar', 'superiorly', 'urgent'] \n",
      "Match: ['0.8', '0.9', '336\\u2005mIU/L', '6/18', '6/18', '6/18 N12', '6/5', '6/6 N5', '6/7.5 N6', '6/9', 'depressed', 'normal', 'normal'] \n",
      "\n",
      "Sex \n",
      "Real: ['man'] \n",
      "Gene: ['man'] \n",
      "Match: ['man'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['attention span', 'concentration', 'lethargy', 'new systemic symptoms', 'risk factors', 'vision'] \n",
      "Gene: ['cupped', 'cupped', 'displaced', 'false-negative error rates', 'fluid levels', 'lethargy', 'mass', 'mass', 'optic disc cupping', 'symptoms'] \n",
      "Match: ['lethargy'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 35-year-old woman was admitted with a 5-day history of increasing drowsiness, fatigue, personality changes, generalised weakness and deteriorating mobility.\n",
      "Her medical history included a 10-year history of schizo-affective disorder, lithium-induced diabetes insipidus and hypothyroidism.\n",
      "Her medications included sodium valproate 1 g twice daily, carbamazepine, quetiapine and lithium.\n",
      "On admission, the patient was drowsy, difficult to rouse (sleeping all night and in the afternoon).\n",
      "Her Glasgow Coma Scale (GCS) was 14/15, her abbreviated mini-mental test score was 6/10 and she was noted to be slow in her responses.\n",
      "Neurological examination showed reduced power in her lower limbs (3–4/5) in a pyramidal distribution, and brisk reflexes with intact sensation in all modalities.\n",
      "Cardiovascular, respiratory and gastrointestinal examinations were unremarkable and there were no demonstrable features of hepatic disease.\n",
      "Routine biochemistry was unremarkable.\n",
      "The patient's alanine transaminase was 7 u/L (1–15), alkaline phosphatase 61 u/L (30–130), γ-glutamyl transferase 17 u/L (<45) and bilirubin was 6 μmol/L (<21).\n",
      "She had normal serum B12 and folate levels, and normal thyroid function tests.\n",
      "Serum lithium levels were within the therapeutic range.\n",
      "Her serum ammonia was 47 μmol/L (11.2–35.4) and valproate concentrations were elevated 140 mg/L (50–100).\n",
      "Brain imaging including CT and MRI did not show any haemorrhage or intracranial mass.\n",
      "Lumbar puncture demonstrated clear cerebrospinal fluid (white cell count <1×106/L, protein 0.25 g/L (0.2–0.4)).\n",
      "EEG demonstrated mild variable slowing and irregularity of background activity (likely due to medication).\n",
      "The differential diagnosis included encephalitis, Guillain-Barré syndrome and acute disseminated encephalomyelitis (ADEM), which were excluded through our investigations.\n",
      "In view of the raised valproate and ammonia levels, a diagnosis of valproate hyperammonaemic encephalopathy (VHE) was made at this point\n",
      "The dose of sodium valproate was reduced by half rather than being completely withdrawn.\n",
      "This was due to the complex and challenging nature of managing the patient's schizo-affective disorder.\n",
      "Within 3 days, there was evident clinical improvement in her mental status.\n",
      "The patient was able to sit and communicate with the team, and her abbreviated mental test score improved to 10/10.The serum ammonia normalised and valproate levels decreased to the therapeutic range.\n",
      "After 4 days she was discharged home.\n",
      "Results of investigations of her urea cycle were awaited.\n",
      "Three weeks postdischarge, the patient was clinically stable in terms of mental functioning, having had no further episodes of confusion or drowsiness.\n",
      "However, she continued to have fatigue and walking difficulties (due to residual leg weakness).\n",
      "Results of her serum amino acids revealed that our patient had carnitine deficiency evident by low acyl-carnitine, low free serum carnitine (14.2 μmol/L, normal range: 23–52) and low total serum carnitine (19.9 μmol/L, normal range: 27–63).\n",
      "She was started on oral carnitine 1 g twice daily.\n",
      "Within 2 weeks, her symptoms resolved completely with normalisation of serum carnitine levels (free carnitine 25.2 μmol/L and total carnitine of 41.3 μmol/L).\n",
      "The sodium valproate was initially stopped completely, but due to a serious relapse of the patient's schizo-affective disorder, it was restarted and she is now well on valproate 1 g once daily and long-term carnitine supplements.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['35-year-old'] \n",
      "Gene: ['35-year-old'] \n",
      "Match: ['35-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['0.25\\u2005g/L', '10/10', '14.2\\u2005μmol/L', '140\\u2005mg/L', '17\\u2005u/L', '19.9\\u2005μmol/L', '25.2\\u2005μmol/L', '41.3\\u2005μmol/L', '47\\u2005μmol/L', '61\\u2005u/L', '6\\u2005μmol/L', '7\\u2005u/L', 'decreased', 'improved', 'improvement', 'low', 'low', 'low', 'normalisation', 'normalised', 'raised', 'stable', 'therapeutic range', 'unremarkable'] \n",
      "Gene: ['10/10', '14/15', '140\\u2005mg/L', '14.2\\u2005μmol/L', '17\\u2005u/L', '19.9\\u2005μmol/L', '25.2\\u2005μmol/L', '3–4/5', '41.3\\u2005μmol/L', '47\\u2005μmol/L', '6/10', '61\\u2005u/L', '6\\u2005μmol/L', '7\\u2005u/L', 'elevated', 'improved', 'intact', 'low', 'low', 'normal', 'normal', 'normal', 'normalised', 'reduced', 'slow', 'slowing', 'unremarkable', 'unremarkable'] \n",
      "Match: ['10/10', '14.2\\u2005μmol/L', '140\\u2005mg/L', '17\\u2005u/L', '19.9\\u2005μmol/L', '25.2\\u2005μmol/L', '41.3\\u2005μmol/L', '47\\u2005μmol/L', '61\\u2005u/L', '6\\u2005μmol/L', '7\\u2005u/L', 'improved', 'low', 'low', 'low', 'normalised', 'unremarkable'] \n",
      "\n",
      "Sex \n",
      "Real: ['woman'] \n",
      "Gene: ['woman'] \n",
      "Match: ['woman'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['confusion', 'difficult to rouse', 'drowsiness', 'drowsiness', 'drowsy', 'fatigue', 'fatigue', 'haemorrhage', 'irregularity of background activity', 'mass', 'personality changes', 'slowing', 'symptoms', 'walking difficulties', 'weakness', 'weakness', 'well'] \n",
      "Gene: ['acute disseminated encephalomyelitis', 'confusion', 'drowsiness', 'drowsiness', 'drowsy', 'fatigue', 'fatigue', 'haemorrhage', 'irregularity of background activity', 'mass', 'personality changes', 'slow in her responses', 'sleeping', 'weakness', 'weakness'] \n",
      "Match: ['confusion', 'drowsiness', 'drowsiness', 'drowsy', 'fatigue', 'fatigue', 'haemorrhage', 'irregularity of background activity', 'mass', 'personality changes', 'weakness', 'weakness'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 33-year-old Moroccan man with a 5-month history of recurring sinusitis presented with temperature, myalgia, cough, and hemoptysis.\n",
      "Thoracic physical examination revealed reduced vesicular murmur, and chest X-ray showed bilateral diffuse alveolar infiltrates.\n",
      "Laboratory analyses showed elevation of acute-phase proteins (CRP 26 mg/dL) and creatinine (1.2 mg/dL) with nonnephrotic proteinuria and microscopic glomerular hematuria.\n",
      "An antibiotic therapy for community acquired and atypical pneumoniae with levofloxacin 500 mg daily iv was established, but on the 3rd day from admission the patient developed respiratory failure unresponsive to noninvasive ventilation and had to be intubated.\n",
      "White lung with consolidations and ground glass areas at computed tomography scan (Fig.1), diffuse airways bleeding at fiberoptic bronchoscopy, mild normocytic anemia (9 g/dL), and ANCA-PR3 positivity (18.9 U/mL) were consistent with the diagnosis of GPA.\n",
      "Despite maximal ventilatory support, gas exchange did not improve (pH 7.33, PaO2 71 mm Hg, PaCO2 51 mm Hg, HCO3– 25 mmol/L, SaO2 94%) requiring extracorporeal membrane oxygenation (ECMO).\n",
      "Treatment with high-dose methylprednisolone (1 g daily for 3 days and 1 mg/kg daily thereafter), cyclophosphamide (1.2 g/pulse every 2 weeks for the 1st 3 pulses, followed by infusions every 3 weeks for the next 2 pulses), and plasma exchange (PE), according to European vasculitis study group recommendations,[4] was immediately started with respiratory improvement that allowed ECMO and orotracheal tube withdrawal and subsequent discharge from intensive care unit.\n",
      "After 9 days from the beginning of treatment ANCA-PR3 levels normalized (3 U/mL).\n",
      "Two months later, after the 5th bolus of cyclophosphamide and with prednisone 0.3 mg/kg/day, the patient complained arthromyalgia and testicular pain with edema, cough, and shortness of breath: laboratory data showed worsening of renal function, mild anemia, and hypoxemia (PaO2 61 mm Hg).\n",
      "Epididymitis was confirmed by ultrasound.\n",
      "Because of rapidly progressive renal failure (estimated glomerular filtration rate 35 mL/min) renal biopsy was performed and showed prominent intracapillary and extracapillary proliferation, with the formation of some fibrocellular crescents and focal capillary necrosis (Fig.2).\n",
      "Lung computed tomography-scan and fiberoptic bronchoscopy also confirmed alveolar hemorrhage relapse.\n",
      "Treatment with high-dose methylprednisolone (1 g daily for 3 consecutive days) and rituximab (375 mg/m2 weekly for 4 weeks) was started with rapid resolution of urologic and pulmonary involvement but delayed and partial improvement of renal function.\n",
      "At a 1-year follow-up, the patient is alive and under treatment with azathioprine (150 mg daily) and prednisone (15 mg daily).\n",
      "Renal function has markedly improved (estimated glomerular filtration rate 94 mL/min) same as gas-exchange (pH 7.42, PaO2 86 mm Hg, PaCO2 36 mm Hg, HCO3– 23 mmol/L, SaO2 95%).\n",
      "The patient signed institutional informed consent form at the time of admission to hospitalization.\n",
      "An approval by ethics committee was not necessary because of the routine health care.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['33-year-old'] \n",
      "Gene: ['33-year-old'] \n",
      "Match: ['33-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['1.2\\u200amg/dL', '18.9\\u200aU/mL', '23\\u200ammol/L', '25\\u200ammol/L', '26\\u200amg/dL', '35\\u200amL/min', '36\\u200amm\\u200aHg', '3\\u200aU/mL', '51\\u200amm\\u200aHg', '61\\u200amm\\u200aHg', '7.33', '7.42', '71\\u200amm\\u200aHg', '86\\u200amm\\u200aHg', '94%', '94\\u200amL/min', '95%', '9\\u200ag/dL', 'did not improve', 'elevation', 'improvement', 'markedly improved', 'normalized', 'partial improvement', 'positivity', 'reduced', 'worsening'] \n",
      "Gene: ['1.2\\u200amg/kg', '15\\u200amg', '18.9\\u200aU/mL', '23\\u200ammol/L', '25\\u200ammol/L', '35\\u200amL/min', '375\\u200amg/m2', '51\\u200amm\\u200aHg', '61\\u200amm\\u200aHg', '7.33', '7.42', '71\\u200amm\\u200aHg', '86\\u200amm\\u200aHg', '9', '94%', '94\\u200amL/min', '95%', '1.2\\u200ag/pulse', '150\\u200amg', '18.9\\u200aU/mL', '26\\u200amg/dL', '375\\u200amg/m2', '500\\u200amg', '7.33', '71\\u200amm\\u200aHg', '86\\u200amm\\u200aHg', '94\\u200amL/min', '95%', 'elevation', 'improved', 'improvement', 'improvement', 'markedly improved', 'negative', 'nonnephrotic', 'normalized', 'positivity', 'reduced', 'worsening'] \n",
      "Match: ['18.9\\u200aU/mL', '23\\u200ammol/L', '25\\u200ammol/L', '26\\u200amg/dL', '35\\u200amL/min', '51\\u200amm\\u200aHg', '61\\u200amm\\u200aHg', '7.33', '7.42', '71\\u200amm\\u200aHg', '86\\u200amm\\u200aHg', '94%', '94\\u200amL/min', '95%', 'elevation', 'improvement', 'markedly improved', 'normalized', 'positivity', 'reduced', 'worsening'] \n",
      "\n",
      "Sex \n",
      "Real: ['man'] \n",
      "Gene: ['man'] \n",
      "Match: ['man'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['alive', 'anemia', 'anemia', 'arthromyalgia', 'bleeding', 'consolidations', 'cough', 'cough', 'edema', 'fibrocellular crescents', 'ground glass areas', 'hematuria', 'hemoptysis', 'hemorrhage', 'hypoxemia', 'infiltrates', 'involvement', 'myalgia', 'necrosis', 'pain', 'proliferation', 'proteinuria', 'shortness of breath', 'temperature'] \n",
      "Gene: ['ANCA-PR3 positivity', 'airways bleeding', 'alveolar infiltrates', 'anemia', 'anemia', 'arthromyalgia', 'cough', 'cough', 'edema', 'fibrocellular crescents', 'focal capillary necrosis', 'hematuria', 'hemoptysis', 'hypoxemia', 'myalgia', 'pain', 'proliferation', 'reduced vesicular murmur', 'respiratory failure', 'shortness of breath', 'sinusitis', 'temperature'] \n",
      "Match: ['anemia', 'anemia', 'arthromyalgia', 'cough', 'cough', 'edema', 'fibrocellular crescents', 'hematuria', 'hemoptysis', 'hypoxemia', 'myalgia', 'pain', 'proliferation', 'shortness of breath', 'temperature'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 69-year-old male diabetic patient was admitted with acute left ventricular (LV) failure and was treated with intensive decongestive therapy.\n",
      "He had severe peripheral edema and bilateral basal crepitations.\n",
      "Initial workup revealed normal levels of troponine (<0.01 ng/ml) and CPK-MB (4.6 ng/ml), thereby ruling out acute coronary syndrome.\n",
      "The level of BNP was 1230 pg/ml confirming the diagnosis of heart failure.\n",
      "Echocardiographic assessment revealed LVEF of 10% with dilated cardiac chambers, and estimated pulmonary artery systolic pressure (PASP) was 70 mm of Hg with features of raised LV diastolic pressure.\n",
      "Angiography was performed after 3 days of decongestive therapy which revealed critical triple vessel disease with left main coronary artery disease (Fig.1).\n",
      "He was referred for CABG.\n",
      "The patient was put on frusemide infusion with oral ramipril and aldactone.\n",
      "He was taken up for elective OPCAB 1 week after angiography.\n",
      "Renal function was monitored regularly to adjust diuretic dose.\n",
      "There was no feature of end organ dysfunction.\n",
      "STS risk of mortality was 2.4%.\n",
      "Repeat echocardiographic assessment revealed improvement of LVEF to 15% and decrease in PASP to 30 mm Hg and decrease in LV filling pressure.\n",
      "OPCAB was performed through median sternotomy.\n",
      "Left ITA was used for grafting left anterior descending artery (LAD).\n",
      "After revascularizing LAD, right ITA (RITA) and RA composite, graft was prepared.\n",
      "The distal end of RITA was used for revascularizing the diagonal artery.\n",
      "There was significant improvement in systolic blood pressure after this distal anastomosis.\n",
      "After these two grafts, heart was lifted for lateral and posterior wall vessels.\n",
      "We make a longer skin incision, open the sternum widely, and release pericardium on the right side to displace the heart gradually.\n",
      "This was supplemented with headlow and lateral tilt position.\n",
      "Liberal doses of inotropic support were used.\n",
      "RA was used for sequential grafting of obtuse marginal, posterior LV branch, and posterior descending artery.\n",
      "All these five distal anastamoses were performed uneventfully.\n",
      "The patient was electively ventilated for 2 days.\n",
      "Echocardiography on 1st POD revealed improvement in LVEF to 20%.\n",
      "He was started on low dose beta blockers after inotropic supports were weaned off.\n",
      "He had slow and uneventful recovery.\n",
      "Echocardiography on 4th POD revealed improvement in LVEF to 30% (Table 1).\n",
      "He had persistent pleural drainage requiring prolonged intensive care unit stay.\n",
      "After pleural drainage subsided and drains were removed, he was discharged on diuretics, low dose beta blocker, and ramipril.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['69-year-old'] \n",
      "Gene: ['69-year-old'] \n",
      "Match: ['69-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['10%', '1230 pg/ml', '15%', '2.4%', '20%', '30 mm Hg', '30%', '4.6 ng/ml', '70 mm of Hg', '<0.01 ng/ml', 'decrease', 'decrease', 'features of raised LV diastolic pressure', 'five', 'improvement', 'improvement', 'improvement', 'improvement', 'normal', 'slow', 'uneventful'] \n",
      "Gene: ['10%', '1230 pg/ml', '15%', '2.4%', '30 mm Hg', '30%', '4.6 ng/ml', '70 mm of Hg', 'decrease', 'decrease', 'improvement', 'improvement', 'improvement', 'normal', 'slow', 'uneventful', 'uneventful'] \n",
      "Match: ['10%', '1230 pg/ml', '15%', '2.4%', '30 mm Hg', '30%', '4.6 ng/ml', '70 mm of Hg', 'decrease', 'decrease', 'improvement', 'improvement', 'improvement', 'improvement', 'normal', 'slow', 'uneventful'] \n",
      "\n",
      "Sex \n",
      "Real: ['male'] \n",
      "Gene: ['male'] \n",
      "Match: ['male'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['crepitations', 'dilated cardiac chambers', 'edema', 'pleural drainage', 'pleural drainage'] \n",
      "Gene: ['acute', 'acute', 'basal crepitations', 'dilated', 'drainage', 'drainage', 'edema', 'failure', 'failure', 'recovery', 'slow', 'systolic blood pressure', 'triple vessel disease'] \n",
      "Match: ['edema'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "This 73-year-old female, with a history of coronary artery disease and situs inversus, had recently noticed left arm exercise weakness.\n",
      "The magnetic resonance angiogram and computer tomography of the upper limb vessels showed critical stenosis of the left-side SCA with calcification near the orifice (Fig.1).\n",
      "Percutaneous revascularization was done via right femoral artery approach, and severe left SCA stenosis was confirmed (Fig.2A) (Supplementary Video 1, only online).\n",
      "Direct stenting in the left SCA was carried out with a balloon-expansible Express LD 10×25 mm stent (Boston Scientific Corporation, Natick, MA, USA) (Fig.2B) up to 8 atmospheres for 13 seconds.\n",
      "However, severe shortness of breath and hypotension developed after stenting.\n",
      "The blood pressure was down to 88/56, and the heart rate was up to 90.\n",
      "The oxygen saturation was down to 74%.\n",
      "Emergent intubation, fluid infusion, inotropic agent with norepinephrine were given.\n",
      "The immediate angiography showed vascular perforation in the stented segment of left SCA (Fig.2C) (Supplementary Video 2, only online).\n",
      "Emergency balloon inflation within the stent was done with an 8×40 mm balloon.\n",
      "The cardiovascular surgeon was also consulted.\n",
      "However, due to the high surgical risk of bypass, endovascular treatment was suggested.\n",
      "We decided to do a retrograde approach through the left brachial artery by surgical cut-down.\n",
      "Since there was no suitable graft stent in our cath lab, we modified and cut the iliac extension of self-expanding Endurant II graft stent 10×82 mm (Medtronic, Inc., Minneapolis, MN, USA), which was originally designed to treat abdominal aortic aneurysms (AAA), to a suitable length (around 30 mm) (Fig.3).\n",
      "After retrograde wiring, an operator-modified Endurant II graft stent was deployed in the stent segment (Fig.2D) and the perforation was sealed successfully.\n",
      "After stenting of the graft stent, however, slow flow of left common carotid artery (LCCA) was noted, and angiography confirmed the near total occlusion of the LCCA, occluded by the SCA graft stent (Fig.2E) (Supplementary Video 3, only online).\n",
      "A firm wire (Conquest pro, Asahi Inc, Seto, Japan) was successfully advanced outside the graft stent and into the LCCA.\n",
      "Despite sequent balloon dilations at the bifurcation site, the flow to the LCCA was still poor.\n",
      "Intravascular ultrasound also confirmed the severe compromise of the LCCA ostium which was caused by the graft stent (Fig.2F).\n",
      "Therefore, we put a balloon-expansible carotid stent, Express 7×37 mm (Boston Scientific Corporation), from the LCCA to the left brachiocephalic artery (Fig.4A).\n",
      "TIMI III flow of the LCCA was restored after stenting (Fig.4B-E) (Supplementary Video 4, only online).\n",
      "Intravascular ultrasound also confirmed proper expansion of the LCCA stent (Fig.4F).\n",
      "The follow-up chest radiography revealed left-side hemothorax.\n",
      "After thoracocentesis with drainage and proper care, the patient was discharged.\n",
      "Till now, the patient has been free from symptoms for six months.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['73-year-old'] \n",
      "Gene: ['73-year-old'] \n",
      "Match: ['73-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['74%', '8 atmospheres for 13 seconds', '88/56', '90', 'poor', 'proper', 'restored', 'slow'] \n",
      "Gene: ['74%', '88/56', '90', 'down', 'down', 'up', 'up'] \n",
      "Match: ['74%', '88/56', '90'] \n",
      "\n",
      "Sex \n",
      "Real: ['female'] \n",
      "Gene: ['female'] \n",
      "Match: ['female'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['calcification', 'compromise', 'expansion', 'hemothorax', 'hypotension', 'occlusion', 'shortness of breath', 'stenosis', 'stenosis', 'symptoms', 'vascular perforation', 'weakness'] \n",
      "Gene: ['calcification', 'critical stenosis', 'flow', 'flow', 'hypotension', 'perforation', 'perforation', 'shortness of breath', 'slow flow', 'symptoms', 'stenosis', 'stenosis', 'weakness'] \n",
      "Match: ['calcification', 'hypotension', 'shortness of breath', 'stenosis', 'stenosis', 'symptoms', 'weakness'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 72-year-old woman underwent RF ablation for the treatment of a metachronous solitary liver metastasis from rectal adenocarcinoma, which had been treated 18 mo before with preoperative chemoradiotherapy and anterior resection.\n",
      "The lesion was located in segment VIII.\n",
      "The procedure went well and the patient was initially haemodynamically stable.\n",
      "One hour later her condition suddenly worsened.\n",
      "She was clinically dizzy and had fatigue and pallor.\n",
      "We detected hypotension and tachycardia.\n",
      "Emergency blood tests showed a haemoglobin level of < 7 g/dL and markedly elevated transaminase levels.\n",
      "A computed tomography examination revealed an accumulation of intraperitoneal fluid in the pelvis and two liver haematomas.\n",
      "The images showed two liver lacerations, one following the path of the needle (Figure ​(Figure1A1A-​-C)C) and another, unexplained laceration leading away from the first (Figure ​(Figure2).2).\n",
      "The patient had no serious coughing or hiccupping after the RF treatment, or any other complications that might have caused increased abdominal pressure and tumour rupture.\n",
      "Subsequently, the patient received a blood transfusion and close monitoring.\n",
      "She was haemodynamically stable after the second blood transfusion and completely recovered 24 h later.\n",
      "The patient remained in bed for 1 wk.\n",
      "No surgical intervention was required.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['72-year-old'] \n",
      "Gene: ['72-year-old'] \n",
      "Match: ['72-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['< 7 g/dL', 'elevated', 'second', 'stable', 'stable', 'two', 'two', 'well'] \n",
      "Gene: ['< 7 g/dL', 'elevated', 'stable', 'stable'] \n",
      "Match: ['< 7 g/dL', 'elevated', 'stable', 'stable'] \n",
      "\n",
      "Sex \n",
      "Real: ['woman'] \n",
      "Gene: ['woman'] \n",
      "Match: ['woman'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['completely recovered', 'coughing', 'dizzy', 'fatigue', 'hiccupping', 'hypotension', 'lacerations', 'pallor', 'tachycardia'] \n",
      "Gene: ['accumulation of intraperitoneal fluid', 'complications', 'coughing', 'dizzy', 'hiccupping', 'hypotension', 'laceration', 'lacerations', 'pallor', 'pressure', 'tachycardia', 'tumour'] \n",
      "Match: ['coughing', 'dizzy', 'hiccupping', 'hypotension', 'lacerations', 'pallor', 'tachycardia'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 54-year-old woman with stage IV NSCLC was treated with carboplatin and paclitaxel without disease response.\n",
      "Molecular analysis of tumor tissue was unavailable at that time.\n",
      "However, her demographic profile (Asian, minimal smoking history, non-small cell histology) predicted her disease would harbor EGFR TKI sensitive cells [2].\n",
      "Therefore, she then initiated standard daily dosing of erlotinib (150 mg) and her disease responded.\n",
      "Twenty-eight months later, she acquired resistance to erlotinib with progression of disease systemically.\n",
      "Following further progression through an experimental angiogenesis inhibitor, she initiated pemetrexed and resumed standard dose erlotinib.\n",
      "After initial response, 11 months later, her disease again progressed.\n",
      "DNA was extracted from biopsy of a progressing lung lesion and examined using established techniques for analysis of EGFR mutations [5].\n",
      "Direct sequencing of exons 18–21 encoding the kinase domain of EGFR revealed the L858R mutation associated with EGFR TKI sensitivity (Fig.1) [2].\n",
      "It also demonstrated the T790M mutation associated with acquired EGFR TKI resistance (Fig.1) [2].\n",
      "She also developed headaches and there was a high clinical suspicion of CNS metastases despite negative imaging (not shown).\n",
      "She refused a lumbar puncture.\n",
      "She initiated empiric temozolomide plus standard dose erlotinib (150 mg daily) for presumed CNS disease, but after one cycle her headaches worsened, and she developed nausea and vomiting concerning for CNS metastases with associated raised intracranial pressure.\n",
      "Magnetic resonance imaging (MRI) of the brain now demonstrated LM (Fig.2) confirmed by CSF cytology (not shown).\n",
      "By direct sequencing, DNA from CSF cells harbored L858R predicting EGFR TKI sensitivity (Fig.3, left panel) but not the T790M resistance mutation (data not shown).\n",
      "Because the result for T790M was negative in this sample, we performed a more sensitive fluorescence detection PCR-based assay that takes advantage of a PCR restriction fragment length polymorphism generated by the specific missense mutation (Fig.3, right panel, arrow, positive control) [6].\n",
      "That result was also negative, as only the wild type peak was detected (Fig.3, right panel, bottom).\n",
      "Therefore, we hypothesized that the LM remained sensitive to an EGFR TKI if sufficiently high concentrations of drug could be achieved in the CSF.\n",
      "The erlotinib concentration required to inhibit growth of cell lines harboring L858R by 50% (IC50) is 100 nM (nM) [2].\n",
      "Standard dose erlotinib (150 mg daily) achieves 3000 nM in plasma [7], but CSF concentrations of EGFR TKIs are as low as 1% plasma levels below the IC50 [3, 8].\n",
      "Increasing the daily dose of gefitinib to enhance CSF penetration has been an effective strategy [3], but gefitinib is no longer available in the United States following failure in phase III NSCLC trials.\n",
      "An analogous increase of the daily erlotinib dose above 150–200 mg daily induces unacceptable toxicity.\n",
      "However, weekly high-dose erlotinib up to 2000 mg is tolerable [4].\n",
      "Pharmacokinetic analysis of CSF from another patient with NSCLC LM (not shown) treated with 1500 mg erlotinib weekly demonstrated a peak plasma concentration of 11,300 nM with a concurrent CSF concentration of 130 nM.\n",
      "Therefore, such high dose weekly administration of erlotinib achieved a CSF concentration exceeding the IC50.\n",
      "Therefore, to increase CSF penetrance over standard daily erlotinib dosing in this patient, we initiated high-dose weekly erlotinib at 1000 mg then 1200 mg; persistent nausea precluded higher doses.\n",
      "Pharmacokinetic analysis was not undertaken in this patient.\n",
      "After 1 month there was a partial radiographic response of LM on brain MRI (Fig.2b) and after 2 months in the cauda equina (not shown).\n",
      "However, hydrocephalus and persistent symptoms referable to increased intracranial pressure led to a VP shunt and whole-brain radiation therapy, after which she resumed treatment with 1500 mg weekly erlotinib.\n",
      "One month later, progressive intra-thoracic disease led to initiation of cetixumab and erlotinib was continued but changed to low dose (100 mg) daily.\n",
      "She survived 14 months following the diagnosis of CNS disease.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['54-year-old'] \n",
      "Gene: ['54-year-old'] \n",
      "Match: ['54-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['100 mg', '1000 mg', '11', '300 nM', '1200 mg', '130 nM', '150 mg', 'increased', 'more sensitive', 'negative', 'negative', 'negative', 'raised', 'worsened'] \n",
      "Gene: ['1%', '11', '3000 nM', '130 nM', '3000 nM', 'L858R', 'L858R', 'T790M', 'T790M', 'acquired', 'negative', 'negative', 'negative', 'positive', 'unavailable', 'worsened'] \n",
      "Match: ['11', '130 nM', 'negative', 'negative', 'negative', 'worsened'] \n",
      "\n",
      "Sex \n",
      "Real: ['woman'] \n",
      "Gene: ['woman'] \n",
      "Match: ['woman'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['headaches', 'headaches', 'nausea', 'nausea', 'survived', 'symptoms', 'vomiting'] \n",
      "Gene: ['LM', 'LM', 'acquired resistance', 'headaches', 'headaches', 'high clinical suspicion', 'intra-thoracic disease', 'nausea', 'nausea', 'progression', 'progression', 'symptoms', 'toxicity', 'vomiting'] \n",
      "Match: ['headaches', 'headaches', 'nausea', 'nausea', 'symptoms', 'vomiting'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 58-year-old white American male presented to the emergency department (ED) with acute onset of confusion, disorientation, inability to walk steadily, and dehydration with associated generalized weakness, polyuria, and polydipsia over the previous week.\n",
      "He had been diagnosed with CML six years earlier, with no hematological response to multiple chemotherapy regimens; he was at that time on treatment with allopurinol and awaiting allogenic bone marrow transplant.\n",
      "Vital signs on admission to the ED included heart rate of 104 bpm, blood pressure of 122/68 mm Hg and temperature of 99.7°F (37.6°C).\n",
      "There were no remarkable findings on physical examination except for altered mental status and dehydration.\n",
      "Laboratory evaluations were: hemoglobin 13.3 g/dL, white blood cell count 18.3×109/L (neutrophils 79.6%, lymphocytes 7.2%, and monocytes 8.3%), and platelet count 910×109/L.\n",
      "Serum levels were: calcium 18.6 mg/dL, phosphate 4.6 mg/dL, sodium 135 mEq/L, potassium 2.7 mg/dL, albumin 4.0 g/dL, creatinine 2.2 mg/dL, total bilirubin 0.9 mg/dL, alkaline phosphatase 125 IU/L, aspartate aminotransferase 41 IU/L, and alanine aminotransferase 71 IU/L.\n",
      "SPEP was negative for an M spike.\n",
      "Hormones and vitamins levels were: PTH 8.5 pg/L (15–65 pg/L), PHTrP 1.4 pg/L (<2 pg/L), 25-OH vitamin D 30.4 ng/mL (20–50 ng/mL) and 1,25 OH vitamin D3 33.7 ng/mL (18–64) ng/mL.\n",
      "Chest CT scan showed diffuse lytic lesions and bone destruction throughout the visualized thoracic skeleton, concerning for diffuse bone marrow involvement.\n",
      "The clinical history and biochemical findings led to a diagnosis of CML-associated hypercalcemia in the context of a blast phase.\n",
      "Treatment with aggressive hydration with 0.9% saline and calcitonin 400 units subcutaneous was administered.\n",
      "Additionally, zoledronic acid 3.3 mg intravenous, adjusted for renal insufficiency, was given.\n",
      "After five days of treatment, normalization of symptoms and serum calcium levels were achieved.\n",
      "After discharge from the hospital, the patient continued with palliative treatment with ponatinib and radiotherapy.\n",
      "Acceptable serum calcium levels were maintained with zoledronic acid 4 mg every eight weeks.\n",
      "Despite these therapies, his cancer progressed and he passed away approximately eight months after the initial admission for hypercalcemia.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['58-year-old'] \n",
      "Gene: ['58-year-old'] \n",
      "Match: ['58-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['0.9 mg/dL', '0.9%', '1.4 pg/L', '104 bpm', '122/68 mm Hg', '125 IU/L', '13.3 g/dL', '135 mEq/L', '18.3×109/L', '18.6 mg/dL', '2.2 mg/dL', '2.7 mg/dL', '30.4 ng/mL', '33.7 ng/mL', '37.6°C', '4.0 g/dL', '4.6 mg/dL', '41 IU/L', '71 IU/L', '8.5 pg/L', '910×109/L', '99.7°F', 'lymphocytes 7.2%', 'maintained', 'monocytes 8.3%', 'negative for an M spike', 'neutrophils 79.6%', 'no remarkable findings', 'normalization'] \n",
      "Gene: ['0.9 mg/dL', '1.4 pg/L', '104 bpm', '122/68 mm Hg', '125 IU/L', '13.3 g/dL', '135 mEq/L', '18.6 mg/dL', '18.3×109/L', '2.2 mg/dL', '2.7 mg/dL', '30.4 ng/mL', '33.7 ng/mL', '4.0 g/dL', '4.6 mg/dL', '41 IU/L', '7.2%', '71 IU/L', '79.6%', '8.3%', '8.5 pg/L', '910×109/L', '99.7°F', 'acceptable', 'dehydration', 'dehydration', 'diffuse', 'diffuse', 'negative', 'normalization', 'serum', 'serum', 'serum'] \n",
      "Match: ['0.9 mg/dL', '1.4 pg/L', '104 bpm', '122/68 mm Hg', '125 IU/L', '13.3 g/dL', '135 mEq/L', '18.3×109/L', '18.6 mg/dL', '2.2 mg/dL', '2.7 mg/dL', '30.4 ng/mL', '33.7 ng/mL', '4.0 g/dL', '4.6 mg/dL', '41 IU/L', '71 IU/L', '8.5 pg/L', '910×109/L', '99.7°F', 'normalization'] \n",
      "\n",
      "Sex \n",
      "Real: ['male'] \n",
      "Gene: ['male'] \n",
      "Match: ['male'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['altered mental status', 'blast phase', 'bone destruction', 'bone marrow involvement', 'confusion', 'dehydration', 'dehydration', 'disorientation', 'hematological response', 'hypercalcemia', 'inability to walk steadily', 'lesions', 'polydipsia', 'polyuria', 'symptoms', 'weakness'] \n",
      "Gene: ['altered mental status', 'bone destruction', 'confusion', 'disorientation', 'inability to walk steadily', 'lesions', 'polyuria', 'polydipsia', 'weakness'] \n",
      "Match: ['altered mental status', 'bone destruction', 'confusion', 'disorientation', 'inability to walk steadily', 'lesions', 'polydipsia', 'polyuria', 'weakness'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 13 year-old boy presented with a 1 month history of intermittent episodes of small volume epistaxis.\n",
      "There was no significant past medical, surgical or family history.\n",
      "Physical examination revealed a left palatal fleshy mass centered at left posterior molars.\n",
      "A nasoendoscopic examination demonstrated a large swelling on the floor of left anterior nasal space, involving the left inferior meatus and abutting the inferior turbinate.\n",
      "Computed Tomography (CT) of the paranasal sinuses showed a lobulated, expansile lytic lesion of left maxilla [Fig 1] with a lucent center and peripheral calcifications.\n",
      "The lesion was associated with a large soft tissue component that demonstrated mild heterogeneous contrast enhancement.\n",
      "The lesion was centered at the left maxillary alveolus, involving the left hard palate, inferior part of the anterior left maxillary wall and medial left maxillary wall.\n",
      "The lesion also involved the roots of the left maxillary molars and pre-molars and was abutting the left orbital floor, with no gross orbital invasion.\n",
      "The patient subsequently underwent 18-FDG PET-CT study.\n",
      "The mass centered in left maxilla was hypermetabolic, with index SUVmax 9.7 [Fig 2].\n",
      "There was no FDG-avid metastatic disease.\n",
      "A trans-oral incisional biopsy of the left palatal mass was performed and was suggestive of a preliminary diagnosis of mesenchymal chondrosarcoma, as evidenced by a biphasic pattern featuring sheets of small round cells with hemangiopericytoma-like vascular pattern and mature cartilaginous islands.\n",
      "The patient underwent left subtotal maxillectomy via lateral rhinotomy and translabial approach.\n",
      "The final histology was conclusive of a 5cm high-grade (Grade 3) mesenchymal chondrosarcoma.\n",
      "One of the resection margins was positive for tumor involvement and the patient was planned for adjuvant chemotherapy.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['13 year-old'] \n",
      "Gene: ['13 year-old'] \n",
      "Match: ['13 year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['Grade 3', 'high-grade', 'index SUVmax 9.7'] \n",
      "Gene: ['9.7', 'Grade 3', 'high-grade', 'positive', 'suggestive'] \n",
      "Match: ['Grade 3', 'high-grade'] \n",
      "\n",
      "Sex \n",
      "Real: ['boy'] \n",
      "Gene: ['boy'] \n",
      "Match: ['boy'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['contrast enhancement', 'epistaxis', 'invasion', 'lesion', 'mass', 'mass', 'swelling', 'tumor involvement'] \n",
      "Gene: ['epistaxis', 'invasion', 'mass', 'mass', 'metastatic disease', 'swelling'] \n",
      "Match: ['epistaxis', 'invasion', 'mass', 'mass', 'swelling'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "An 18-year-old lady presented with history of progressive swaying while walking, worse in night since 5 years.\n",
      "Speech was normal and there was not any tremulousness of upper limbs.\n",
      "Also, there was no history of weakness or numbness.\n",
      "The family history was significant as her elder sister had similar difficulty in walking of 7-year duration and died because of poorly controlled diabetes mellitus at an age of 20 years.\n",
      "No other members of the kin were affected and there was no history of consanguinity in the parents.\n",
      "The cognitive function was normal and her Mini Mental Status Score was 30/30.\n",
      "Visual acuity and fundus examination were normal.\n",
      "The extraocular movements including saccades, pursuits were normal, and no saccadic intrusions or nystagmus were noticed.\n",
      "The tone and power were normal in all four limbs; however, there was generalised exaggeration of myotatic reflexes along with bilateral extensor plantar response.\n",
      "Vibration and joint position sense were intact but Romberg's was positive.\n",
      "Patient swayed to either side while performing tandem walk.\n",
      "Fasting and postprandial blood glucose levels were normal.\n",
      "Haemogram, renal, liver and thyroid function tests were normal.\n",
      "Serum vitamin B12 and folic acid levels were within normal limits.\n",
      "Nerve conduction study showed absent sensory nerve action potential with normal distal latency, conduction velocity and amplitude of compound muscle action potential (CMAP) in all nerves trunks.\n",
      "MRI showed marked atrophy of the cervical cord as compared to cerebellum which was normal (figure 1).\n",
      "The genetic testing using PCR disclosed expansion of GAA repeat in both alleles (254 and 298) of FXN gene.\n",
      "A variety of causes of autosomal recessive ataxias need to be considered in the differential diagnosis of FRDA.\n",
      "These were systemically excluded here.\n",
      "Ataxia with vitamin E deficiency (AVED) has a phenotype quite similar to FRDA, though titubation and hyperkinesia are more common in AVED.4 A strong family history of ataxia and diabetes mellitus in elder sister favoured FRDA.\n",
      "Ataxia with oculomotor apraxia (AOA) types 1 and 2 were not considered owing to lack of apraxia of ocular movements, distal amyotrophy (severe axonal sensorimotor polyneuropathy), atrophy of cerebellum and involuntary movements.5 Ataxia telengiectesia has presentation very similar to AOA types 1 and 2.\n",
      "Muco-cutaneous markers, sino-pulmonary infections, hypogammaglobulinaemia and radiosensitivity along with predisposition to development of a variety of neoplasms are features of ataxia telengiectasia not seen in AOAs and FRDA.\n",
      "The other important causes of autosomal recessive spastic ataxia apart from FRDA would be autosomal recessive spastic ataxia of Charlevoix-Saguenay and Marinesco-Sjögren syndrome.\n",
      "Charcot-Marie-Tooth disease, an inherited polyneuropathy has gait ataxia in common with FRDA; however, in this case, CMT was improbable due to following features in this case: lack of motor weakness, retained reflexes, bilaterally extensor plantar and pure sensory neuropathy on nerve conduction study.\n",
      "Idebenone was started at a dose of 450 mg twice a day for its antioxidant properties.\n",
      "For gait ataxia, physiotherapy—Frenkel's exercise—was initiated.\n",
      "Blood glucose is monitored regularly.\n",
      "At 4 months, there was no significant change in the gait difficulty.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['18-year-old', '20\\u2005years'] \n",
      "Gene: ['18-year-old'] \n",
      "Match: ['18-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['30/30', 'absent', 'exaggeration', 'intact', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'poorly controlled', 'positive', 'worse'] \n",
      "Gene: ['30/30', 'absent', 'intact', 'normal', 'normal', 'normal', 'normal', 'normal', 'positive', 'retained', 'within normal limits'] \n",
      "Match: ['30/30', 'absent', 'intact', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'normal', 'positive'] \n",
      "\n",
      "Sex \n",
      "Real: ['lady'] \n",
      "Gene: ['lady'] \n",
      "Match: ['lady'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['Muco-cutaneous markers', 'apraxia of ocular movements', 'atrophy', 'atrophy', 'axonal sensorimotor polyneuropathy', 'bilateral extensor plantar response', 'bilaterally extensor plantar', 'difficulty in walking', 'distal amyotrophy', 'expansion of GAA repeat', 'gait ataxia', 'hyperkinesia', 'hypogammaglobulinaemia', 'infections', 'involuntary movements', 'motor weakness', 'neoplasms', 'numbness', 'pure sensory neuropathy', 'radiosensitivity', 'retained reflexes', 'swayed', 'swaying', 'titubation', 'tremulousness', 'weakness'] \n",
      "Gene: ['absent sensory nerve action potential', 'atrophy', 'atrophy', 'atrophy', 'atrophy', 'axonal sensorimotor polyneuropathy', 'change in the gait difficulty', 'difficulty in walking', 'exaggeration', 'extensor plantar response', 'extensor plantar response', 'gait ataxia', 'gait ataxia', 'generalised', 'hyperkinesia', 'involuntary movements', 'motor weakness', 'nystagmus', 'numbness', 'saccades', 'saccadic intrusions', 'tremulousness', 'weakness'] \n",
      "Match: ['atrophy', 'atrophy', 'axonal sensorimotor polyneuropathy', 'difficulty in walking', 'gait ataxia', 'hyperkinesia', 'involuntary movements', 'motor weakness', 'numbness', 'tremulousness', 'weakness'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 24-year-old Malay male patient was referred to the respiratory clinic because of an abnormal pre-employment chest radiograph.\n",
      "He had been smoking about 10 cigarettes a day since he was 21 years of age.\n",
      "He stopped smoking 10 months ago after he noticed he had being having reduced effort tolerance for the past three years.\n",
      "He was an office worker and did not have a history of exposure to organic or inorganic dusts.\n",
      "His two siblings were asymptomatic.\n",
      "On examination, the patient was not tachypnoeic.\n",
      "There were no signs of finger clubbing or pulmonary hypertension.\n",
      "His oxygen saturation on room air at rest was 94% and dropped to 92% after climbing up four flights of stairs.\n",
      "Spirometry testing revealed a restrictive pattern of lung disease with a forced expiratory volume in 1 second (FEV1) and a forced vital capacity (FVC) of 2.7 L (69% of predicted) and 3.2 L (68% of predicted), respectively.\n",
      "The FEV1/FVC ratio was 85%.\n",
      "His haemoglobin (168 g/L), serum parathyroid hormone (2.9 pmol/L [normal, 1.1-7.3]) and calcium (2.34 mmol/L) levels were normal.\n",
      "24-hour urine calcium was also normal 6.9 mmol with a 24-hour urine volume of 2.8 L.\n",
      "His chest radiograph (Fig.1A) revealed dense micronodular opacities distributed symmetrically and predominantly in the middle to lower zones of both lungs giving the classical \"sandstorm\" appearance.\n",
      "The cardiac borders were obscured by the sand-like opacities.\n",
      "A high-resolution computed tomography (HRCT) scan of the lungs (Fig.1B) showed widespread tiny microcalcifications throughout the lungs with a preponderance of microliths in the lower lobes.\n",
      "There were associated areas of interlobular septal thickening and ground-glass changes.\n",
      "Subpleural cystic changes were also seen in both lower lobes giving rise to the 'black pleura sign' (Fig.1C) (2).\n",
      "No pneumothorax or pleural effusion was present.\n",
      "Both the bronchial system (including the small bronchioles) and the size of the pulmonary vessels were normal.\n",
      "As this was diffuse parenchymal lung disease, videoassisted thoracic surgical (VATS) lung biopsy was planned but the procedure was converted into a mini-thoracotomy because there was difficulty in manoeuvering the endostapler.\n",
      "There was a moderate pneumothorax postmini-thoracotomy from which the patient fully recovered after 5 days in the ward.\n",
      "The lung biopsy specimen revealed features consistent with PAM, with numerous calcospherites within the alveolar spaces (Fig.1D).\n",
      "The intervening alveolar septae were congested and showed mild fibrosis with infiltrates of mainly lymphoplasmacytic cells.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['24-year-old'] \n",
      "Gene: ['24-year-old'] \n",
      "Match: ['24-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['168 g/L', '2.34 mmol/L', '2.7 L', '2.8 L', '2.9 pmol/L', '3.2 L', '6.9 mmol', '85%', '92%', '94%', 'abnormal', 'dropped', 'normal', 'normal', 'normal'] \n",
      "Gene: ['168 g/L', '2.34 mmol/L', '2.7 L', '2.9 pmol/L', '3.2 L', '85%', '92%', '94%', 'dropped', 'high', 'normal', 'normal', 'normal', 'reduced'] \n",
      "Match: ['168 g/L', '2.34 mmol/L', '2.7 L', '2.9 pmol/L', '3.2 L', '85%', '92%', '94%', 'dropped', 'normal', 'normal', 'normal'] \n",
      "\n",
      "Sex \n",
      "Real: ['male'] \n",
      "Gene: ['male'] \n",
      "Match: ['male'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['asymptomatic', 'black pleura sign', 'calcospherites', 'congested', 'cystic changes', 'fibrosis', 'finger clubbing', 'ground-glass changes', 'infiltrates', 'microcalcifications', 'microliths', 'obscured', 'opacities', 'opacities', 'reduced effort tolerance', 'septal thickening', 'tachypnoeic'] \n",
      "Gene: ['black pleura sign', 'clubbing', 'difficulty in manoeuvering', 'fibrosis', 'ground-glass changes', 'hypertension', 'infiltrates', 'microliths', 'opacities', 'opacities', 'pneumothorax', 'pneumothorax', 'recovered', 'thickening'] \n",
      "Match: ['black pleura sign', 'fibrosis', 'ground-glass changes', 'infiltrates', 'microliths', 'opacities', 'opacities'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 48-year-old man presented at our hospital complaining of intracranial bruits.\n",
      "The cranial nerve and laboratory examinations were normal.\n",
      "A DVAF was considered, conventional digital subtraction angiography was performed, which showed a Cognard II DVAF in the location of right hypoglossal canal (HC).\n",
      "The DVAF was fed by the meningeal branches of the bilateral ascending pharyngeal arteries, and the right anterior condylar veins within the hypoglossal canal that was the fistulous point mainly drained into the jugular vein.\n",
      "Endovascular treatment was insisted upon by the patient because of the associated intolerable intracranial bruit.\n",
      "Treatment was performed by transarterial approach under general anesthesia using a biplane angiographic unit.\n",
      "Standard coaxial techniques were used.\n",
      "The guide catheter was navigated into the left ascendingtrawt pharyngeal artery.\n",
      "Marathon flow directed catheter (eV3) was subsequently navigated over a Mirage.008 microwire (eV3) to reach as near as possible to the fistula.\n",
      "The microcatheter was flushed with 10 mL of normal saline.\n",
      "The dead space of the microcatheter was subsequently filled with dimethyl sulfoxide (DMSO).\n",
      "Onyx-18 was then injected over two minutes to fill the microcatheter and to replace the DMSO in the dead space.\n",
      "The injection was stopped until unwanted flow into branches of the right ascending pharyngeal artery was observed.\n",
      "The follow-up angiogram showed no residual shunt (Figure ​1).\n",
      "The intracranial bruits disappeared immediately after the operation.\n",
      "For the toxicity of DMSO, 10 mg dexamethasone was intravenously injected for three days.\n",
      "On the second day after endovascular treatment, the patient complainted of difficulty moving his tongue to the left.\n",
      "Cranial nerve examination was only significant for a right hypoglossal nerve palsy.\n",
      "The patient was discharged from our hospital four days after treatment without continuative pharmacotherapy.\n",
      "At two months follow-up, the patient's main complaint were difficulty swallowing (dysphagia) and slurred speech (dysarthria).\n",
      "In addition, his tongue deviated toward the right during tongue protrusion, and marked right-sided hemiatrophy was observed (Figure ​2).\n",
      "The cranial nerve examation showed right hypoglossal nerve palsy persisted.\n",
      "We telephone the patient at three months follow-up and requested a vocal cord evaluation by an ENT specialist in a local hospital.\n",
      "Vocal cord paralysis was not found, and there was no evidence of vagus nerve dysfunction.\n",
      "Oral Vitamin B12 was used and his prognosis was still at a long follow-up.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['48-year-old'] \n",
      "Gene: ['48-year-old'] \n",
      "Match: ['48-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['normal', 'tongue deviated toward the right'] \n",
      "Gene: ['normal'] \n",
      "Match: ['normal'] \n",
      "\n",
      "Sex \n",
      "Real: ['man'] \n",
      "Gene: ['man'] \n",
      "Match: ['man'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['bruits', 'bruits', 'difficulty moving his tongue', 'dysarthria', 'dysfunction', 'dysphagia', 'fistula', 'fistulous point', 'hemiatrophy', 'paralysis', 'residual shunt', 'right hypoglossal nerve palsy'] \n",
      "Gene: ['bruits', 'bruits', 'dysphagia', 'dysarthria', 'dysphagia', 'hemiatrophy', 'palsy', 'palsy', 'shunt', 'slurred speech'] \n",
      "Match: ['bruits', 'bruits', 'dysarthria', 'dysphagia', 'hemiatrophy'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 21-year-old male presented with a 2-year history of progressive shortness of breath on exertion and dry cough.\n",
      "At physical examination, auscultation of the lungs has revealed random wheezes and coarse crackles.\n",
      "Cardiac auscultation was normal, and no cyanosis or peripheral edema was observed.\n",
      "There was no history of smoking or previous known pulmonary disease.\n",
      "On routine blood examination, blood counts and serum chemistries were found to be normal.\n",
      "Arterial blood gas analysis and echocardiography showed no important abnormalities.\n",
      "Pulmonary function tests (PFT) showed a mild restrictive ventilatory defect, with a reduced total lung capacity of 79% (5.94 L), forced vital capacity of 80% (4.18 L) and a forced expiratory volume in one second of 83% (3.72 L).\n",
      "The sputum was negative for acid-alcohol resistant bacillus and human immunodeficiency virus testing was negative as well.\n",
      "The chest plain films revealed a diffuse symmetric dense bilateral micronodular pattern (Figure 1).\n",
      "Based on this finding, HRCT scan was obtained, revealing diffuse ground glass attenuation and septal thickening, more pronounced in lower pulmonary regions, with calcifications along the interlobar septa and subpleural regions.\n",
      "Subpleural cysts were also noticed (Figure 2).\n",
      "The patient underwent a fiberoptic bronchoscopy with bronchoalveolar lavage and transbronchial lung biopsy.\n",
      "The lavage fluid was negative for tuberculosis or fungi.\n",
      "Microliths were not found.\n",
      "Histology revealed round, concentrically laminated microliths in the alveoli associated with slightly thickened interstitial septa, consistent with the diagnosis of PAM.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['21-year-old'] \n",
      "Gene: ['21-year-old'] \n",
      "Match: ['21-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['3.72\\u2009L', '4.18\\u2009L', '5.94\\u2009L', '79%', '80%', '83%', 'negative', 'negative', 'negative', 'no important abnormalities', 'normal', 'reduced'] \n",
      "Gene: ['79%', '80%', '83%', 'diffuse', 'diffuse', 'negative', 'negative', 'normal', 'normal', 'pronounced', 'reduced', 'slightly thickened'] \n",
      "Match: ['79%', '80%', '83%', 'negative', 'negative', 'negative', 'normal', 'reduced'] \n",
      "\n",
      "Sex \n",
      "Real: ['male'] \n",
      "Gene: ['male'] \n",
      "Match: ['male'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['Microliths', 'calcifications', 'coarse crackles', 'cyanosis', 'cysts', 'dry cough', 'ground glass attenuation', 'microliths', 'micronodular pattern', 'peripheral edema', 'restrictive ventilatory defect', 'septal thickening', 'shortness of breath', 'thickened interstitial septa', 'wheezes'] \n",
      "Gene: ['Microliths', 'acid-alcohol resistant bacillus', 'calcifications', 'cyanosis', 'cough', 'cysts', 'edema', 'microliths', 'micronodular pattern', 'peripheral edema', 'restrictive ventilatory defect', 'shortness of breath', 'thickened interstitial septa', 'wheezes'] \n",
      "Match: ['Microliths', 'calcifications', 'cyanosis', 'cysts', 'microliths', 'micronodular pattern', 'peripheral edema', 'restrictive ventilatory defect', 'shortness of breath', 'thickened interstitial septa', 'wheezes'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "The affected individual, a male, was the fourth child born to an Arab Muslim family of Palestinian origin.\n",
      "Parents are first cousins.\n",
      "One sibling, a girl died at the age of 1 month due to SIDS according to parents, without any investigation.\n",
      "Informed consent for muscle, skin biopsy and exome sequencing was obtained from the parents, and the study was approved by the local institutional review board.\n",
      "During pregnancy, the patient was suspected to have IUGR and was delivered by Cesarian section at 36+2 gestational weeks with Apgar scores 9 and 10.\n",
      "Birth weight was 2.3 kg (10th percentile) and head circumference 32 cm (10th percentile).\n",
      "Physical exam after birth, detected a systolic murmur 3/6.\n",
      "On the first day of life, the patient developed metabolic acidosis with lactate level of 27 mm (normal range 0.5–2.4 mm) and elevated ammonia levels 277 μm (normal range 10–60 μm).\n",
      "Metabolic work up showed normal acylcarnitines, elevated alanine level 1400 μm (normal <700 μm), and urine organic acids showed increased lactic acidosis and ketones.\n",
      "Chest X-ray revealed cardiomegaly, and cardiac Echo revealed symmetrical left ventricular hypertrophy tricuspid regurge and pulmonary hypertension.\n",
      "The patient was transported to the NICU and treated with bicarbonate, acetate and diuretics.\n",
      "At the age of 4 days, quadriceps muscle biopsy was performed.\n",
      "Pathology showed decreased cytochrome c oxidase stain.\n",
      "Brain ultrasound performed at age 5 days was normal.\n",
      "Subsequently, head circumference grew rapidly adding 13 centimeters by age 35 days, repeated brain US revealed dilated ventricles confirmed by head CT, which also showed subcortical and white matter cortical hemorrhage in the occipital region.\n",
      "At 15 weeks, a ventriculoperitoneal (VP) shunt was inserted.\n",
      "Head CT performed at age 2 years showed prominent dilatation of the ventricles (Figure 1).\n",
      "Due to hypotonia with feeding difficulties and recurrent aspirations, a percutaneous endoscopic gastrostomy (PEG) feeding tube was inserted at age 10 weeks.\n",
      "Eye examination at the age of 15 weeks revealed cortical blindness.\n",
      "Hearing test ABR was normal.\n",
      "Repeated cardiac Echo continued to show mild tricuspid regurge, mild left ventricular hypertrophy (LVH) and pulmonary hypertension (PHTN) in the first months of life.\n",
      "At the age of 4 months, he started developing hypertrophic obstructive cardiomyopathy, he was started on diuretics and beta blockers and later calcium channel blockers were added; follow-up Echo at the age of 2 years revealed improvement in the LVH and no PHTN.\n",
      "Coenzyme Q10 supplementation (60 mg twice daily) was initiated at 4 months but discontinued by the parents.\n",
      "In the following months, the patient was admitted several times to the hospital mainly due to chest infections, shunt malfunctioning and fevers.\n",
      "He died at the age of 30 months due to severe hypoxemia after a febrile illness with chest infection and pulmonary edema; parents refused mechanical ventilation and refused reanimation.\n",
      "\n",
      "\n",
      "Lab_value \n",
      "Real: ['10', '10th percentile', '10th percentile', '1400\\u2009μm', '2.3\\u2009kg', '277\\u2009μm', '27\\u2009mm', '3/6', '32\\u2009cm', '9', 'adding 13 centimeters', 'decreased', 'elevated', 'elevated', 'grew rapidly', 'normal', 'normal', 'normal'] \n",
      "Gene: ['10', '13 centimeters', '277\\u2009μm', '9', 'grew rapidly', 'normal', 'normal', 'normal', 'rapidly', 'subcortial and white matter'] \n",
      "Match: ['10', '277\\u2009μm', '9', 'grew rapidly', 'normal', 'normal', 'normal'] \n",
      "\n",
      "Sex \n",
      "Real: ['male'] \n",
      "Gene: ['male'] \n",
      "Match: ['male'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['LVH', 'LVH', 'PHTN', 'PHTN', 'cardiomegaly', 'dilatation', 'dilated', 'edema', 'fevers', 'hemorrhage', 'hypoxemia', 'ketones', 'lactic acidosis', 'left ventricular hypertrophy', 'left ventricular hypertrophy', 'pulmonary hypertension', 'pulmonary hypertension', 'systolic murmur', 'tricuspid regurge', 'tricuspid regurge'] \n",
      "Gene: ['cardiomegaly', 'dilatation', 'dilatation', 'febrile illness', 'febres', 'hypoxemia', 'hypertrophic', 'hypertrophy', 'hypertrophy', 'hypertrophy', 'hypotonia', 'infections', 'left ventricular hypertrophy', 'left ventricular hypertrophy', 'metabolic acidosis', 'murmur', 'pulmonary edema', 'pulmonary hypertension', 'pulmonary hypertension', 'shunt malfunctioning', 'subcortial and white matter cortical hemorrhage'] \n",
      "Match: ['cardiomegaly', 'dilatation', 'hypoxemia', 'left ventricular hypertrophy', 'left ventricular hypertrophy', 'pulmonary hypertension', 'pulmonary hypertension'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 76-year-old Japanese man had experienced a decreased appetite, nausea, and abdominal distension since late February 2011.\n",
      "The patient passed loose, soft stools that were not watery once daily.\n",
      "The results of blood tests performed by the primary care physician (PCP) between April and June were all normal, but upper and lower gastrointestinal endoscopy showed atrophic gastritis.\n",
      "Therefore, we assessed the patient for Helicobacter pylori antibodies and performed the rapid urease test and a histopathological evaluation, all of which were negative.\n",
      "In June, the patient received medication from a mental health clinic, but because the symptoms did not improve and his body weight further decreased by 23 kg, he stopped the medication; he was referred to our department in November.\n",
      "The patient had been suffering from hypertension since his 30s.\n",
      "Amlodipine (5 mg/day) administration was initiated in 2005, and olmesartan from May 2008; the patient is currently receiving 30 mg/day of olmesartan.\n",
      "He denied the use of any other new medications or nonsteroidal anti-inflammatory drugs.\n",
      "He had no history of smoking or alcohol consumption.\n",
      "Physical examination showed that the patient had a height of 165 cm, a body weight of 47 kg, and body mass index of 17 kg/m2.\n",
      "His body temperature was 36.3℃; pulse rate, 101/min; blood pressure, 101/81 mmHg; respiratory rate, 12 breaths/min; and saturation from pulse oximetry (SpO2) was 98%.\n",
      "The patient had no heart murmur, third heart sound or jugular venous distension.\n",
      "The patient had bilateral pitting edema on his lower legs and presented with unilateral gaze-evoked nystagmus as well as a mildly reduced tactile sensation and thermal nociception in the toes and dorsal regions of both feet.\n",
      "Moreover, the finger-to-nose test results and tandem gait were poor, and his patellar and Achilles tendon reflexes had disappeared.\n",
      "Confabulation was observed in the patient, and the revised Hasegawa Dementia Scale (HDS-R) score was 17/30 (cut-off point: 20).\n",
      "The hematologic findings were as follows: white blood cell count, 6,200 /μL; hemoglobin level, 12.5 g/dL; mean corpuscular volume, 88.4; platelet count, 230,000 /μL; sodium level, 136 mEq/L; potassium level, 3.8 mEq/L; chlorine level, 103 mEq/L; iron level, 57 μg/dL (reference value: 64-187 μg/dL); ferritin level, 366 ng/mL (reference value: 50-200 ng/mL); B-type natriuretic peptide (BNP) level, 125.3 pg/mL (reference value: -18.4 pg/mL); and vitamin B1 level, 8 ng/mL (reference value: 24-66 ng/mL).\n",
      "An electrocardiogram was normal and chest X-rays showed a normal cardiothoracic ratio (40.8%) without either pulmonary congestion or pleural effusion.\n",
      "Cranial fluid-attenuated inversion recovery-magnetic resonance imaging findings revealed periaqueductal hyperintensities (Fig.1); therefore, Wernicke encephalopathy was diagnosed.\n",
      "Moreover, as sinus tachycardia and a tendency towards hypotension were noted, no clear symptoms of heart failure or dehydration were observed; it was thus inferred that the vitamin B1 deficiency had likely played a role in both of the conditions.\n",
      "The antihypertensive agents were discontinued, and 10 days after the intravenous administration of vitamin B1, the patient's loss of appetite, nausea, and gait disturbance disappeared, and his body weight increased by 3 kg.\n",
      "Nystagmus was ameliorated on physical examination, but the patient still had confabulations, and the HDS-R score and absence of deep tendon reflexes did not improve.\n",
      "Since the gastrointestinal symptoms were ameliorated and his blood pressure increased to 160/90 mmHg, the PCP resumed the administration of olmesartan on late December, 2011.\n",
      "One week later, the patient complained of recurrent decreased appetite and nausea.\n",
      "After experiencing diarrhea (five bowel movements during a 2-day period), he passed soft stools once daily, and his body weight decreased to 47 kg so he came to our hospital again 3 weeks after the resumption of olmesartan for treatment.\n",
      "His vital signs were as follows: body temperature, 35.1℃; pulse rate, 93/min; and blood pressure, 132/75 mmHg.\n",
      "The neurologic findings showed no worsening of his symptoms.\n",
      "The laboratory findings were as follows: sodium level, 139 mEq/L; potassium level, 2.9 mEq/L; and chlorine level, 116 mEq/L.\n",
      "Hyperchloremia was noted, and the serum sodium level minus chloride level (139-116) was 23 mEq/L; additionally, the arterial blood gas findings were as follows: pH, 7.25; PCO2, 25 mmHg, and HCO3, 11 mmol/L.\n",
      "The urinalysis findings were as follows: sodium level, 15 mEq/L; potassium level, 13 mEq/L; chlorine level, 100 mEq/L; and urine anion gap, -72 mEq/L, and there was no apparent increase in bowel movements; however, this seemed to be due to the absence of HCO3- from the gastrointestinal tract (6).\n",
      "Plain abdominal radiography and abdominal computed tomography findings revealed the continuous dilation of the entire intestinal tract (Fig.2); therefore, the patient was diagnosed with intestinal pseudo-obstruction.\n",
      "After discontinuing olmesartan, the nausea and body weight decreases were promptly ameliorated.\n",
      "Duodenal biopsy findings after discontinuing olmesartan showed that the inflammation observed before discontinuation as well as the fold structure had both clearly ameliorated (Fig.3).\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['76-year-old'] \n",
      "Gene: ['76-year-old'] \n",
      "Match: ['76-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['-72 mEq/L', '100 mEq/L', '101/81 mmHg', '101/min', '103 mEq/L', '11 mmol/L', '116 mEq/L', '12 breaths/min', '12.5 g/dL', '125.3 pg/mL', '13 mEq/L', '132/75 mmHg', '136 mEq/L', '139 mEq/L', '15 mEq/L', '160/90 mmHg', '17 kg/m2', '17/30', '2.9 mEq/L', '230', '000 /μL', '25 mmHg', '3.8 mEq/L', '35.1℃', '36.3℃', '366 ng/mL', '40.8%', '47 kg', '57 μg/dL', '6', '200 /μL', '7.25', '8 ng/mL', '88.4', '93/min', '98%', 'absence', 'decreased', 'decreased by 23 kg', 'did not improve', 'disappeared', 'five bowel movements during a 2-day period', 'increased', 'increased by 3 kg', 'loose', 'negative', 'no apparent increase', 'no worsening', 'normal', 'normal', 'normal', 'not watery', 'poor', 'serum sodium level minus chloride level (139-116) was 23 mEq/L', 'soft', 'soft'] \n",
      "Gene: ['101/81 mmHg', '101/min', '11 mmol/L', '12.5 g/dL', '125.3 pg/mL', '132/75 mmHg', '136 mEq/L', '139 mEq/L', '139-116', '14', '15 mEq/L', '160/90 mmHg', '17/30', '17 kg/m2', '2.9 mEq/L', '23 kg', '230', '000 /μL', '3.8 mEq/L', '35.1℃', '36.3℃', '366 ng/mL', '47 kg', '47 kg', '57 μg/dL', '6', '200 /μL', '7.25', '8 ng/mL', '88.4', '93/min', '98%', 'decreased', 'decreased', 'decreased', 'decreased', 'increased', 'increased', 'increased', 'lost', 'normal', 'normal', 'poor', 'reduced', 'worsening'] \n",
      "Match: ['101/81 mmHg', '101/min', '11 mmol/L', '12.5 g/dL', '125.3 pg/mL', '132/75 mmHg', '136 mEq/L', '139 mEq/L', '15 mEq/L', '160/90 mmHg', '17 kg/m2', '17/30', '2.9 mEq/L', '230', '000 /μL', '3.8 mEq/L', '35.1℃', '36.3℃', '366 ng/mL', '47 kg', '57 μg/dL', '6', '200 /μL', '7.25', '8 ng/mL', '88.4', '93/min', '98%', 'decreased', 'increased', 'normal', 'normal', 'normal', 'poor'] \n",
      "\n",
      "Sex \n",
      "Real: ['man'] \n",
      "Gene: ['man'] \n",
      "Match: ['man'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['Confabulation', 'Hyperchloremia', 'Nystagmus', 'absence of HCO3-', 'body weight decreases', 'confabulations', 'decreased appetite', 'decreased appetite', 'dehydration', 'diarrhea', 'dilation', 'distension', 'distension', 'edema', 'fold structure', 'gait disturbance', 'gastrointestinal symptoms', 'heart murmur', 'hyperintensities', 'hypotension', 'inflammation', 'loss of appetite', 'nausea', 'nausea', 'nausea', 'nausea', 'nociception', 'nystagmus', 'pulmonary congestion', 'reduced tactile sensation', 'symptoms', 'tachycardia', 'third heart sound'] \n",
      "Gene: ['Confabulation', 'Wernicke encephalopathy', 'abdominal distension', 'abdominal fold structure', 'abdominal hyperintensities', 'abdominal inflammation', 'appetite', 'appetite', 'appetite', 'confabulations', 'diarrhea', 'deep tendon reflexes', 'dehydration', 'distension', 'edema', 'fold structure', 'gait disturbance', 'gaze-evoked nystagmus', 'gastrointestinal fold structure', 'gastrointestinal symptoms', 'gastrointestinal tract', 'heart murmur', 'hypotension', 'hyperintensities', 'loss of appetite', 'nausea', 'nausea', 'nausea', 'nystagmus', 'pleural effusion', 'pitting edema', 'pulmonary congestion', 'reflexes', 'reflexes', 'reflexes', 'sinus tachycardia', 'stools', 'stools', 'symptoms', 'symptoms', 'symptoms', 'symptoms', 'tendency towards hypotension', 'tendon reflexes', 'tendon reflexes', 'tactile sensation', 'thermal nociception', 'vitamin B1 deficiency'] \n",
      "Match: ['Confabulation', 'confabulations', 'dehydration', 'diarrhea', 'distension', 'distension', 'edema', 'fold structure', 'gait disturbance', 'gastrointestinal symptoms', 'heart murmur', 'hyperintensities', 'hypotension', 'loss of appetite', 'nausea', 'nausea', 'nausea', 'nausea', 'nystagmus', 'pulmonary congestion', 'symptoms'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 16-year-old female suffered from an abdominal pain at right upper quadrant, lasting for more than 10 days.\n",
      "Shortly ahead of her medical consultation, 4 ascaris-like worms were vomited out, individually with a length of about 10 cm.\n",
      "She had a previous onset of ascariasis, during which low-grade fever (38.4°C) occurred without apparent jaundice, diarrhea, or anemia.\n",
      "Physical examination revealed tenderness at right upper quadrant.\n",
      "Her total leukocyte count was 11.2 G/L consisting of 5.2% eosinophils.\n",
      "Serum and urine amylase were 386 and 928 U/L, respectively.\n",
      "In terms of liver functionality, the level of total bilirubin rose up to 23.2 μm/L, meanwhile the hepatic enzymes were similarly elevated (alanine aminotransferase 163 U/L; aspartate aminotransferase 96 U/L).\n",
      "Abdominal ultrasound described the enlargement of the gallbladder, upper segment of common bile duct (1.5 cm in diameter) and intrahepatic bile duct (1.3 cm in diameter).\n",
      "Furthermore, the intrahepatic bile duct was also discovered to be filled with echogenic and banded substances.\n",
      "Based on such evidences, she was clinically diagnosed as biliary ascariasis with a concomitant manifestation of biliary duct infection.\n",
      "Through a laparoscopic exploration, the enlarged gallbladder and common bile duct were visually confirmed, as well as the massive storage of worms within the biliary tract.\n",
      "Subsequently, a 1 cm incision was made on the common bile duct in order to facilitate the extraction of the pathogenic worms, which were then placed into a specimen bag laparoscopically (Fig.1).\n",
      "Once the contained worms were virtually cleared out under laparoscope, chodedochoscopy was employed to seek for more hidden worms inside the biliary system (Fig.2).\n",
      "After finishing the removal of ascarides, the biliary duct incision was well sutured, with a T-tube placed in situ (Fig.3).\n",
      "Eventually, a standard laparoscopic cholecystectomy was performed, followed by the removal of specimen bags and drainage tube placement (Fig.4).\n",
      "The total amount of worms inside the common bile duct summed up to more than 100 (Fig.5).\n",
      "Postoperatively, she was treated with oral antihelminthic drugs for 3 days.\n",
      "The T-tube was rinsed with saline every day, and during the first 3 days, 6 more worms were additionally washed out.\n",
      "The patient experienced a smooth postoperative recovery and was finally discharged from hospital without any complications.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['16-year-old'] \n",
      "Gene: ['16-year-old'] \n",
      "Match: ['16-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['11.2\\u200aG/L', '163\\u200aU/L', '23.2\\u200aμm/L', '38.4°C', '4', '5.2% eosinophils', '6', '928\\u200aU/L', '96\\u200aU/L', 'elevated', 'low-grade', 'more than 100', 'rose up', 'smooth'] \n",
      "Gene: ['1.3\\u200acm in diameter', '1.5\\u200acm in diameter', '10\\u200acm', '11.2\\u200aG/L', '163\\u200aU/L', '23.2\\u200aμm/L', '38.4°C', '386', '928\\u200aU/L', '5.2% eosinophils', '96\\u200aU/L', 'rose up', 'similarly elevated'] \n",
      "Match: ['11.2\\u200aG/L', '163\\u200aU/L', '23.2\\u200aμm/L', '38.4°C', '5.2% eosinophils', '928\\u200aU/L', '96\\u200aU/L', 'rose up'] \n",
      "\n",
      "Sex \n",
      "Real: ['female'] \n",
      "Gene: ['female'] \n",
      "Match: ['female'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 35-year-old woman presented to the ear, nose and throat outpatient department with headache and a blocked left-sided nasal canal with epistaxis for 2 years.\n",
      "Medical history revealed secondary amenorrhoea for the past 19 years with regression of secondary sexual characteristics with no galactorrhoea.\n",
      "Physical and ophthalmological examinations were within normal limits.\n",
      "Nasal examination revealed a left-sided choanal mass.\n",
      "As the mass was bleeding on palpation, imaging was advised prior to performing a biopsy.\n",
      "MRI of the head showed a sellar and infra-sellar mass extending into the nasopharynx (figure 1).\n",
      "The patient was then referred to the endocrinologist for further hormonal investigation.\n",
      "Her serum prolactin was very high—7443 µg/L (N<20).\n",
      "Other pituitary hormones were within normal limits.\n",
      "A final diagnosis of giant prolactinoma was made and the patient was prescribed cabergoline 0.25 mg twice per week with the associated risks explained in advance.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['35-year-old'] \n",
      "Gene: ['35-year-old'] \n",
      "Match: ['35-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['7443\\u2005µg/L', 'very high', 'within normal limits', 'within normal limits'] \n",
      "Gene: ['7443\\u2005µg/L', 'within normal limits', 'within normal limits'] \n",
      "Match: ['7443\\u2005µg/L', 'within normal limits', 'within normal limits'] \n",
      "\n",
      "Sex \n",
      "Real: ['woman'] \n",
      "Gene: ['woman'] \n",
      "Match: ['woman'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['amenorrhoea', 'blocked left-sided nasal canal', 'epistaxis', 'galactorrhoea', 'headache', 'mass', 'mass', 'regression of secondary sexual characteristics'] \n",
      "Gene: ['blocked', 'choal', 'epistaxis', 'galactorrhoea', 'headache', 'mass', 'mass', 'regression of secondary sexual characteristics'] \n",
      "Match: ['epistaxis', 'galactorrhoea', 'headache', 'mass', 'mass', 'regression of secondary sexual characteristics'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 24-year-old Azeri male presented with a chief complaint of blurred vision in both eyes.\n",
      "He was diagnosed with achondroplasia in early childhood.\n",
      "He was born at term with normal birth weight.\n",
      "There was no history of hereditary ocular or systemic disease.\n",
      "This patient had characteristic features of achondroplasia (i.e., short stature, lumbar lordosis, short proximal long bones, large hands, frontal bossing, and flat nasal bridge) [Figure 1].\n",
      "His height was 108 cm.\n",
      "Best corrected visual acuity was 20/70 in the right eye with −0.75 + 1.25 × 60° and 20/70 in the left eye with OS: +0.25 + 0.50 × 130°.\n",
      "Pupils were equal in size, round and reactive to light with no reverse afferent pupillary defect.\n",
      "Ocular motility was normal in both eyes and slit lamp examination was unremarkable.\n",
      "The intraocular pressure was 12 mm Hg by Goldmann applanation tonometry.\n",
      "Central corneal thickness values of the right and left eyes were, 561 and 557 μ, respectively.\n",
      "Dilated fundus examination revealed a well-demarcated circular macular lesion in both eyes.\n",
      "In the right eye, this lesion was approximately 2.5 disc diameters (DD) in width and 2.7 DD in length.\n",
      "In the left eye, this lesion was approximately 2.5 DD in width and 2.8 DD in length.\n",
      "The border of the lesions was pigmented.\n",
      "Underlying choroidal and overlying retinal vasculature appeared normal.\n",
      "In addition, a stalk of tissue emanated from the optic disc and coursed into the vitreous [Figure 2].\n",
      "Fluorescein angiography of the posterior segment and optical coherence tomography of the optic nerve head (Stratus; Carl Zeiss GmbH, Jena, Germany) were performed [Figures ​[Figures3 and 4].\n",
      "Due to the background maculopathy, the patient could not appropriately fixate on the target during image acquisition resulting in slightly decentered fluorescein angiography images [Figure 3].\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['24-year-old'] \n",
      "Gene: ['24-year-old'] \n",
      "Match: ['24-year-old'] \n",
      "\n",
      "Sex \n",
      "Real: ['male'] \n",
      "Gene: ['male'] \n",
      "Match: ['male'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['blurred vision', 'characteristic features of achondroplasia', 'lesion', 'lesion', 'lesion', 'lesions', 'reverse afferent pupillary defect', 'stalk of tissue'] \n",
      "Gene: ['blurred vision', 'decentered', 'fixate', 'frontal bossing', 'large hands', 'macular lesion', 'pigmented', 'reverse afferent pupillary defect', 'short stature', 'slightly'] \n",
      "Match: ['blurred vision', 'reverse afferent pupillary defect'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 28-year-old woman was referred with a 4-week history of continuous, moderate right upper quadrant pain associated with jaundice, as well as weight loss (10 kg over 3 months) and a liver mass identified by ultrasonography.\n",
      "The pain sensation seemed different from previous colicky attacks the patient had experienced before she underwent laparoscopic cholecystectomy 7 years previously.\n",
      "On physical examination, she was obese (body mass index 37.8), with icterus noted over the conjunctivae, oral mucosa and skin.\n",
      "Imaging modalities included computed tomography (CT), positron emission tomography (PET) and endoscopic cholangiopancreatography (ERCP) (Fig.1).\n",
      "The patient underwent exploratory laparotomy.\n",
      "Intraoperative ultrasonography revealed a cystic lesion measuring 3.5 2.5 cm within the central portion of the liver, anterior to the porta hepatis.\n",
      "Intraoperative cholangiography demonstrated an extensive stricture obliterating the left hepatic duct, with partial occlusion of the right hepatic duct.\n",
      "An extended left lobectomy was done en bloc with the biliary confluence.\n",
      "On frozen-section examination, all margins of the excised specimen were free of malignant cells.\n",
      "Reconstruction was performed with a Roux-en-Y cholangiojejunostomy to 3 second bile duct radicals in the right side.\n",
      "Intraoperatively, radiotherapy was applied to the surgical margins.\n",
      "After this, the patient underwent 6 weeks of image-guided external beam radiation centred on the resection field labelled at surgery.\n",
      "The final pathology described this tumour as an infiltrating, moderately differentiated squamous cell carcinoma associated with severe dysplasia of the bile-duct epithelium (Fig.2).\n",
      "The patient recovered without complications and was doing well 18 months after the initial surgical procedure, with an unremarkable CT scan.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['28-year-old'] \n",
      "Gene: ['28-year-old'] \n",
      "Match: ['28-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['10 kg', '37.8', 'different', 'free of malignant cells', 'moderately', 'unremarkable'] \n",
      "Gene: ['10 kg', '37.8', 'doing well', 'icterus', 'moderate', 'over 3 months', 'unremarkable'] \n",
      "Match: ['10 kg', '37.8', 'unremarkable'] \n",
      "\n",
      "Sex \n",
      "Real: ['woman'] \n",
      "Gene: ['woman'] \n",
      "Match: ['woman'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['colicky attacks', 'complications', 'doing well', 'icterus', 'jaundice', 'lesion', 'mass', 'obese', 'occlusion', 'pain', 'recovered', 'stricture', 'weight loss'] \n",
      "Gene: ['icterus', 'infiltrating', 'jaundice', 'lesion', 'mass', 'malignant cells', 'obese', 'pain', 'pain', 'stricture', 'tumour', 'weight loss'] \n",
      "Match: ['icterus', 'jaundice', 'lesion', 'mass', 'obese', 'pain', 'stricture', 'weight loss'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "A 47-year-old female patient presented progressively worsening pain in the chest wall, back, and bilateral lower extremities as well as muscle weakness.\n",
      "She did not have a family history of bone disease or fractures.\n",
      "Physical examination was unremarkable excluding bone and muscular abnormalities.\n",
      "Bone scintigraphy using technetium-99m methylenediphosphate showed increased uptake in the shoulder, multiple ribs, thoracic and lumbar spines, bilateral sacroiliac joints, left ilium, and left foot (Fig.1).\n",
      "Plain radiographs revealed osteoporosis change of thoracic and lumbar spine as well as compression change of several spines, consistent with findings of magnetic resonance imaging and computed tomography (CT) (Fig.1).\n",
      "Decreased bone mass was presented in the lumbar spine (T-score: −3.4), femoral neck (T-score: −3.1), and hip (T-score: −3.5).\n",
      "Main laboratory data are shown in Table ​1.\n",
      "She demonstrated hypophosphatemia, hypokalemia, hypouricemia, elevated level of ALP, C-telopeptides and chloride, and decreased carbon dioxide combining power.\n",
      "The urinalysis revealed a high pH value, increased level of potassium, and decreased level of specific gravity, chlorine, and phosphate.\n",
      "Persistent glycosuria and proteinuria were repeatedly found, despite normal HbA1c and plasma glucose level.\n",
      "Other laboratory test results including thyroid function test, serum parathyroid hormone, 25-hydroxyvitamin D, and protein electrophoresis were within normal range.\n",
      "In addition, all of serum tumor markers were negative.\n",
      "Due to limited technique, we cannot determine the level of serum fibroblast growth factor 23.\n",
      "With the clinical diagnosis of HO, fluorodeoxyglucose positron emission tomography/computed tomography (18F-FDG PET/CT) and technetium-99m octreotide (99mTc-OCT) scintigraphy were performed to confirm whether the occult causative tumor exist.\n",
      "However, the results of these 2 tests were negative except that mild uptake in the seventh rib was found on PET/CT, which identified no evidence of a neoplastic lesion potentially responsible for HO (Fig.2).\n",
      "The immunological examination showed elevated level of serum IgG, IgM, and IgA, as well as positive antinuclear antibody, anti-SSA antibody, and rheumatic factor.\n",
      "Subsequently, Schirmer test was abnormal and lip biopsy supported the diagnosis of SS (Fig.3).\n",
      "Eventually, this patient was diagnosed with HO secondary to SS, and she was then treated with alkalinization (citrate 4 g/day and potassium citrate 3 g/day for 2 weeks), steroids (prednisone 20 mg/day for 1 month, 10 mg/day for 4 months), neutral phosphate (1.0 g/day for 5 months), calcium supplements (600 mg/day for 5 months), and together with activated vitamin D (0.5 g/day for 5 months).\n",
      "So far, she recovered uneventfully with relieved pain and increased serum phosphorus level.\n",
      "This case report was approved by the ethics committee of West China Hospital of Sichuan University, Chengdu, China, and the written informed consent was obtained.\n",
      "\n",
      "\n",
      "Age \n",
      "Real: ['47-year-old'] \n",
      "Gene: ['47-year-old'] \n",
      "Match: ['47-year-old'] \n",
      "\n",
      "Lab_value \n",
      "Real: ['Decreased', 'T-score: −3.1', 'T-score: −3.4', 'T-score: −3.5', 'abnormal', 'decreased', 'decreased', 'elevated', 'elevated', 'high', 'increased', 'increased', 'negative', 'negative', 'normal', 'positive', 'unremarkable', 'within normal range'] \n",
      "Gene: ['+3.1', '+3.4', 'decreased', 'decreased', 'elevated', 'elevated', 'increased', 'increased', 'negative', 'negative', 'normal', 'positive', 'relieved', 'uneventfully', 'within normal range'] \n",
      "Match: ['decreased', 'decreased', 'elevated', 'elevated', 'increased', 'increased', 'negative', 'negative', 'normal', 'positive', 'within normal range'] \n",
      "\n",
      "Sex \n",
      "Real: ['female'] \n",
      "Gene: ['female'] \n",
      "Match: ['female'] \n",
      "\n",
      "Sign_symptom \n",
      "Real: ['compression', 'glycosuria', 'hypokalemia', 'hypophosphatemia', 'hypouricemia', 'lesion', 'muscle weakness', 'pain', 'pain', 'proteinuria', 'recovered', 'tumor', 'uptake', 'uptake'] \n",
      "Gene: ['abnormal', 'change', 'compression', 'decreased', 'decreased', 'decreased', 'elevated', 'elevated', 'hypokalemia', 'hypouricemia', 'increased', 'increased', 'increased', 'increased uptake', 'lesion', 'pain', 'pain', 'pain', 'recovered', 'weakness'] \n",
      "Match: ['compression', 'hypokalemia', 'hypouricemia', 'lesion', 'pain', 'pain', 'recovered'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(compare)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A 16-year-old female suffered from an abdominal pain at right upper quadrant, lasting for more than 10 days.\\nShortly ahead of her medical consultation, 4 ascaris-like worms were vomited out, individually with a length of about 10\\u200acm.\\nShe had a previous onset of ascariasis, during which low-grade fever (38.4°C) occurred without apparent jaundice, diarrhea, or anemia.\\nPhysical examination revealed tenderness at right upper quadrant.\\nHer total leukocyte count was 11.2\\u200aG/L consisting of 5.2% eosinophils.\\nSerum and urine amylase were 386 and 928\\u200aU/L, respectively.\\nIn terms of liver functionality, the level of total bilirubin rose up to 23.2\\u200aμm/L, meanwhile the hepatic enzymes were similarly elevated (alanine aminotransferase 163\\u200aU/L; aspartate aminotransferase 96\\u200aU/L).\\nAbdominal ultrasound described the enlargement of the gallbladder, upper segment of common bile duct (1.5\\u200acm in diameter) and intrahepatic bile duct (1.3\\u200acm in diameter).\\nFurthermore, the intrahepatic bile duct was also discovered to be filled with echogenic and banded substances.\\nBased on such evidences, she was clinically diagnosed as biliary ascariasis with a concomitant manifestation of biliary duct infection.\\nThrough a laparoscopic exploration, the enlarged gallbladder and common bile duct were visually confirmed, as well as the massive storage of worms within the biliary tract.\\nSubsequently, a 1\\u200acm incision was made on the common bile duct in order to facilitate the extraction of the pathogenic worms, which were then placed into a specimen bag laparoscopically (Fig.1).\\nOnce the contained worms were virtually cleared out under laparoscope, chodedochoscopy was employed to seek for more hidden worms inside the biliary system (Fig.2).\\nAfter finishing the removal of ascarides, the biliary duct incision was well sutured, with a T-tube placed in situ (Fig.3).\\nEventually, a standard laparoscopic cholecystectomy was performed, followed by the removal of specimen bags and drainage tube placement (Fig.4).\\nThe total amount of worms inside the common bile duct summed up to more than 100 (Fig.5).\\nPostoperatively, she was treated with oral antihelminthic drugs for 3 days.\\nThe T-tube was rinsed with saline every day, and during the first 3 days, 6 more worms were additionally washed out.\\nThe patient experienced a smooth postoperative recovery and was finally discharged from hospital without any complications.\\n'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "F1score = defaultdict(list)\n",
    "compare = \"\"\n",
    "\n",
    "for instance in range(len(test_dataset['input'][:1])):\n",
    "    #for x in json.loads(test_dataset['output'][instance]):\n",
    "    #    print(x)\n",
    "    prompt = formatting_input_func(test_dataset[instance])\n",
    "    compare = compare + test_dataset[instance][\"input\"] + \"\\n\\n\"\n",
    "    #print(prompt)\n",
    "\n",
    "    inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "    outputs = model.generate(input_ids=inputs, max_length=2000)[0]\n",
    "\n",
    "    answer_start = int(inputs.shape[-1])\n",
    "    pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "    pred = pred.replace(\"\\'\", \"\\\"\")\n",
    "    sep = \"}\"\n",
    "    pred = pred.split(sep, 1)[0] + sep\n",
    "    pred = pred.replace(\"\\\\\",\" \")\n",
    " \n",
    "    try:\n",
    "        json_pred = json.loads(pred)\n",
    "    except:\n",
    "        print(\"Something went wrong with json.loads(pred)\")\n",
    "    else:\n",
    "        #json_pred = json.loads(pred)\n",
    "        json_pred = json.loads(pred)\n",
    "        \n",
    "        precision = 0\n",
    "        recall = 0\n",
    "        #naredi dict listov za vsak atrubut posebej F1\n",
    "        for x in json.loads(test_dataset['output'][instance]):\n",
    "\n",
    "            if x not in json_pred:\n",
    "                continue\n",
    "\n",
    "            real = json.loads(test_dataset['output'][instance])[x]\n",
    "            gene = json_pred[x]\n",
    "\n",
    "            match = [r for r in real if r in gene]\n",
    "\n",
    "\n",
    "            lenmatch = len(match)\n",
    "            lenreal = len(real)\n",
    "            lengene = len(gene)\n",
    "            \n",
    "            text = f\"{x} \\nReal: {json.loads(test_dataset['output'][instance])[x]} \\nGene: {json_pred[x]} \\nMatch: {match} \\n\\n\"\n",
    "            compare = compare + text\n",
    "\n",
    "            precision = lenmatch/lengene\n",
    "            recall = lenmatch/lenreal\n",
    "            f1 = 2*(precision*recall) / (precision+recall)\n",
    "            #print(precision, recall, f1)\n",
    "\n",
    "            F1score[x].append(f1) \n",
    "            \n",
    "            \n",
    "        compare = compare + \"\\n\\n\\n\"\n",
    "    finally:\n",
    "        continue\n",
    "    #print(pred)\n",
    "    #print(test_dataset['output'][instance])\n",
    "\n",
    "    for x in test_dataset['output'][instance]:\n",
    "        text = f\"{x} \\nReal: {test_dataset['output'][instance][x]} \\nGene: {json_pred[x]} \\n\\n\"\n",
    "        compare = compare + text\n",
    "    compare = compare + \"\\n\\n\\n\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[78], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m#print(prompt)\u001b[39;00m\n\u001b[1;32m     11\u001b[0m inputs \u001b[38;5;241m=\u001b[39m tokenizer(prompt, return_tensors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpt\u001b[39m\u001b[38;5;124m\"\u001b[39m, add_special_tokens\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\u001b[38;5;241m.\u001b[39minput_ids\u001b[38;5;241m.\u001b[39mto(model\u001b[38;5;241m.\u001b[39mdevice)\n\u001b[0;32m---> 12\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2000\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     14\u001b[0m answer_start \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(inputs\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n\u001b[1;32m     15\u001b[0m pred \u001b[38;5;241m=\u001b[39m tokenizer\u001b[38;5;241m.\u001b[39mdecode(outputs[answer_start:], skip_special_tokens\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/peft/peft_model.py:1130\u001b[0m, in \u001b[0;36mPeftModelForCausalLM.generate\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m   1128\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbase_model\u001b[38;5;241m.\u001b[39mgeneration_config \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgeneration_config\n\u001b[1;32m   1129\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1130\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbase_model\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1131\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m   1132\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbase_model\u001b[38;5;241m.\u001b[39mprepare_inputs_for_generation \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbase_model_prepare_inputs_for_generation\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/utils/_contextlib.py:115\u001b[0m, in \u001b[0;36mcontext_decorator.<locals>.decorate_context\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m    113\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdecorate_context\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    114\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m ctx_factory():\n\u001b[0;32m--> 115\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/utils.py:1479\u001b[0m, in \u001b[0;36mGenerationMixin.generate\u001b[0;34m(self, inputs, generation_config, logits_processor, stopping_criteria, prefix_allowed_tokens_fn, synced_gpus, assistant_model, streamer, negative_prompt_ids, negative_prompt_attention_mask, **kwargs)\u001b[0m\n\u001b[1;32m   1462\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39massisted_decoding(\n\u001b[1;32m   1463\u001b[0m         input_ids,\n\u001b[1;32m   1464\u001b[0m         candidate_generator\u001b[38;5;241m=\u001b[39mcandidate_generator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1475\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mmodel_kwargs,\n\u001b[1;32m   1476\u001b[0m     )\n\u001b[1;32m   1477\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m generation_mode \u001b[38;5;241m==\u001b[39m GenerationMode\u001b[38;5;241m.\u001b[39mGREEDY_SEARCH:\n\u001b[1;32m   1478\u001b[0m     \u001b[38;5;66;03m# 11. run greedy search\u001b[39;00m\n\u001b[0;32m-> 1479\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgreedy_search\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1480\u001b[0m \u001b[43m        \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1481\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlogits_processor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprepared_logits_processor\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1482\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstopping_criteria\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprepared_stopping_criteria\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1483\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpad_token_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpad_token_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1484\u001b[0m \u001b[43m        \u001b[49m\u001b[43meos_token_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meos_token_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1485\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_scores\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moutput_scores\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1486\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_dict_in_generate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgeneration_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreturn_dict_in_generate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1487\u001b[0m \u001b[43m        \u001b[49m\u001b[43msynced_gpus\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msynced_gpus\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1488\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstreamer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstreamer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1489\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmodel_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1490\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1492\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m generation_mode \u001b[38;5;241m==\u001b[39m GenerationMode\u001b[38;5;241m.\u001b[39mCONTRASTIVE_SEARCH:\n\u001b[1;32m   1493\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m model_kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muse_cache\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/utils.py:2340\u001b[0m, in \u001b[0;36mGenerationMixin.greedy_search\u001b[0;34m(self, input_ids, logits_processor, stopping_criteria, max_length, pad_token_id, eos_token_id, output_attentions, output_hidden_states, output_scores, return_dict_in_generate, synced_gpus, streamer, **model_kwargs)\u001b[0m\n\u001b[1;32m   2337\u001b[0m model_inputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprepare_inputs_for_generation(input_ids, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mmodel_kwargs)\n\u001b[1;32m   2339\u001b[0m \u001b[38;5;66;03m# forward pass to get next token\u001b[39;00m\n\u001b[0;32m-> 2340\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2341\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mmodel_inputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2342\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   2343\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2344\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2345\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2347\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m synced_gpus \u001b[38;5;129;01mand\u001b[39;00m this_peer_finished:\n\u001b[1;32m   2348\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m  \u001b[38;5;66;03m# don't waste resources running the code we don't need\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/accelerate/hooks.py:165\u001b[0m, in \u001b[0;36madd_hook_to_module.<locals>.new_forward\u001b[0;34m(module, *args, **kwargs)\u001b[0m\n\u001b[1;32m    163\u001b[0m         output \u001b[38;5;241m=\u001b[39m module\u001b[38;5;241m.\u001b[39m_old_forward(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    164\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 165\u001b[0m     output \u001b[38;5;241m=\u001b[39m \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_old_forward\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m module\u001b[38;5;241m.\u001b[39m_hf_hook\u001b[38;5;241m.\u001b[39mpost_forward(module, output)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:1183\u001b[0m, in \u001b[0;36mLlamaForCausalLM.forward\u001b[0;34m(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, labels, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1180\u001b[0m return_dict \u001b[38;5;241m=\u001b[39m return_dict \u001b[38;5;28;01mif\u001b[39;00m return_dict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39muse_return_dict\n\u001b[1;32m   1182\u001b[0m \u001b[38;5;66;03m# decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)\u001b[39;00m\n\u001b[0;32m-> 1183\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1184\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1185\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1186\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1187\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpast_key_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1188\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs_embeds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs_embeds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1189\u001b[0m \u001b[43m    \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1190\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1191\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1192\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1193\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1195\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m outputs[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1196\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mpretraining_tp \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/accelerate/hooks.py:165\u001b[0m, in \u001b[0;36madd_hook_to_module.<locals>.new_forward\u001b[0;34m(module, *args, **kwargs)\u001b[0m\n\u001b[1;32m    163\u001b[0m         output \u001b[38;5;241m=\u001b[39m module\u001b[38;5;241m.\u001b[39m_old_forward(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    164\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 165\u001b[0m     output \u001b[38;5;241m=\u001b[39m \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_old_forward\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m module\u001b[38;5;241m.\u001b[39m_hf_hook\u001b[38;5;241m.\u001b[39mpost_forward(module, output)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:1070\u001b[0m, in \u001b[0;36mLlamaModel.forward\u001b[0;34m(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1060\u001b[0m     layer_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_gradient_checkpointing_func(\n\u001b[1;32m   1061\u001b[0m         decoder_layer\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__call__\u001b[39m,\n\u001b[1;32m   1062\u001b[0m         hidden_states,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1067\u001b[0m         use_cache,\n\u001b[1;32m   1068\u001b[0m     )\n\u001b[1;32m   1069\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1070\u001b[0m     layer_outputs \u001b[38;5;241m=\u001b[39m \u001b[43mdecoder_layer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1071\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1072\u001b[0m \u001b[43m        \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1073\u001b[0m \u001b[43m        \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1074\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpast_key_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1075\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1076\u001b[0m \u001b[43m        \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1077\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1079\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m layer_outputs[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1081\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m use_cache:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/accelerate/hooks.py:165\u001b[0m, in \u001b[0;36madd_hook_to_module.<locals>.new_forward\u001b[0;34m(module, *args, **kwargs)\u001b[0m\n\u001b[1;32m    163\u001b[0m         output \u001b[38;5;241m=\u001b[39m module\u001b[38;5;241m.\u001b[39m_old_forward(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    164\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 165\u001b[0m     output \u001b[38;5;241m=\u001b[39m \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_old_forward\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m module\u001b[38;5;241m.\u001b[39m_hf_hook\u001b[38;5;241m.\u001b[39mpost_forward(module, output)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:798\u001b[0m, in \u001b[0;36mLlamaDecoderLayer.forward\u001b[0;34m(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache, **kwargs)\u001b[0m\n\u001b[1;32m    795\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minput_layernorm(hidden_states)\n\u001b[1;32m    797\u001b[0m \u001b[38;5;66;03m# Self Attention\u001b[39;00m\n\u001b[0;32m--> 798\u001b[0m hidden_states, self_attn_weights, present_key_value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mself_attn\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpast_key_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    803\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[43m    \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    805\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    806\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    807\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m residual \u001b[38;5;241m+\u001b[39m hidden_states\n\u001b[1;32m    809\u001b[0m \u001b[38;5;66;03m# Fully Connected\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/accelerate/hooks.py:165\u001b[0m, in \u001b[0;36madd_hook_to_module.<locals>.new_forward\u001b[0;34m(module, *args, **kwargs)\u001b[0m\n\u001b[1;32m    163\u001b[0m         output \u001b[38;5;241m=\u001b[39m module\u001b[38;5;241m.\u001b[39m_old_forward(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    164\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 165\u001b[0m     output \u001b[38;5;241m=\u001b[39m \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_old_forward\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m module\u001b[38;5;241m.\u001b[39m_hf_hook\u001b[38;5;241m.\u001b[39mpost_forward(module, output)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:693\u001b[0m, in \u001b[0;36mLlamaSdpaAttention.forward\u001b[0;34m(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache)\u001b[0m\n\u001b[1;32m    682\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mforward(\n\u001b[1;32m    683\u001b[0m         hidden_states\u001b[38;5;241m=\u001b[39mhidden_states,\n\u001b[1;32m    684\u001b[0m         attention_mask\u001b[38;5;241m=\u001b[39mattention_mask,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    688\u001b[0m         use_cache\u001b[38;5;241m=\u001b[39muse_cache,\n\u001b[1;32m    689\u001b[0m     )\n\u001b[1;32m    691\u001b[0m bsz, q_len, _ \u001b[38;5;241m=\u001b[39m hidden_states\u001b[38;5;241m.\u001b[39msize()\n\u001b[0;32m--> 693\u001b[0m query_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mq_proj\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    694\u001b[0m key_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mk_proj(hidden_states)\n\u001b[1;32m    695\u001b[0m value_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mv_proj(hidden_states)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/peft/tuners/lora/bnb.py:309\u001b[0m, in \u001b[0;36mLinear4bit.forward\u001b[0;34m(self, x, *args, **kwargs)\u001b[0m\n\u001b[1;32m    306\u001b[0m     expected_dtype \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39mdtype\n\u001b[1;32m    307\u001b[0m     x \u001b[38;5;241m=\u001b[39m x\u001b[38;5;241m.\u001b[39mto(lora_A\u001b[38;5;241m.\u001b[39mweight\u001b[38;5;241m.\u001b[39mdtype)\n\u001b[0;32m--> 309\u001b[0m output \u001b[38;5;241m=\u001b[39m lora_B(lora_A(\u001b[43mdropout\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m))\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_conversion:\n\u001b[1;32m    311\u001b[0m     output \u001b[38;5;241m=\u001b[39m output\u001b[38;5;241m.\u001b[39mto(expected_dtype)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1521\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1520\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_call_impl\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m-> 1521\u001b[0m     forward_call \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_slow_forward \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_C\u001b[49m\u001b[38;5;241m.\u001b[39m_get_tracing_state() \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mforward)\n\u001b[1;32m   1522\u001b[0m     \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m     \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m             \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m             \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "F1score = defaultdict(list)\n",
    "compare = \"\"\n",
    "\n",
    "for instance in range(len(test_dataset['input'][:1])):\n",
    "    prompt = formatting_input_func(test_dataset[instance])\n",
    "    compare = compare + test_dataset[instance][\"input\"] + \"\\n\\n\"\n",
    "    #print(prompt)\n",
    "\n",
    "    inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "    outputs = model.generate(input_ids=inputs, max_length=2000)[0]\n",
    "\n",
    "    answer_start = int(inputs.shape[-1])\n",
    "    pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "\n",
    "    print(pred)\n",
    "    pred = pred.replace(\"\\'\", \"\\\"\")\n",
    "    sep = \"}\"\n",
    "    pred = pred.split(sep, 1)[0] + sep\n",
    "    pred = pred.replace(\"\\\\\",\" \")\n",
    " \n",
    "    try:\n",
    "        json_pred = json.loads(pred)\n",
    "    except:\n",
    "        print(\"Something went wrong with json.loads(pred)\")\n",
    "    else:\n",
    "        #json_pred = json.loads(pred)\n",
    "        json_pred = json.loads(pred)\n",
    "        \n",
    "        precision = 0\n",
    "        recall = 0\n",
    "        #naredi dict listov za vsak atrubut posebej F1\n",
    "        for x in json_pred['output'][instance]:\n",
    "            print(x)\n",
    "        \n",
    "            real = test_dataset['output'][instance][x]\n",
    "            gene = json_pred[x]\n",
    "\n",
    "            match = [r for r in real if r in gene]\n",
    "\n",
    "            lenmatch = len(match)\n",
    "            lenreal = len(real)\n",
    "            lengene = len(gene)\n",
    "            \n",
    "            text = f\"{x} \\nReal: {test_dataset['output'][instance][x]} \\nGene: {json_pred[x]} \\nMatch: {match} \\n\\n\"\n",
    "            compare = compare + text\n",
    "\n",
    "            precision = lenmatch/lengene\n",
    "            recall = lenmatch/lenreal\n",
    "            f1 = 2*(precision*recall) / (precision+recall)\n",
    "            #print(precision, recall, f1)\n",
    "\n",
    "            F1score[x].append(f1) \n",
    "            \n",
    "            \n",
    "        compare = compare + \"\\n\\n\\n\"\n",
    "    finally:\n",
    "        continue\n",
    "    #print(pred)\n",
    "    #print(test_dataset['output'][instance])\n",
    "\n",
    "    for x in test_dataset['output'][instance]:\n",
    "        text = f\"{x} \\nReal: {test_dataset['output'][instance][x]} \\nGene: {json_pred[x]} \\n\\n\"\n",
    "        compare = compare + text\n",
    "    compare = compare + \"\\n\\n\\n\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#F1score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"Diagnostic_procedure\": [\"Computerised perimetry\", \"IOPs\", \"MRI\", \"T1 and T2 signals\", \"Visual acuities\", \"acuities\", \"acuities\", \"blood analysis\", \"colour vision\", \"eye movements\", \"false-negative error rates\", \"field test\", \"inflammatory markers\", \"intraocular pressures\", \"perimetry\", \"pupils\", \"scanning\", \"visual acuity\", \"visual acuity\", \"visual acuity\", \"visual field examination\", \"visual field test\", \"visual fields\", \"visual fields\"]}\n"
     ]
    }
   ],
   "source": [
    "print(pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Diagnostic_procedure': ['Computerised perimetry',\n",
       "  'IOPs',\n",
       "  'MRI',\n",
       "  'T1 and T2 signals',\n",
       "  'Visual acuities',\n",
       "  'acuities',\n",
       "  'acuities',\n",
       "  'blood analysis',\n",
       "  'colour vision',\n",
       "  'eye movements',\n",
       "  'false-negative error rates',\n",
       "  'field test',\n",
       "  'inflammatory markers',\n",
       "  'intraocular pressures',\n",
       "  'perimetry',\n",
       "  'pupils',\n",
       "  'scanning',\n",
       "  'visual acuity',\n",
       "  'visual acuity',\n",
       "  'visual acuity',\n",
       "  'visual field examination',\n",
       "  'visual field test',\n",
       "  'visual fields',\n",
       "  'visual fields']}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{ \"Sex\": [\"man\"], \"Biological_structure\": [\"both optic nerves\", \"head\", \"infrasellar sphenoid sinuses\", \"pituitary fossa\", \"right eye\", \"suprasellar cistern\"], \"Sign_symptom\": [\"cupped\", \"cupping\", \"displaced\", \"deteriorate\", \"deterioration\", \"fluid levels\", \"heterogeneous\", \"hypertension\", \"hypercholesterolaemia\", \"increased\", \"lethargy\", \"mass\", \"mass\", \"multiple\", \"optic disc cupping\", \"prolactin levels\", \"recovery\", \"risk factors\", \"symptoms\", \"vasculopathy\"], \"Diagnostic_procedure\": [\"Clinical examination\", \"Computerised perimetry\", \"MRI\", \"T1 and T2 signals\", \"Visual acuities\", \"acuities\", \"acuities\", \"blood analysis\", \"colour vision\", \"eye movements\", \"false-negative error rates\", \"field test\", \"inflammatory markers\", \"intraocular pressures\", \"pupils\", \"visual acuity\", \"visual acuity\", \"visual acuity\", \"visual field examination\", \"visual field test\", \"visual fields\", \"visual fields\"], \"Detailed_description\": [\"advanced\", \"amblyopic\", \"bilateral\", \"compressed\", \"considerably raised\", \"depressed\", \"equal and reactive\", \"false-negative\", \"giant\", \"heterogeneous\", \"hypertension\", \"increased\", \"low-signal components\", \"multiple\", \"normal\", \"normal\", \"reactive\", \"superiorly\", \"type 2\", \"urgent\", \"with testing compounded\"], \"Lab_value\": [\"12\", \"18\", \"18\", \"18\", \"200 000 mIU/L\", \"336 mIU/L\", \"30%\", \"4.8\", \"50 pack-years\", \"6/18\", \"6/18\", \"6/18\", \"6/18\", \"6/5\", \"6/6\", \"6/7.5 N6\", \"6/9\", \"6/9\", \"6/18 N12\", \"6/18 RE\", \"6/18 RE\", \"6/24\", \"6/5\", \"6/7.5\", \"80%\", \"8 mg\", \"decreased\", \"greater than 200 000 mIU/L\", \"loss of one Snellen line\", \"loss of two Snellen lines\", \"normal\", \"raised\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\", \"similar\",}\n"
     ]
    }
   ],
   "source": [
    "sep = \"}\"\n",
    "pred.split(sep, 1)[0] + sep\n",
    "print(\"{\" + pred.split(sep, 1)[1] + \"}\")\n",
    "\n",
    "#haha = json.loads(pred)\n",
    "#haha[\"Age\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 68-year-old man was referred by his optometrist to HES with suspected LTG due to repeatedly irregular visual field test results, advanced optic disc cupping, normal intraocular pressures (IOPs) and a family history of glaucoma.\n",
      "The patient subjectively felt that vision in his ‘good’ left eye (LE), which normally had a visual acuity of 6/6 N5, started to deteriorate 6 months earlier; at the point of referral it was best corrected to 6/7.5 N6.\n",
      "His right eye (RE) was known to be amblyopic with a visual acuity of 6/18 N12.\n",
      "His medical history included considerable risk factors for systemic vasculopathy, such as hypertension, hypercholesterolaemia, 50 pack-years of smoking and type 2 diabetes with no diabetic retinopathy.\n",
      "Despite detailed questioning, he denied any new systemic symptoms apart from experiencing increased lethargy.\n",
      "Clinical examination at HES revealed advanced bilateral cupped optic discs with a cup-to-disc ratio right 0.9 (90%) and left 0.8 (80%; figure 1).\n",
      "Furthermore, the findings from the referring optometrist of equal and reactive pupils, normal eye movements and IOPs in mid-teens were confirmed.\n",
      "Computerised perimetry showed constricted visual fields in both eyes, albeit with considerably raised false-negative error rates in excess of 30% (figure 2A).\n",
      "Since this can be a result of poor concentration and attention span with testing compounded, especially with an amblyopic RE, the field test was repeated 4 weeks later with similar results (figure 2B).\n",
      "At this visit, the patient experienced further deterioration of his ‘good’ LE.\n",
      "Visual acuities measured 6/24 (loss of one Snellen line) N36 in his amblyopic RE, and 6/18 (loss of two Snellen lines) N36 in his LE, which now also showed depressed colour vision.\n",
      "An urgent scan of the patient's head and orbit was requested to rule out a compressive lesion and blood analysis was carried out to screen for other causes of optic neuropathy including vitamin B12, folate levels and inflammatory markers.\n",
      "MRI revealed a large mass centred at the pituitary fossa measuring approximately 4.5 cm×5.1 cm× 4.8 cm (figure 3A–C) extending into the suprasellar cistern and infrasellar sphenoid sinuses.\n",
      "The optic chiasm and both optic nerves were displaced superiorly and compressed on both sides.\n",
      "The mass showed heterogeneous T1 and T2 signals with multiple fluid levels most likely due to previous haemorrhage and low-signal components suggesting calcifications.\n",
      "The appearance was probably due to a craniopharyngioma or haemorrhagic pituitary adenoma.\n",
      "The patient was referred urgently to the neurosurgery unit and was found to have markedly raised prolactin levels of greater than 200 000 mIU/L, confirming the diagnosis of a giant prolactinoma.\n",
      "In cooperation with the local endocrinology service, he was started on cabergoline 500 µg and dexamethasone 8 mg.\n",
      "Within a week this led to a dramatic improvement of the patient's visual fields and acuities to 6/18 RE and to 6/9 LE.\n",
      "His prolactin levels showed a similar dramatic response with levels coming down to 336 mIU/L.\n",
      "Three months later his visual acuity in his RE and LE were 6/18 and 6/5, respectively, and visual field examination showed striking recovery with almost normal results (figure 2C).\n",
      "\n",
      "\n",
      "Lab_value \n",
      "Real: ['0.8', '0.9', '336\\u2005mIU/L', '50 pack-years', '6/18', '6/18', '6/18 N12', '6/5', '6/6 N5', '6/7.5 N6', '6/9', '80%', '90%', 'both', 'both', 'considerably raised', 'depressed', 'dramatic improvement', 'dramatic response with levels coming down', 'further deterioration', 'greater than 200\\u2005000\\u2005mIU/L', 'in excess of 30%', 'increased', 'markedly raised', 'multiple', 'normal', 'normal'] \n",
      "Gene: ['0.8', '0.9', '12', '18', '18', '18', '200\\u2005000\\u2005mIU/L', '336\\u2005mIU/L', '30%', '6/18', '6/18', '6/18', '6/5', '6/6', '6/6 N5', '6/7.5 N6', '6/9', '6/9', '6/18 N12', '6/18 RE', '6/24', '6/9 LE', '6/9 RE', 'depressed', 'equal', 'false-negative', 'greater than', 'loss of one Snellen line', 'loss of two Snellen lines', 'normal', 'normal', 'normal', 'raised', 'raised', 'reactive', 'similar', 'similar', 'similar', 'superiorly', 'urgent'] \n",
      "Match: ['0.8', '0.9', '336\\u2005mIU/L', '6/18', '6/18', '6/18 N12', '6/5', '6/6 N5', '6/7.5 N6', '6/9', 'depressed', 'normal', 'normal'] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(compare)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"../results/comparePerEntity2.txt\", \"w\")\n",
    "f.write(compare)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Age\n",
      "Biological_structure\n",
      "Detailed_description\n",
      "Diagnostic_procedure\n",
      "Lab_value\n",
      "Sex\n",
      "Sign_symptom\n"
     ]
    }
   ],
   "source": [
    "for x in test_dataset['output'][0]:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['36-yr-old']"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dataset['output'][0][\"Age\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'dict' object has no attribute 'replace'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[61], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mtest_dataset\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43moutput\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreplace\u001b[49m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\'\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\\"\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'dict' object has no attribute 'replace'"
     ]
    }
   ],
   "source": [
    "test_dataset['output'][0].replace(\"\\'\", \"\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\n"
     ]
    }
   ],
   "source": [
    "prompt = \"\"\"\n",
    "### Question:\n",
    "A 58-year-old man had been suffering from general fatigue and severe anemia for several months.\n",
    "\n",
    "### Answer:\n",
    "\"\"\"\n",
    "\n",
    "input_ids = tokenizer(prompt, return_tensors=\"pt\", truncation=True).input_ids.cuda()\n",
    "outputs = model.generate(input_ids=input_ids, max_new_tokens=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated Response:\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"Generated Response:\\n{tokenizer.batch_decode(outputs.detach().cpu().numpy(), skip_special_tokens=False)[0][len(prompt):]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.llms import HuggingFacePipeline\n",
    "import torch\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline, AutoModelForSeq2SeqLM\n",
    "from kor import create_extraction_chain, Object, Text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The model 'PeftModelForCausalLM' is not supported for text-generation. Supported models are ['BartForCausalLM', 'BertLMHeadModel', 'BertGenerationDecoder', 'BigBirdForCausalLM', 'BigBirdPegasusForCausalLM', 'BioGptForCausalLM', 'BlenderbotForCausalLM', 'BlenderbotSmallForCausalLM', 'BloomForCausalLM', 'CamembertForCausalLM', 'LlamaForCausalLM', 'CodeGenForCausalLM', 'CpmAntForCausalLM', 'CTRLLMHeadModel', 'Data2VecTextForCausalLM', 'ElectraForCausalLM', 'ErnieForCausalLM', 'FalconForCausalLM', 'FuyuForCausalLM', 'GitForCausalLM', 'GPT2LMHeadModel', 'GPT2LMHeadModel', 'GPTBigCodeForCausalLM', 'GPTNeoForCausalLM', 'GPTNeoXForCausalLM', 'GPTNeoXJapaneseForCausalLM', 'GPTJForCausalLM', 'LlamaForCausalLM', 'MarianForCausalLM', 'MBartForCausalLM', 'MegaForCausalLM', 'MegatronBertForCausalLM', 'MistralForCausalLM', 'MixtralForCausalLM', 'MptForCausalLM', 'MusicgenForCausalLM', 'MvpForCausalLM', 'OpenLlamaForCausalLM', 'OpenAIGPTLMHeadModel', 'OPTForCausalLM', 'PegasusForCausalLM', 'PersimmonForCausalLM', 'PhiForCausalLM', 'PLBartForCausalLM', 'ProphetNetForCausalLM', 'QDQBertLMHeadModel', 'Qwen2ForCausalLM', 'ReformerModelWithLMHead', 'RemBertForCausalLM', 'RobertaForCausalLM', 'RobertaPreLayerNormForCausalLM', 'RoCBertForCausalLM', 'RoFormerForCausalLM', 'RwkvForCausalLM', 'Speech2Text2ForCausalLM', 'TransfoXLLMHeadModel', 'TrOCRForCausalLM', 'WhisperForCausalLM', 'XGLMForCausalLM', 'XLMWithLMHeadModel', 'XLMProphetNetForCausalLM', 'XLMRobertaForCausalLM', 'XLMRobertaXLForCausalLM', 'XLNetLMHeadModel', 'XmodForCausalLM'].\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:392: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:397: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'data': {},\n",
       " 'raw': \" {'patient': ['A 32-year-old man', 'Emily Johnson', '42-year-old woman', 'a 32-year-old man', 'a 50-year-old woman','man', 'woman'], 'age': ['32-year-old', 'escalating', 'persistent'], 'biological structure': ['abdomen','abdomen','abdominal','abdominal','chest','chest','gallbladder','gallbladder','left arm','left upper quadrant','upper right quadrant'], 'diagnostic procedure': ['CT','Blood levels of lactate and pyruvate','Discomfort','Heart rate','Physical examination','Blood levels of lactate and pyruvate','Computed tomography','diagnostic ultrasound','electrocardiogram','lactate','pyruvate','physical examination','sinus rhythm','ultrasound'], 'lab value': ['0.096\\\\u2009mmol/l','0.096\\\\u2009mmol/l','1.6','1.6','130/minute','38.4 °C','38.4 °C','unremarkable findings'],'sign symptom': ['discomfort','discomfort','escalating','gallstones','lost strength','nausea','pain','pain','persistent','sporadic','tumor-free','vomiting']}\",\n",
       " 'errors': [],\n",
       " 'validated_data': {}}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain.prompts import PromptTemplate\n",
    "\n",
    "pipe = pipeline(\n",
    "    \"text-generation\",\n",
    "    model=model, \n",
    "    tokenizer=tokenizer, \n",
    "    max_length=2048,\n",
    "    #do_sample=True,\n",
    "    #temperature=0.0001,\n",
    "    #max_new_tokens=300\n",
    ")\n",
    "\n",
    "local_llm = HuggingFacePipeline(pipeline=pipe)\n",
    "\n",
    "#[INST] <<SYS>>\\n<</SYS>> [/INST]\n",
    "\n",
    "instruction_template = PromptTemplate(\n",
    "    input_variables=[\"format_instructions\", \"type_description\"],\n",
    "    template=(\n",
    "        \"[INST]\\n Your goal is to extract structured information from the user's input that matches the form described below. When extracting information please make sure it matches the type information exactly. Do not add any attributes that do not appear in the schema shown below.\\n\\n\"\n",
    "        \"{type_description}\\n\\n\" # Can comment out\n",
    "        \"{format_instructions}\\n\\n\"\n",
    "        #\"Here are some examples:\\n\\n\"\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "schema = Object(\n",
    "    id=\"patient\",\n",
    "    description=(\n",
    "        \"A patient is an individual admitted to or seeking medical attention\"\n",
    "        \" at a hospital or healthcare facility.\"\n",
    "    ),\n",
    "    attributes=[\n",
    "        Text(\n",
    "            id=\"age\",\n",
    "            description=\"Patients' age.\",\n",
    "            examples=[\n",
    "                (\"A 32-year-old man\", \"32-year-old\")\n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"sex\",\n",
    "            description=\"Patients' sex.\",\n",
    "            examples=[\n",
    "                (\"A 32-year-old man.\", \"Male\"),\n",
    "                (\"A 50-year-old woman.\", \"Female\")\n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"sign symptom\",\n",
    "            description=\"Sign symptom are crucial in the diagnostic process as they provide valuable information for understanding a patient's health concerns.\",\n",
    "            examples=[\n",
    "                  (\"The patient has a fever temperature of 38.\", \"fever\"),\n",
    "                  (\"The patient complains of abdominal pain.\", \"abdominal pain\")\n",
    "            ],\n",
    "              many=True,\n",
    "          ),\n",
    "         Text(\n",
    "             id=\"biological structure\",\n",
    "             description=\"Biological structures refer to the anatomical components and organs.\",\n",
    "             examples=[\n",
    "                 (\"The patient lost strength in the left arm.\", \"arm\"),\n",
    "                 (\"Computed tomography (CT) of the chest.\", \"chest\"),\n",
    "                 (\"Discomfort in the left upper quadrant of the abdomen.\", \"abdomen\"),\n",
    "             ],\n",
    "             many=True,\n",
    "         ),\n",
    "        Text(\n",
    "            id=\"diagnostic procedure\",\n",
    "            description=\"A diagnostic procedure is a medical test or examination.\",\n",
    "            examples=[\n",
    "                (\"An electrocardiogram revealed normal sinus rhythm\", \"electrocardiogram\"),\n",
    "                (\"Physical examination yielded unremarkable findings.\", \"physical examination\"),\n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "        Text(\n",
    "            id=\"lab value\",\n",
    "            description=\"Laboratory values refer to the results of tests.\",\n",
    "            examples=[\n",
    "                (\"Surgical margins were tumor-free.\", \"tumor-free\"),\n",
    "                (\"His temperature was 38.4 °C.\", \"38.4 °C\"),\n",
    "                (\"Heart rate 130/minute\", \"130/minute\"),\n",
    "                (\"Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\", \"0.096 mmol/l\"), \n",
    "                (\"Blood levels of lactate and pyruvate were 1.6 and 0.096 mmol/l\", \"1.6\"), \n",
    "            ],\n",
    "            many=True,\n",
    "        ),\n",
    "    ],\n",
    "    many=False,\n",
    ")\n",
    "\n",
    "user_input = \"\"\"\\n\n",
    "Emily Johnson, a 42-year-old woman, visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month. \n",
    "Alongside the discomfort, she experienced sporadic nausea and vomiting. \n",
    "During the examination, a diagnostic ultrasound revealed gallstones in the gallbladder, indicative of cholecystitis. \n",
    "[/INST]\"\"\"\n",
    "\n",
    "prompt = f\"{user_input}\"\n",
    "\n",
    "#prompt = user_input\n",
    "\n",
    "chain = create_extraction_chain(local_llm, schema, instruction_template=instruction_template, encoder_or_encoder_class='json')\n",
    "chain.run(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Guidance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f07b646b3f07400e867c44a1cd5c6012",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n",
      "  return self.fget.__get__(instance, owner)()\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:392: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:397: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from peft import PeftConfig, PeftModel\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
    "import json\n",
    "import tqdm\n",
    "from guidance import models\n",
    "\n",
    "device = \"cuda\"\n",
    "\n",
    "base_model_name = \"AdaptLLM/medicine-chat\" #path/to/your/model/or/name/on/hub\"\n",
    "adapter_model_name = \"../models/AdaptLLM/medicine-chat-Erik-shuffled\"\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(base_model_name, load_in_4bit=True)\n",
    "mode l = PeftModel.from_pretrained(model, adapter_model_name).to(device)\n",
    "#model = model.merge_and_unload()\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(base_model_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HELOOOOOOOOO\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "586bdb3acf4c44b1994093b8b42a829b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n",
      "  return self.fget.__get__(instance, owner)()\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:392: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.9` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n",
      "/home/erikc/code/medical-llm/venv/lib/python3.9/site-packages/transformers/generation/configuration_utils.py:397: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.6` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from guidance import models, gen, select\n",
    "\n",
    "base_model_name = \"AdaptLLM/medicine-chat\" #path/to/your/model/or/name/on/hub\"\n",
    "adapter_model_name = \"../models/AdaptLLM/medicine-chat-Erik-shuffled\"\n",
    "#medicine = models.Transformers(\"AdaptLLM/medicine-chat\")\n",
    "medicine = models.Transformers(model=base_model_name, adapter=adapter_model_name, device=\"cuda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>Question: Luke has ten balls. He gives three to his brother.\n",
       "How many balls does he have left?\n",
       "Answer:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>/</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>]</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> Luke</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> has</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> ten</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> balls</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> to</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> begin</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> with</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>.</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> He</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> gives</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> three</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> balls</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> to</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> his</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> brother</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> which</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> means</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 7\u001b[0m\n\u001b[1;32m      5\u001b[0m lm \u001b[38;5;241m=\u001b[39m medicine \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mQuestion: Luke has ten balls. He gives three to his brother.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      6\u001b[0m lm \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHow many balls does he have left?\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 7\u001b[0m lm \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAnswer: \u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m gen(stop\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/guidance/models/_model.py:302\u001b[0m, in \u001b[0;36mModel.__add__\u001b[0;34m(self, value)\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[38;5;66;03m# run stateless functions (grammar nodes)\u001b[39;00m\n\u001b[1;32m    301\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(value, StatelessFunction):\n\u001b[0;32m--> 302\u001b[0m     out \u001b[38;5;241m=\u001b[39m \u001b[43mlm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_stateless\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    304\u001b[0m \u001b[38;5;66;03m# run stateful functions\u001b[39;00m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    306\u001b[0m     out \u001b[38;5;241m=\u001b[39m value(lm)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/guidance/models/_model.py:465\u001b[0m, in \u001b[0;36mModel._run_stateless\u001b[0;34m(lm, stateless_function, temperature, top_p, n)\u001b[0m\n\u001b[1;32m    463\u001b[0m delayed_bytes \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    464\u001b[0m \u001b[38;5;66;03m# last_is_generated = False\u001b[39;00m\n\u001b[0;32m--> 465\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m new_bytes, is_generated, new_bytes_prob, capture_groups, capture_group_log_probs, new_token_count \u001b[38;5;129;01min\u001b[39;00m gen_obj:\n\u001b[1;32m    466\u001b[0m \n\u001b[1;32m    467\u001b[0m     \u001b[38;5;66;03m# we make everything full probability if we are not computing uncertainty\u001b[39;00m\n\u001b[1;32m    468\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m lm\u001b[38;5;241m.\u001b[39mcompute_log_probs:\n\u001b[1;32m    469\u001b[0m         new_bytes_prob \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1.0\u001b[39m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/guidance/models/_model.py:785\u001b[0m, in \u001b[0;36mModel.__call__\u001b[0;34m(self, grammar, max_tokens, n, top_p, temperature, ensure_bos_token)\u001b[0m\n\u001b[1;32m    783\u001b[0m grammar_temp \u001b[38;5;241m=\u001b[39m parser\u001b[38;5;241m.\u001b[39mnext_byte_temperature()\n\u001b[1;32m    784\u001b[0m current_temp \u001b[38;5;241m=\u001b[39m grammar_temp \u001b[38;5;28;01mif\u001b[39;00m grammar_temp \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m temperature \u001b[38;5;66;03m# we prefer to use the grammar temp when it is specified\u001b[39;00m\n\u001b[0;32m--> 785\u001b[0m logits \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_logits\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtoken_ids\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbytes\u001b[49m\u001b[43m[\u001b[49m\u001b[43mstart_pos\u001b[49m\u001b[43m:\u001b[49m\u001b[43mforced_pos\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcurrent_temp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    786\u001b[0m is_generated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    788\u001b[0m \u001b[38;5;66;03m# if requested we compute the log probabilities so we can track the probabilities of each node\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/guidance/models/transformers/_transformers.py:146\u001b[0m, in \u001b[0;36mTransformers._get_logits\u001b[0;34m(self, token_ids, forced_bytes, current_temp)\u001b[0m\n\u001b[1;32m    144\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(new_token_ids) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    145\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[0;32m--> 146\u001b[0m         model_out \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel_obj\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    147\u001b[0m \u001b[43m            \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtensor\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnew_token_ids\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43munsqueeze\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    148\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpast_key_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cache_state\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpast_key_values\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    149\u001b[0m \u001b[43m            \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    150\u001b[0m \u001b[43m            \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marange\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpast_length\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpast_length\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mnew_token_ids\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43munsqueeze\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    151\u001b[0m \u001b[43m            \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mones\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpast_length\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mnew_token_ids\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    152\u001b[0m \u001b[43m            \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    153\u001b[0m \u001b[43m            \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[43m            \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\n\u001b[1;32m    155\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    157\u001b[0m     \u001b[38;5;66;03m# save the results\u001b[39;00m\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cache_state[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpast_key_values\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m model_out\u001b[38;5;241m.\u001b[39mpast_key_values\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/peft/peft_model.py:1073\u001b[0m, in \u001b[0;36mPeftModelForCausalLM.forward\u001b[0;34m(self, input_ids, attention_mask, inputs_embeds, labels, output_attentions, output_hidden_states, return_dict, task_ids, **kwargs)\u001b[0m\n\u001b[1;32m   1062\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mforward in MPTForCausalLM does not support inputs_embeds\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1063\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbase_model(\n\u001b[1;32m   1064\u001b[0m             input_ids\u001b[38;5;241m=\u001b[39minput_ids,\n\u001b[1;32m   1065\u001b[0m             attention_mask\u001b[38;5;241m=\u001b[39mattention_mask,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1070\u001b[0m             \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   1071\u001b[0m         )\n\u001b[0;32m-> 1073\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbase_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1074\u001b[0m \u001b[43m        \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1075\u001b[0m \u001b[43m        \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1076\u001b[0m \u001b[43m        \u001b[49m\u001b[43minputs_embeds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs_embeds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1077\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlabels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1078\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1079\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1080\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1081\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1082\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1084\u001b[0m batch_size \u001b[38;5;241m=\u001b[39m _get_batch_size(input_ids, inputs_embeds)\n\u001b[1;32m   1085\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m attention_mask \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1086\u001b[0m     \u001b[38;5;66;03m# concat prompt attention mask\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:103\u001b[0m, in \u001b[0;36mBaseTuner.forward\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    102\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mforward\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs: Any, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any):\n\u001b[0;32m--> 103\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mforward\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:1183\u001b[0m, in \u001b[0;36mLlamaForCausalLM.forward\u001b[0;34m(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, labels, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1180\u001b[0m return_dict \u001b[38;5;241m=\u001b[39m return_dict \u001b[38;5;28;01mif\u001b[39;00m return_dict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39muse_return_dict\n\u001b[1;32m   1182\u001b[0m \u001b[38;5;66;03m# decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)\u001b[39;00m\n\u001b[0;32m-> 1183\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1184\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1185\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1186\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1187\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpast_key_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1188\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs_embeds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs_embeds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1189\u001b[0m \u001b[43m    \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1190\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1191\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1192\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1193\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1195\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m outputs[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1196\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mpretraining_tp \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:1070\u001b[0m, in \u001b[0;36mLlamaModel.forward\u001b[0;34m(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1060\u001b[0m     layer_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_gradient_checkpointing_func(\n\u001b[1;32m   1061\u001b[0m         decoder_layer\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__call__\u001b[39m,\n\u001b[1;32m   1062\u001b[0m         hidden_states,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1067\u001b[0m         use_cache,\n\u001b[1;32m   1068\u001b[0m     )\n\u001b[1;32m   1069\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1070\u001b[0m     layer_outputs \u001b[38;5;241m=\u001b[39m \u001b[43mdecoder_layer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1071\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1072\u001b[0m \u001b[43m        \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1073\u001b[0m \u001b[43m        \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1074\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpast_key_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1075\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1076\u001b[0m \u001b[43m        \u001b[49m\u001b[43muse_cache\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_cache\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1077\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1079\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m layer_outputs[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m   1081\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m use_cache:\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:812\u001b[0m, in \u001b[0;36mLlamaDecoderLayer.forward\u001b[0;34m(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache, **kwargs)\u001b[0m\n\u001b[1;32m    810\u001b[0m residual \u001b[38;5;241m=\u001b[39m hidden_states\n\u001b[1;32m    811\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpost_attention_layernorm(hidden_states)\n\u001b[0;32m--> 812\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmlp\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhidden_states\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    813\u001b[0m hidden_states \u001b[38;5;241m=\u001b[39m residual \u001b[38;5;241m+\u001b[39m hidden_states\n\u001b[1;32m    815\u001b[0m outputs \u001b[38;5;241m=\u001b[39m (hidden_states,)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:268\u001b[0m, in \u001b[0;36mLlamaMLP.forward\u001b[0;34m(self, x)\u001b[0m\n\u001b[1;32m    266\u001b[0m     down_proj \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msum\u001b[39m(down_proj)\n\u001b[1;32m    267\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 268\u001b[0m     down_proj \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdown_proj(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mact_fn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgate_proj\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m*\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mup_proj(x))\n\u001b[1;32m    270\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m down_proj\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1518\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1516\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1517\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1518\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/module.py:1527\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1523\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1524\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1525\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1526\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1527\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1530\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/modules/activation.py:393\u001b[0m, in \u001b[0;36mSiLU.forward\u001b[0;34m(self, input)\u001b[0m\n\u001b[1;32m    392\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mforward\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m: Tensor) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Tensor:\n\u001b[0;32m--> 393\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mF\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msilu\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minplace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minplace\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/torch/nn/functional.py:2072\u001b[0m, in \u001b[0;36msilu\u001b[0;34m(input, inplace)\u001b[0m\n\u001b[1;32m   2070\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inplace:\n\u001b[1;32m   2071\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_C\u001b[38;5;241m.\u001b[39m_nn\u001b[38;5;241m.\u001b[39msilu_(\u001b[38;5;28minput\u001b[39m)\n\u001b[0;32m-> 2072\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_C\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_nn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msilu\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "user_input = \"\"\"\\n\n",
    "Emily Johnson, a 42-year-old woman, visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month. Alongside the discomfort, she experienced sporadic nausea and vomiting. During the examination, a diagnostic ultrasound revealed gallstones in the gallbladder, indicative of cholecystitis. \n",
    "\"\"\"\n",
    "\n",
    "lm = medicine + 'Question: Luke has ten balls. He gives three to his brother.\\n'\n",
    "lm += 'How many balls does he have left?\\n'\n",
    "lm += 'Answer: ' + gen(stop='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>Question: Emily Johnson, a 42-year-old woman, visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month. Alongside the discomfort, she experienced sporadic nausea and vomiting. During the examination, a diagnostic ultrasound revealed gallstones in the gallbladder, indicative of cholecystitis.\n",
       "Your goal is to extract structured information from the question. Please extract Age.\n",
       "Answer:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> Johnson</span></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lm = medicine + 'Question: Emily Johnson, a 42-year-old woman, visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month. Alongside the discomfort, she experienced sporadic nausea and vomiting. During the examination, a diagnostic ultrasound revealed gallstones in the gallbladder, indicative of cholecystitis.\\n'\n",
    "lm += \"Your goal is to extract structured information from the question. Please extract Age.\\n\" #Sex, Diagnostic Procedure, Symptomes.\n",
    "lm += 'Answer: ' + gen(stop='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A 28-year-old previously healthy man presented with a 6-week history of palpitations. The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea. Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings. An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway. Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2). The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead). Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D). The patient underwent an electrophysiologic study with mapping of the accessory pathway, followed by radiofrequency ablation (interruption of the pathway using the heat generated by electromagnetic waves at the tip of an ablation catheter). His post-ablation ECG showed a prolonged PR interval and an odd “second” QRS complex in leads III, aVF and V2–V4 (Fig.1Bottom), a consequence of abnormal impulse conduction in the “atrialized” right ventricle. The patient reported no recurrence of palpitations at follow-up 6 months after the ablation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>### Your goal is to extract structured information from the user&#x27;s input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Age, Sex, Symptom. Do not add any extra entities.\n",
       "\n",
       "### Input: Erik, 53. Visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month.\n",
       "\n",
       " ### Output:\n",
       "<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> {&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>Age</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>5</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>3</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;],</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>S</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ex</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>man</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;],</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>Sym</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>pt</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>om</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>es</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>cal</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ating</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>p</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ain</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>pers</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>istent</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;]</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>}</span></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lm = medicine + \"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Age, Sex, Symptom. Do not add any extra entities.\\n\\n### Input: Erik, 53. Visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month.\\n\\n ### Output:\"\n",
    "lm += \"\\n \" + gen(\"output\", max_tokens=250, stop=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(lm[\"output\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Invalid file path or buffer object type: <class 'NoneType'>",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[66], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_json\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlm\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43moutput\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/util/_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[0;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/util/_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    330\u001b[0m     )\n\u001b[0;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/io/json/_json.py:733\u001b[0m, in \u001b[0;36mread_json\u001b[0;34m(path_or_buf, orient, typ, dtype, convert_axes, convert_dates, keep_default_dates, numpy, precise_float, date_unit, encoding, encoding_errors, lines, chunksize, compression, nrows, storage_options)\u001b[0m\n\u001b[1;32m    730\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m convert_axes \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m orient \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtable\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    731\u001b[0m     convert_axes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m--> 733\u001b[0m json_reader \u001b[38;5;241m=\u001b[39m \u001b[43mJsonReader\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    734\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    735\u001b[0m \u001b[43m    \u001b[49m\u001b[43morient\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morient\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    736\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtyp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    737\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    738\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconvert_axes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_axes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    739\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconvert_dates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_dates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    740\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeep_default_dates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeep_default_dates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    741\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnumpy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnumpy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    742\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprecise_float\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprecise_float\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    743\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdate_unit\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_unit\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    744\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    745\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlines\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlines\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    746\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    747\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    748\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnrows\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnrows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    749\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    750\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding_errors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding_errors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    751\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    753\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize:\n\u001b[1;32m    754\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m json_reader\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/io/json/_json.py:818\u001b[0m, in \u001b[0;36mJsonReader.__init__\u001b[0;34m(self, filepath_or_buffer, orient, typ, dtype, convert_axes, convert_dates, keep_default_dates, numpy, precise_float, date_unit, encoding, lines, chunksize, compression, nrows, storage_options, encoding_errors)\u001b[0m\n\u001b[1;32m    815\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlines:\n\u001b[1;32m    816\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnrows can only be passed if lines=True\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 818\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_data_from_filepath\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    819\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdata \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_preprocess_data(data)\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/io/json/_json.py:858\u001b[0m, in \u001b[0;36mJsonReader._get_data_from_filepath\u001b[0;34m(self, filepath_or_buffer)\u001b[0m\n\u001b[1;32m    851\u001b[0m filepath_or_buffer \u001b[38;5;241m=\u001b[39m stringify_path(filepath_or_buffer)\n\u001b[1;32m    852\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    853\u001b[0m     \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(filepath_or_buffer, \u001b[38;5;28mstr\u001b[39m)\n\u001b[1;32m    854\u001b[0m     \u001b[38;5;129;01mor\u001b[39;00m is_url(filepath_or_buffer)\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;129;01mor\u001b[39;00m is_fsspec_url(filepath_or_buffer)\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;129;01mor\u001b[39;00m file_exists(filepath_or_buffer)\n\u001b[1;32m    857\u001b[0m ):\n\u001b[0;32m--> 858\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    859\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m        \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    863\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    864\u001b[0m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding_errors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    865\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    866\u001b[0m     filepath_or_buffer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n\u001b[1;32m    867\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m (\n\u001b[1;32m    868\u001b[0m     \u001b[38;5;28misinstance\u001b[39m(filepath_or_buffer, \u001b[38;5;28mstr\u001b[39m)\n\u001b[1;32m    869\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m filepath_or_buffer\u001b[38;5;241m.\u001b[39mlower()\u001b[38;5;241m.\u001b[39mendswith(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    872\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m file_exists(filepath_or_buffer)\n\u001b[1;32m    873\u001b[0m ):\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/io/common.py:713\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    710\u001b[0m     codecs\u001b[38;5;241m.\u001b[39mlookup_error(errors)\n\u001b[1;32m    712\u001b[0m \u001b[38;5;66;03m# open URLs\u001b[39;00m\n\u001b[0;32m--> 713\u001b[0m ioargs \u001b[38;5;241m=\u001b[39m \u001b[43m_get_filepath_or_buffer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    714\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    715\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    716\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    717\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    718\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    719\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    721\u001b[0m handle \u001b[38;5;241m=\u001b[39m ioargs\u001b[38;5;241m.\u001b[39mfilepath_or_buffer\n\u001b[1;32m    722\u001b[0m handles: \u001b[38;5;28mlist\u001b[39m[BaseBuffer]\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/pandas/io/common.py:451\u001b[0m, in \u001b[0;36m_get_filepath_or_buffer\u001b[0;34m(filepath_or_buffer, encoding, compression, mode, storage_options)\u001b[0m\n\u001b[1;32m    447\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\n\u001b[1;32m    448\u001b[0m     \u001b[38;5;28mhasattr\u001b[39m(filepath_or_buffer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mread\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(filepath_or_buffer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwrite\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    449\u001b[0m ):\n\u001b[1;32m    450\u001b[0m     msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInvalid file path or buffer object type: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(filepath_or_buffer)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m--> 451\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m    453\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m IOArgs(\n\u001b[1;32m    454\u001b[0m     filepath_or_buffer\u001b[38;5;241m=\u001b[39mfilepath_or_buffer,\n\u001b[1;32m    455\u001b[0m     encoding\u001b[38;5;241m=\u001b[39mencoding,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    458\u001b[0m     mode\u001b[38;5;241m=\u001b[39mmode,\n\u001b[1;32m    459\u001b[0m )\n",
      "\u001b[0;31mValueError\u001b[0m: Invalid file path or buffer object type: <class 'NoneType'>"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.read_json(lm[\"output\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>### Your goal is to extract structured information from the user&#x27;s input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Age, Sex, Symptom. Do not add any extra entities.\n",
       "\n",
       "### Input: Erik, 53. Visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month.\n",
       "\n",
       " ### Output:</pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lm = medicine + \"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Age, Sex, Symptom. Do not add any extra entities.\\n\\n### Input: Erik, 53. Visited the clinic reporting persistent abdominal pain in the upper right quadrant, escalating over the past month.\\n\\n ### Output:\"\n",
    "\n",
    "lm = gen(max_tokens=250)\n",
    "#lm += \"\\n \" + gen(max_tokens=250, stop=\"\\n\")\n",
    "#print(str(lm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>### Your goal is to extract structured information from the user&#x27;s input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Sex, Age. Do not add any extra entities.\n",
       "\n",
       "### Input: A 28-year-old previously healthy man presented with a 6-week history of palpitations. The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea. Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings. An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway. Transthoracic echocardiography demonstrated the presence of Ebstein&#x27;s anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2). The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead). Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
       "\n",
       " ### Output:\n",
       "<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> {&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>S</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ex</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>man</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;],</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>Age</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>2</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>8</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>-</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>year</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>-</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>old</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;]</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>}</span></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lm = medicine + \"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Sex, Age. Do not add any extra entities.\\n\\n### Input: A 28-year-old previously healthy man presented with a 6-week history of palpitations. The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea. Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings. An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway. Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2). The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead). Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\\n\\n ### Output:\"\n",
    "lm += \"\\n \" + gen(max_tokens=250, stop=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>\n",
       "### Your goal is to extract structured information from the user&#x27;s input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entity: Diagnostic_procedure. Do not add any extra entities. \n",
       "\n",
       "### Input: A 36-year-old female presented at the emergency department with aggravating right upper abdominal pain for 2 hours. The patient was diagnosed hepatitis B virus (HBV) carrier for several years and non-alcoholics. No other specific personal and familial medical history was noted. Initial blood pressure was 100/60 mmHg, pulse rate 70/min, respiration rate 20/min, body temperature 37.\n",
       "\n",
       "### Output:\n",
       "\n",
       "<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> {&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>Di</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>agnostic</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>_</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>pro</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>cedure</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>blo</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>od</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> pressure</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>body</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> temperature</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>p</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ulse</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> rate</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>resp</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>iration</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> rate</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;]</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>}</span></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lm = medicine + \"\"\"\n",
    "### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entity: Diagnostic_procedure. Do not add any extra entities. \n",
    "\n",
    "### Input: A 36-year-old female presented at the emergency department with aggravating right upper abdominal pain for 2 hours. The patient was diagnosed hepatitis B virus (HBV) carrier for several years and non-alcoholics. No other specific personal and familial medical history was noted. Initial blood pressure was 100/60 mmHg, pulse rate 70/min, respiration rate 20/min, body temperature 37.\n",
    "\n",
    "### Output:\n",
    "\"\"\"\n",
    "\n",
    "lm += \"\\n \" + gen(max_tokens=2500, stop=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>\n",
       "### Your goal is to extract structured information from the user&#x27;s input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entity: Sign_symptom. Do not add any extra entities.\n",
       "\n",
       "### Input: Patient Name: John Doe Age: 45 Gender: Male Date of Admission: 2023-03-01 Diagnosis: Right-sided ischemic stroke with resulting hemiparesis and aphasia. Medical History: Hypertension, Type 2 Diabetes Mellitus. Rehabilitation Goals: Improve motor function in right arm and leg, enhance speech and communication abilities, increase independence in daily activities.\n",
       "\n",
       "### Output:\n",
       "\n",
       "<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> {&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>Sign</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>_</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>sym</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>pt</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>om</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;:</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> [&quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>hem</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ip</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ares</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>is</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>hem</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ip</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ares</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>is</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>hy</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>pert</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ension</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>isch</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>em</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ic</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> stroke</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ap</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>has</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ia</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ap</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>has</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ia</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>in</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>dep</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>end</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ence</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>mot</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>or</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> function</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>mot</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>or</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> function</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>spe</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ech</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;,</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>spe</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ech</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>&quot;]</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>}</span></pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lm = medicine + \"\"\"\n",
    "### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entity: Sign_symptom. Do not add any extra entities.\n",
    "\n",
    "### Input: Patient Name: John Doe Age: 45 Gender: Male Date of Admission: 2023-03-01 Diagnosis: Right-sided ischemic stroke with resulting hemiparesis and aphasia. Medical History: Hypertension, Type 2 Diabetes Mellitus. Rehabilitation Goals: Improve motor function in right arm and leg, enhance speech and communication abilities, increase independence in daily activities.\n",
    "\n",
    "### Output:\n",
    "\"\"\"\n",
    "\n",
    "lm += \"\\n \" + gen(max_tokens=150, stop=\"\\n\", temperature=0.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style='margin: 0px; padding: 0px; vertical-align: middle; padding-left: 8px; margin-left: -8px; border-radius: 0px; border-left: 1px solid rgba(127, 127, 127, 0.2); white-space: pre-wrap; font-family: ColfaxAI, Arial; font-size: 15px; line-height: 23px;'>    &quot;### Your goal is to extract structured information from the user&#x27;s input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Sex, Age. Do not add any extra entities.\n",
       "\n",
       "### Input: A previously healthy man presented with a 6-week history of palpitations. The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea. Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings. An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway. Transthoracic echocardiography demonstrated the presence of Ebstein&#x27;s anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2). The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead). Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\n",
       "\n",
       " ### Output:&quot;\n",
       "    ```json\n",
       "    {\n",
       "        \n",
       "        &quot;age&quot;:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>3</span>4&quot;,\n",
       "        &quot;sex&quot;:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>male</span>&quot;,\n",
       "        &quot;Biological_structure&quot;:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>T</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ran</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>st</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>hor</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ac</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ic</span>&quot;,\n",
       "        &quot;Sign_symptom&quot;:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>b</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ub</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>bles</span>&quot;,\n",
       "        &quot;Diagnostic_procedure&quot;:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>Con</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>tr</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ast</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> e</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ch</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>oc</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ardi</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ography</span>&quot;,\n",
       "        &quot;Lab_value&quot;:<span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> &quot;</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>W</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ol</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ff</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>–</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> Park</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>inson</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>–</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> White</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> pre</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>-</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>ex</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>c</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'>itation</span><span style='background-color: rgba(0.0, 165.0, 0, 0.15); border-radius: 3px;' title='1.0'> pattern</span>&quot;,\n",
       "    }```</pre>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "AttributeError",
     "evalue": "'generator' object has no attribute '_inplace_append'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[74], line 17\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m lm\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m#a = time.time()\u001b[39;00m\n\u001b[0;32m---> 17\u001b[0m lm \u001b[38;5;241m=\u001b[39m \u001b[43mlm\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcharacter_maker\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmedicine\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;66;03m#time.time() - a\u001b[39;00m\n\u001b[1;32m     19\u001b[0m \u001b[38;5;66;03m#\"age\": {gen('age', regex='[0-9]+', stop=',')},\u001b[39;00m\n",
      "File \u001b[0;32m~/code/medical-llm/venv/lib/python3.9/site-packages/guidance/models/_model.py:311\u001b[0m, in \u001b[0;36mModel.__add__\u001b[0;34m(self, value)\u001b[0m\n\u001b[1;32m    308\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mA guidance function did not return a model object! Did you forget to return the new lm at the end of your function?\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    310\u001b[0m \u001b[38;5;66;03m# this flushes the display\u001b[39;00m\n\u001b[0;32m--> 311\u001b[0m \u001b[43mout\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_inplace_append\u001b[49m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    313\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m out\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'generator' object has no attribute '_inplace_append'"
     ]
    }
   ],
   "source": [
    "#@guidance\n",
    "def character_maker(lm):\n",
    "    lm += f\"\"\"\\\n",
    "    \"### Your goal is to extract structured information from the user's input that matches json format. When extracting information please make sure it matches the type information exactly. Extract the following entities: Sex, Age. Do not add any extra entities.\\n\\n### Input: A previously healthy man presented with a 6-week history of palpitations. The symptoms occurred during rest, 2–3 times per week, lasted up to 30 minutes at a time and were associated with dyspnea. Except for a grade 2/6 holosystolic tricuspid regurgitation murmur (best heard at the left sternal border with inspiratory accentuation), physical examination yielded unremarkable findings. An electrocardiogram (ECG) revealed normal sinus rhythm and a Wolff– Parkinson– White pre-excitation pattern (Fig.1: Top), produced by a right-sided accessory pathway. Transthoracic echocardiography demonstrated the presence of Ebstein's anomaly of the tricuspid valve, with apical displacement of the valve and formation of an “atrialized” right ventricle (a functional unit between the right atrium and the inlet [inflow] portion of the right ventricle) (Fig.2). The anterior tricuspid valve leaflet was elongated (Fig.2C, arrow), whereas the septal leaflet was rudimentary (Fig.2C, arrowhead). Contrast echocardiography using saline revealed a patent foramen ovale with right-to-left shunting and bubbles in the left atrium (Fig.2D).\\n\\n ### Output:\"\n",
    "    ```json\n",
    "    {{\n",
    "        \n",
    "        \"age\": \"{select(options=['Not available','55', '34', 'None', 'Not available'], name='sex')}\",\n",
    "        \"sex\": \"{select(options=['male', 'female', ''], name='sex')}\",\n",
    "        \"Biological_structure\": \"{gen(stop='\"')}\",\n",
    "        \"Sign_symptom\": \"{gen('Sign_symptom', stop='\"')}\",\n",
    "        \"Diagnostic_procedure\": \"{gen('Diagnostic_procedure', stop='\"')}\",\n",
    "        \"Lab_value\": \"{gen('Lab_value', stop='\"')}\",\n",
    "    }}```\"\"\"\n",
    "    return lm\n",
    "#a = time.time()\n",
    "lm = lm + character_maker(medicine)\n",
    "#time.time() - a\n",
    "#\"age\": {gen('age', regex='[0-9]+', stop=',')},"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inputs = tokenizer(prompt, return_tensors=\"pt\", add_special_tokens=False).input_ids.to(model.device)\n",
    "outputs = model.generate(input_ids=inputs, max_length=2000)[0]\n",
    "\n",
    "answer_start = int(inputs.shape[-1])\n",
    "pred = tokenizer.decode(outputs[answer_start:], skip_special_tokens=True)\n",
    "print(pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datasets import load_from_disk\n",
    "\n",
    "hg_dataset = load_from_disk(\"../data/datasetTestShuffledEntire\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A 18-year-old man presented with a history of a palpable lump in the presternal area, which was found incidentally after weight reduction.\\nThe patient had no other relevant medical or trauma history.\\nOn physical examination, the lesion was found to be a nonmovable, firm mass with no tenderness or associated skin changes, detected at the midline position over the sternum (at the manubrium level).\\nThere was no visible fistulous opening or discharge from the lesion.\\nOn ultrasonography, we detected a well-circumscribed, oval, anechoic mass, with posterior acoustic enhancement, that measured about 3.3\\u200a×\\u200a1.7\\u200a×\\u200a3.1\\u200acm, and was located in the subcutaneous fat layer over the sternum.\\nIn the dependent portion of the mass was an internal, well-circumscribed, heterogeneously hypoechoic, egg-shaped lesion (Fig.1A and B) showing a movement according to patient movement.\\nThe mass could be compressed using the linear transducer (Fig.2A and B).\\nA color Doppler study showed no vascularity within the cystic mass or the internal hypoechoic lesion (Fig.1C).\\nSurgical excision of the mass was performed without postoperative complications.\\nGrossly, the excised mass was a well-defined, ovoid, cystic mass gray-tan in color.\\nOn section, it was found to be a unilocular cyst filled with whitish mucous material.\\nMicroscopically, the mass was lined with ciliated pseudostratified columnar epithelium suggestive of respiratory type-mucosa (Fig.3).\\nThere was nonspecific collagenous fibrosis around the cyst.\\nThe differential diagnosis of subcutaneous cyst included epidermal inclusion cyst, thyroglossal duct cyst, branchial cleft cyst, and dermoid cyst.\\nThere was no histological evidence of squamous epithelium, keratin, thyroid tissue, or skin appendage in the cystic wall of the present case.\\nThe mass was diagnosed as a bronchogenic cyst.\\n'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg_dataset[\"input\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "gr_examples = [[example[\"input\"], example[\"output\"]] for example in hg_dataset]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 68-year-old man was referred by his optometrist to HES with suspected LTG due to repeatedly irregular visual field test results, advanced optic disc cupping, normal intraocular pressures (IOPs) and a family history of glaucoma.\n",
      "The patient subjectively felt that vision in his ‘good’ left eye (LE), which normally had a visual acuity of 6/6 N5, started to deteriorate 6months earlier; at the point of referral it was best corrected to 6/7.5 N6.\n",
      "His right eye (RE) was known to be amblyopic with a visual acuity of 6/18 N12.\n",
      "His medical history included considerable risk factors for systemic vasculopathy, such as hypertension, hypercholesterolaemia, 50 pack-years of smoking and type 2 diabetes with no diabetic retinopathy.\n",
      "Despite detailed questioning, he denied any new systemic symptoms apart from experiencing increased lethargy.\n",
      "Clinical examination at HES revealed advanced bilateral cupped optic discs with a cup-to-disc ratio right 0.9 (90%) and left 0.8 (80%; figure 1).\n",
      "Furthermore, the findings from the referring optometrist of equal and reactive pupils, normal eye movements and IOPs in mid-teens were confirmed.\n",
      "Computerised perimetry showed constricted visual fields in both eyes, albeit with considerably raised false-negative error rates in excess of 30% (figure 2A).\n",
      "Since this can be a result of poor concentration and attention span with testing compounded, especially with an amblyopic RE, the field test was repeated 4weeks later with similar results (figure 2B).\n",
      "At this visit, the patient experienced further deterioration of his ‘good’ LE.\n",
      "Visual acuities measured 6/24 (loss of one Snellen line) N36 in his amblyopic RE, and 6/18 (loss of two Snellen lines) N36 in his LE, which now also showed depressed colour vision.\n",
      "An urgent scan of the patient's head and orbit was requested to rule out a compressive lesion and blood analysis was carried out to screen for other causes of optic neuropathy including vitamin B12, folate levels and inflammatory markers.\n",
      "MRI revealed a large mass centred at the pituitary fossa measuring approximately 4.5cm×5.1cm× 4.8cm (figure 3A–C) extending into the suprasellar cistern and infrasellar sphenoid sinuses.\n",
      "The optic chiasm and both optic nerves were displaced superiorly and compressed on both sides.\n",
      "The mass showed heterogeneous T1 and T2 signals with multiple fluid levels most likely due to previous haemorrhage and low-signal components suggesting calcifications.\n",
      "The appearance was probably due to a craniopharyngioma or haemorrhagic pituitary adenoma.\n",
      "The patient was referred urgently to the neurosurgery unit and was found to have markedly raised prolactin levels of greater than 200000mIU/L, confirming the diagnosis of a giant prolactinoma.\n",
      "In cooperation with the local endocrinology service, he was started on cabergoline 500µg and dexamethasone 8mg.\n",
      "Within a week this led to a dramatic improvement of the patient's visual fields and acuities to 6/18 RE and to 6/9 LE.\n",
      "His prolactin levels showed a similar dramatic response with levels coming down to 336mIU/L.\n",
      "Three months later his visual acuity in his RE and LE were 6/18 and 6/5, respectively, and visual field examination showed striking recovery with almost normal results (figure 2C).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(gr_examples[0][0].replace(\" \", \"\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
